<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="ru_RU">
<context>
    <name>AlignToReferenceBlastDialog</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="20"/>
        <source>Input data</source>
        <translation>Входные данные</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="26"/>
        <source>Reference</source>
        <translation>Референсная последовательность</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="14"/>
        <source>Map Sanger Reads to Reference</source>
        <translation>Выравнивание данных секвенирования по Сэнгеру</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="38"/>
        <source>Input a file with a reference sequence</source>
        <translation>Укажите файл с референсной последовательностью</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="45"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="169"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="54"/>
        <source>Reads</source>
        <translation>Прочтения</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="79"/>
        <source>Add</source>
        <translation>Добавить</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="86"/>
        <source>Remove</source>
        <translation>Удалить</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="98"/>
        <source>Settings</source>
        <translation>Настройки</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="114"/>
        <source>Mapping min similarity</source>
        <translation>Минимальное сходство при выравнивании</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="134"/>
        <source>Trimming quality threshold</source>
        <translation>Порог качества для обрезания концов</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="141"/>
        <source>Read name in result alignment</source>
        <translation>Имя прочтения в выравнивании</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="162"/>
        <source>Input the result file path</source>
        <translation>Укажите путь до выходного файла</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="121"/>
        <source>%</source>
        <translation>%</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="154"/>
        <source>Result alignment</source>
        <translation>Сохранить в</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.ui" line="178"/>
        <source>Add to project</source>
        <translation>Добавить в проект</translation>
    </message>
</context>
<context>
    <name>BlastAllSupportDialog</name>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="20"/>
        <source>Request to Local BLAST Database</source>
        <translation>Запрос к локальной базе данных BLAST</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="48"/>
        <source>General options</source>
        <translation>Основные параметры</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="62"/>
        <source>Select search:</source>
        <translation>Тип поиска:</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="70"/>
        <source>blastn</source>
        <extracomment>qwe</extracomment>
        <translation>blastn</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="75"/>
        <source>blastp</source>
        <translation>blastp</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="85"/>
        <source>blastx</source>
        <translation>blastx</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="90"/>
        <source>tblastn</source>
        <translation>tblastn</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="95"/>
        <source>tblastx</source>
        <translation>tblastx</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="103"/>
        <source>Search for short, nearly exact matches</source>
        <translation>Поиск коротких последовательностей</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="110"/>
        <source>Expectation value </source>
        <translation>Математическое ожидание</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="117"/>
        <source>Megablast</source>
        <translation>Megablast</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="134"/>
        <source>Off</source>
        <translation>Выключить</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="274"/>
        <source>Number of CPUs being used</source>
        <translation>Использовать процессоров</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="318"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="544"/>
        <source>Word size</source>
        <translation>Размер слова</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="616"/>
        <source>Gap costs</source>
        <translation>Вес пробела</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="421"/>
        <source>Match scores</source>
        <translation>Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="480"/>
        <source>1 -3</source>
        <translation>1 -3</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="171"/>
        <source>Both strands</source>
        <translation>Обе цепи</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="181"/>
        <source>Direct</source>
        <translation>Прямая</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="188"/>
        <source>Complement</source>
        <translation>Комплементарная</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="485"/>
        <source>1 -4</source>
        <translation>1 -4</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="490"/>
        <source>1 -2</source>
        <translation>1 -2</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="495"/>
        <source>1 -1</source>
        <translation>1 -1</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="500"/>
        <source>2 -7</source>
        <translation>2 -7</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="505"/>
        <source>2 -5</source>
        <translation>2 -5</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="510"/>
        <source>2 -3</source>
        <translation>2 -3</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="515"/>
        <source>4 -5</source>
        <translation>4 -5</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="520"/>
        <source>5 -4</source>
        <translation>5 -4</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="463"/>
        <source>Matrix</source>
        <translation>Матрица</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="80"/>
        <source>gpu-blastp</source>
        <translation>gpu-blastp</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="429"/>
        <source>BLOSUM62</source>
        <translation>BLOSUM62</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="434"/>
        <source>BLOSUM45</source>
        <translation>BLOSUM45</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="439"/>
        <source>BLOSUM80</source>
        <translation>BLOSUM80</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="444"/>
        <source>PAM30</source>
        <translation>PAM30</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="449"/>
        <source>PAM70</source>
        <translation>PAM70</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="408"/>
        <source>Service</source>
        <translation>Сервис</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="124"/>
        <source>Max hits:</source>
        <translation>Макс. число совпадений:</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="371"/>
        <source>plain</source>
        <translation>простой</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="376"/>
        <source>psi</source>
        <translation>psi</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="381"/>
        <source>phi</source>
        <translation>phi</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="531"/>
        <source>PHI pattern</source>
        <translation>Шаблон PHI</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="600"/>
        <source>Threshold for extending hits</source>
        <translation>Порог для расширенных совпадений</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="603"/>
        <source>Threshold</source>
        <translation>Порог</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="623"/>
        <source>Composition-based statistics</source>
        <translation>Статистика состава</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="336"/>
        <source>D or d: default (equivalent to 2 )</source>
        <translation>D или d: по умолчанию (равно 2)</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="341"/>
        <source>0 or F or f: No composition-based statistics</source>
        <translation>0 или F или f: Нет статистики состава</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="346"/>
        <source>1: Composition-based statistics as in NAR 29:2994-3005, 2001</source>
        <translation>1: Статистика состава как в NAR 29:2994-3005, 2001</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="351"/>
        <source>2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties</source>
        <translation>2 или T или t : Статистика как в in Bioinformatics 21:902-911, 2005, основанная на свойствах последовательности</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="356"/>
        <source>3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally</source>
        <translation>3: Статистика как в Bioinformatics 21:902-911, 2005, безусловная</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="637"/>
        <source>Filters</source>
        <translation>Фильтры</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="646"/>
        <source>Low complexity filter</source>
        <translation>Фильтр низкой сложности</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="653"/>
        <source>Human repeats filter</source>
        <translation>Фильтр человеческих повторов</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="663"/>
        <source>Masks</source>
        <translation>Маски</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="672"/>
        <source>Mask for lookup table only</source>
        <translation>Только маска таблицы поиска</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="682"/>
        <source>Mask lower case letters</source>
        <translation>Маска символов нижн. регистра</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="717"/>
        <source>Extension options</source>
        <translation>Возможности расширения</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="731"/>
        <source>X dropoff value (in bits)</source>
        <translation>X dropoff значение (в битах)</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="740"/>
        <source>For gapped alignment</source>
        <translation>Для выравниваний с пробелами</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="754"/>
        <source>For ungapped extensions</source>
        <translation>Для расширений без пробелов</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="761"/>
        <source>For final gapped alignment</source>
        <translation>Для финального выравнивания</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="788"/>
        <source>Multiple Hits window size</source>
        <translation>Множественные пересечения окна</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastAllSupportDialog.ui" line="800"/>
        <source>Perform gapped alignment</source>
        <translation>Выполнить выравнивание</translation>
    </message>
</context>
<context>
    <name>BlastDBCmdDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="32"/>
        <source>Fetch Sequences from BLAST Database</source>
        <translation>Извлечь последовательность из базы данных BLAST</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="44"/>
        <source>Query ID</source>
        <translation>Идентификатор запроса</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="121"/>
        <source>Select database</source>
        <translation>Выберите базу данных</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="71"/>
        <source>Output path</source>
        <translation>Выходной путь</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="99"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.ui" line="158"/>
        <source>Add to project</source>
        <translation>Добавить в проект</translation>
    </message>
</context>
<context>
    <name>BlastDBSelectorWidget</name>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidget.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidget.ui" line="43"/>
        <source>Database path</source>
        <translation>Путь до базы данных</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidget.ui" line="61"/>
        <source>Database file</source>
        <translation>Файл базы данных</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidget.ui" line="70"/>
        <source>Base name for BLAST DB files</source>
        <translation>Базовое имя для файлов базы данных BLAST</translation>
    </message>
</context>
<context>
    <name>Bowtie2Settings</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="41"/>
        <source>Parameters</source>
        <translation>Параметры</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="60"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="76"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, Bowtie2 performs end-to-end read alignment. That is, it searches for alignments involving all of the read characters. This is also called an &amp;quot;untrimmed&amp;quot; or &amp;quot;unclipped&amp;quot; alignment.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the --local option is specified, Bowtie2 performs local read alignment. In this mode, Bowtie2 might &amp;quot;trim&amp;quot; or &amp;quot;clip&amp;quot; some read characters from one or both ends of the alignment if doing so maximizes the alignment score.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;По умолчанию, Bowtie2 выполняет end-to-end выравнивание. Т.е. он ищет выравнивания с с участием всех символов. Это также называется &amp;quot;необрезное&amp;quot; или &amp;quot;необрезаемое&amp;quot; выравнивание.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Когда опция --local установлена, Bowtie2 выполняет локальное выравнивание. В этом режиме Bowtie2 может &amp;quot;обрезать&amp;quot; или &amp;quot;отсечь&amp;quot; некоторые символы одного или обоих концов выравнивания если это увеличит качество выравнивания.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="63"/>
        <source>Mode</source>
        <translation>Режим</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="86"/>
        <source>--end-to-end</source>
        <translation>--end-to-end</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="91"/>
        <source>--local</source>
        <translation>--local</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="106"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="174"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Pads&amp;quot; dynamic programming problems by specified number of columns on either side to allow gaps.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;&amp;quot;Подкладки&amp;quot; проблем динамического программирования по указанному числу столбцов с обеих сторон для разрешения пробелов.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="109"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation>Добавляет столбцы, чтобы допустить пробелы (--dpad)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="119"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="187"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1. Setting this higher makes alignment slower (often much slower) but increases sensitivity.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Задает количество несовпадений с разрешенным в выравнивании. Может быть 0 или 1. Повышение этого параметра делает выранивание более медленным (чаще всего очень медленным) но повышает точность.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="132"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="197"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Запуск определенного количество параллельных потоков поиска. Потоки будут работать на отдельных процессорах/ ядрах и синхронизоваться при разборе ридов и выводе выравнивания.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="148"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="252"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Disallow gaps within specified number of positions of the beginning or end of the read.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Запретить пробелы в указанных позициях в начале или в конце рида.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="161"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="229"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Use specified value as the seed for pseudo-random number generator.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Использовать указанное значение в качестве источника для псевдо-генератора случайных чисел.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="164"/>
        <source>Seed (--seed)</source>
        <translation>Затравка (--seed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="190"/>
        <source>Number of mismatches</source>
        <translation>Число несоответствий</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="200"/>
        <source>Threads</source>
        <translation>Потоки</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="210"/>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="239"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Устанавливает длину подстрок затравки для выравнивания. Меньшие значения делают выравнивание медленнее, но более чувствительны.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="242"/>
        <source>Seed length (--L)</source>
        <translation>Длина затравки (--L)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="255"/>
        <source>Disallow gaps (--gbar)</source>
        <translation>Запретить пробелы (--gbar)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="281"/>
        <source>Flags</source>
        <translation>Флаги</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="293"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments for the constituent mates. This is called &amp;quot;mixed mode.&amp;quot; To disable mixed mode, set this option.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 runs a little faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Если Bowtie2 не сможет найти пару для выполнения парного выравнивания по умолчанию будет сделано непарное выравнивание. Это называется &amp;quot;смешанным режимом.&amp;quot; Чтобы отключить смешанный режим, установите эту опцию.&lt;/span&gt;&lt;/p&gt;&lt;p&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Bowtie2 запустится немного быстрее в смешанном режиме чем для парного выравнивания.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="296"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation>Только парное выравнивание (--no-mixed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="303"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;По умолчанию, Bowtie2 ищет несогласованные выравнивания, если он не может найти какие-либо согласованные выравнивания. Несогласованное выравнивание это выравнивание, в котором оба рида выровнены однозначно, но они не удовлетворяют условию парности. Это опция отключает это поведение.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="306"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation>Только согласованные выравнивания (--no-discordant)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="313"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the forward reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Если выбрана эта опция, Bowtie2 не будет пытаться выровнять прямую цепь.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="316"/>
        <source>No forward orientation (--nofw)</source>
        <translation>Не прямая ориентация (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="323"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If this option is selected, Bowtie2 will not attempt to align against the reverse-complement reference strand.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Если выбрана эта опция, Bowtie2 не будет пытаться выровнять обратно-комплементарную цепь.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="326"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>Не обратно-комплементарная ориентация (--norc)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="333"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Если один участник выравнивания перекрывается с другим, то это несогласованное выравни вание. По умолчанию: участники могут перекрываться в согласованном выравнивании.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="336"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation>Без перекрытий (--no-overlap)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="343"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the other in a concordant alignment.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-size:9pt;&quot;&gt;Если один участник выравнивания содержит другой, то это несогласованное выравнивание. По умолчанию одна часть выравнивания может содержать другую в согласованном выравнивании.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Settings.ui" line="346"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation>Без вложений (--no-contain)</translation>
    </message>
</context>
<context>
    <name>BowtieBuildSettings</name>
    <message>
        <location filename="../src/bowtie/BowtieBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieBuildSettings.ui" line="29"/>
        <source>Colorspace</source>
        <translation>Цвета</translation>
    </message>
</context>
<context>
    <name>BowtieSettings</name>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="41"/>
        <source>Parameters</source>
        <translation>Параметры</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="110"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;not just in the &amp;quot;seed&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The default is 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;rounding can be disabled with --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Максимальн допустимое общее значение качества всех несогласованных позиций ридов на протяжении всего выравнивания, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;не только в &amp;quot;затравках&amp;quot;. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;По умолчанию 70. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Подобно Maq, значения качества близки к 10 и максимальны к значению 30; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;округление может быть отключено при помощи --nomaqround.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="121"/>
        <source>Maq error (--maqerr)</source>
        <translation>Ошибка Maq (--maqerr)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="144"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The &amp;quot;seed length&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;i.e., the number of bases on the high-quality end of the read to which the -n ceiling applies.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;&amp;quot;Длина затравки&amp;quot;; &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;т.е., число оснований в высококачественном конце рида, для которого применяется -n.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Наименьшее допустимое значение 5, по умолчанию 28. Большие значения -l увеличивают скорость.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="176"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;The maximum insert size for valid paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;E.g. if -X 100 is specified and a paired-end alignment consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;that alignment is considered valid (as long as -I is also satisfied). &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;A 61-bp gap would not be valid in that case. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;If trimming options -3 or -5 are also used, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;the -X constraint is applied with respect to the untrimmed mates, not the trimmed mates.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Default: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Максимальный размер вставки для выравнивания paired-end. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Если -X 100 и выравнивание состоит из участков по 20 нуклеотидов и расстоянием между ними в 60 нуклеотидов, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;такое выравнивание считается действительным. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;В таком случае 61 пропуск не будет считаться допустимым. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Если также используются опции обрезки -3 или -5, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt; -X ограничение применяется к необрезанным участкам.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;По умолчанию: 250.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="215"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. 
Best-first search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost. 
Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases. 
If you receive an error message saying that chunk memory has been exhausted in --best mode, 
try adjusting this parameter up to dedicate more memory to the descriptors. 
Default: 64.</source>
        <translation>Количество мегабайт памяти для данного потока задается для хранения дескрипторов пути в --best режиме.
Поиск по первому наилучшему совпадению должен отслеживать множество путей одновременно, убедитесь, что расширение пути происходит с наименьшей стоимостью.
Bowtie пытается свести к минимуму воздействие памяти дескрипторов, но они по-прежнему могут сильно возрастать в некоторых случаях.
Если вы получаете сообщение об ошибке, говорящее, что участок памяти исчерпан в --best режиме,
попробуйте отрегулировать этот параметр и выделить больше памяти для дескрипторов.
Значение по умолчанию: 64.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="246"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator</source>
        <translation>Используйте &lt;int&gt; в качестве затравки для псевдо-случайного генератора</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="60"/>
        <source>Mode:  </source>
        <translation>Режим:  </translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="67"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;When the -n option is specified (which is the default), &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie determines which alignments are valid according to the following policy, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;which is similar to Maq&apos;s default policy.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;In -v mode, alignments may have no more than V mismatches, where V may be a number from 0 through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Когда указана опция -n, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;bowtie определяет какое выравнивание подходит под указанные условия, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;которое схоже с условиями Maq по умолчанию.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;В режиме-v, выравнивание может иметь не более чем V несовпадений, где V может быть числом от 0 до 3. Опция -v является взаимоисключающей с опцией -n&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="82"/>
        <source>-n mode</source>
        <translation>режим -n</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="87"/>
        <source>-v mode</source>
        <translation>режим -v</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="153"/>
        <source>Seed length (--seedlen)</source>
        <translation>Длина затравки (--seedlen)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="189"/>
        <source>Maximum of backtracks (--maxbts)</source>
        <translation>Максимальное число отступов (--maxbts)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="223"/>
        <source>Descriptors memory usage (--chunkmbs)</source>
        <translation>Использование памяти дескрипторами (--chunkmbs)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="249"/>
        <source>Seed (--seed)</source>
        <translation>Затравка (--seed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="266"/>
        <source>Threads</source>
        <translation>Потоки</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="300"/>
        <source>Flags</source>
        <translation>Флаги</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="312"/>
        <source>Colorspace (--color)</source>
        <translation>Цвета (--color)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="329"/>
        <source>No Maq rounding (--nomaqround)</source>
        <translation>Без округления Maq (--nomaqround)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="339"/>
        <source>No forward orientation (--nofw)</source>
        <translation>Не прямая ориентация (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="349"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>Не обратно-комплементарная ориентация (--norc)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="369"/>
        <source>Try as hard (--tryhard)</source>
        <translation>Упорно пытаться (--tryhard)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="383"/>
        <source>Best alignments (--best)</source>
        <translation>Лучшие выравнивания (--best)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="393"/>
        <source>All alignments (--all)</source>
        <translation>Все выравнивания (--all)</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="319"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq accepts quality values in the Phred quality scale, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but internally rounds values to the nearest 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;with a maximum of 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Maq принимает значения качества в масштабе качества Phred, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;но внутренние значения близки к 10, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;с максимумом около 30. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;По умолчанию, bowtie также вокруг этих значений. --nomaqround препятствует этому округлению в bowtie.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="336"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>Если установлена опция --nofw, bowtie  не будет пытаться выровнять прямую цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="346"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>Если установлена опция --norc, bowtie не будет пытаться выровнять обратно-комплементарную цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="356"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Try as hard as possible to find valid alignments when they exist, including paired-end alignments. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This is equivalent to specifying very high values for the --maxbts and --pairtries options.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is generally much slower than the default settings, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;but can be useful for certain problems. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;This mode is slower when (a) the reference is very repetitive, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) the reads are low quality, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;or (c) not many reads have valid alignments.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Пытаться пока это возможно найти подходящее выравнивание, когда они есть, включая paired-end выравнивания. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Это равносильно очень высоким значениям параметров --maxbts и --pairtries.&lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Этот режим более медленный, чем настройки по умолчанию, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;но может быть применим для некоторых случаев. &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;Этот режим медленнее, когда (a) референсная последовательность очень сильно состоит из повторяющихся участков, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;(b) риды низкого качества, &lt;/span&gt;&lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-size:8pt;&quot;&gt;или (c) не очень много ридов имеют подходящее выравнивание.&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="376"/>
        <source>Make Bowtie guarantee that reported singleton alignments are &quot;best&quot; in terms of stratum
(i.e. number of mismatches, or mismatches in the seed in the case of -n mode)
and in terms of the quality values at the mismatched position(s).
bowtie is somewhat slower when --best is specified.
           </source>
        <translation>Создать Bowtie гарантию того, что одиночные выравнивания являются &quot;лучшими&quot; в плане слоя
(т.е. число несовпадений, или несоответствия в затравке в -n режиме)
и с точки зрения значений качества на несогласованной позиции (ях).
bowtie медленнее, когда указан режим --best.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieSettings.ui" line="390"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation>Сообщить о всех допустимых выравниваниях для рида или пары.</translation>
    </message>
</context>
<context>
    <name>BwaBuildSettings</name>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="63"/>
        <source>Index algorithm (-a)</source>
        <translation>Алгоритм (-а)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="77"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="82"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="87"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaBuildSettings.ui" line="92"/>
        <source>is</source>
        <translation>is</translation>
    </message>
</context>
<context>
    <name>BwaIndexAlgorithmWarningReporter</name>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="77"/>
        <source>NOTE: &quot;is&quot; index algorithm is not supposed to work with reference sequences having size larger than 2 GB. In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;bwtsw&quot;</source>
        <translation>NOTE: &quot;is&quot; алгоритм не поддерживается для работы с реверенсными последовательностями, размер которых превышает 2 GB. В таком случае рекомендуется выбрать алгоритм &quot;bwtsw&quot;</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="84"/>
        <source>NOTE: &quot;bwtsw&quot; index algorithm is not supposed to work with reference sequences having size smaller than 10 MB.
In order to achieve stable BWA performance it is strongly recommend to set the index algorithm to &quot;is&quot;</source>
        <translation>NOTE: &quot;bwtsw&quot; алгоритм не поддерживается для работы с рефересными последовательностями, размер которых меньше 10 MB. В таком случае рекомендуется изменить алгоритм на &quot;is&quot;</translation>
    </message>
</context>
<context>
    <name>BwaMemSettings</name>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="567"/>
        <source>Penalty for a mismatch</source>
        <translation>Штраф за несовпадение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="449"/>
        <source>Score for a sequence match</source>
        <translation>Вес совпадения</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="551"/>
        <source>Mismatch penalty (-B)</source>
        <translation>Штраф за несовпадение (-B)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="484"/>
        <source>Gap open penalty (-O)</source>
        <translation>Штраф за открытие пробела (-O)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="625"/>
        <source>Penalty for an unpaired read pair</source>
        <translation>Штраф за непарные риды</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="746"/>
        <source>Minimum score to output</source>
        <translation>Минимальная оценка для вывода</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="609"/>
        <source>Penalty unpaired (-U)</source>
        <translation>Штраф за непарные риды (-U)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="529"/>
        <source>Penalty for clipping </source>
        <translation>Штраф за отсечение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="500"/>
        <source>Gap open penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="596"/>
        <source>Gap extension penalty; a gap of size k cost {-O} + {-E}*k</source>
        <translation>Штраф за расширение пробела; пробел размера k стоит {-O} + {-E}*k</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="513"/>
        <source>Penalty for clipping (-L)</source>
        <translation>Штраф за отсечение (-L)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="730"/>
        <source>Score threshold (-T)</source>
        <translation>Порог оценки (-T)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="132"/>
        <source>Min seed length (-k)</source>
        <translation>Минимальная длина затравки (-k)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="319"/>
        <source>Look for internal seeds inside a seed longer than {-k} * FLOAT</source>
        <translation>Поиск внутренние затравки внутри затравки длиннее чем {-k} * FLOAT</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="386"/>
        <source>Number of threads</source>
        <translation>Число потоков</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="148"/>
        <source>Off-diagonal X-dropoff</source>
        <translation>Снижение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="354"/>
        <source>Band width for banded alignment</source>
        <translation>Ширина полосы для выравнивания</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="271"/>
        <source>Band width (-w)</source>
        <translation>Ширина полосы (-w)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="202"/>
        <source>Minimum seed length</source>
        <translation>Минимальная длина затравки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="370"/>
        <source>Number of threads (-t)</source>
        <translation>Число потоков (-t)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="68"/>
        <source>Dropoff (-d)</source>
        <translation>Снижение (-d)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="186"/>
        <source>Skip seeds threshold (-c)</source>
        <translation>Порог затравки (-c)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="84"/>
        <source>Skip seeds with more than INT occurrences </source>
        <translation>Пропускает затравки с более чем с указанным количеством вхождений</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="338"/>
        <source>Drop chain threshold (-D)</source>
        <translation>Порог снижения цепи (-D)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="306"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain</source>
        <translation>Снижение цепи короче FLOAT фракции самого длинного перекрытия цепи</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="45"/>
        <source>Algorithm</source>
        <translation>Алгоритм</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="290"/>
        <source>Internal seeds length (-r)</source>
        <translation>Внутренняя длина затравки (-r)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="435"/>
        <source>Scoring Options</source>
        <translation>Настройки весов</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="580"/>
        <source>Gap extention penalty (-E)</source>
        <translation>Штраф за расширение пробела (-E)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="654"/>
        <source>Other</source>
        <translation>Другие</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="668"/>
        <source>Index algorithm (-a)</source>
        <translation>Алгоритм (-a)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="695"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="700"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="705"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="710"/>
        <source>is</source>
        <translation>is</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="465"/>
        <source>Matching score (-A)</source>
        <translation>Веса совпадений (-A)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="164"/>
        <source>Rounds of mate rescues (-m)</source>
        <translation>Округление за освобождение участника (-m)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="116"/>
        <source>Perform at most INT rounds of mate rescues for each read</source>
        <translation>Выполняет для большинства INT округлений освобождений участников для каждого рида</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="681"/>
        <source>Skip mate rescue</source>
        <translation>Пропустить освобождение участника</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="684"/>
        <source>Skip mate rescue (-S)</source>
        <translation>Пропустить освобождение участника (-S)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="233"/>
        <source>Skip pairing; mate rescue performed unless -S also in use</source>
        <translation>Пропустить спаривание; участник освобождается кроме случая когда -S также используется</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="236"/>
        <source>Skip pairing (-P)</source>
        <translation>Пропустить спаривание (-P)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemSettings.ui" line="410"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
</context>
<context>
    <name>BwaSettings</name>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="135"/>
        <source>  Best hits (-R)</source>
        <translation>Лучшие попытки (-R)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="336"/>
        <source>Max gap extensions (-e)</source>
        <translation>Максимальное расширение пробела (-e)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="346"/>
        <source>Indel offset (-i)</source>
        <translation>Смещение (-i)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="263"/>
        <source>Max long deletion extensions (-d)</source>
        <translation>Максимальная длина сокрашения пропуска (-d)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="194"/>
        <source>Seed length (-l)</source>
        <translation>Длина затравки (-l)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="482"/>
        <source>Max seed differences (-k)</source>
        <translation>Максимальное различие затравок (-k)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="283"/>
        <source>Max queue entries (-m)</source>
        <translation>Максимальное число записей в очереди (-m)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="356"/>
        <source>Enable long gaps</source>
        <translation>Допустимы длинные пробелы</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="171"/>
        <source>Missing prob (-n)</source>
        <translation>Доля потери (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="154"/>
        <source>Max #diff (-n)</source>
        <translation>Максимальное расстояние редактирования (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="42"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation>Долгосрочный штраф за длительный пробел (-L)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="241"/>
        <source>Advanced</source>
        <translation>Дополнительно</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="395"/>
        <source>Gap open penalty (-O)</source>
        <translation>Штраф за открытие пробела (-O)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="303"/>
        <source>Barcode length (-B)</source>
        <translation>Длина штрих-кода (-B)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="435"/>
        <source>Quality threshold (-q)</source>
        <translation>Порог качества (-q)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="375"/>
        <source>Mismatch penalty (-M)</source>
        <translation>Штраф за несовпадение (-M)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="415"/>
        <source>Gap extension penalty (-E)</source>
        <translation>Штраф за расширение пробела (-E)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="462"/>
        <source>Threads (-t)</source>
        <translation>Потоки (-t)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="49"/>
        <source>Non-iterative mode (-N)</source>
        <translation>Неитеративный режим (-N)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="72"/>
        <source>Max gap opens (-o)</source>
        <translation>Максимальное число пробелов (-o)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Основные параметры</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="85"/>
        <source>Index algorithm (-a)</source>
        <translation>Алгоритм (-а)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="102"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="107"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="112"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSettings.ui" line="117"/>
        <source>is</source>
        <translation>is</translation>
    </message>
</context>
<context>
    <name>BwaSwSettings</name>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Основные параметры</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="288"/>
        <source>Score threshold (divided by match score) (-T)</source>
        <translation>Порог оценки (-T)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="233"/>
        <source>Number of threads (-t)</source>
        <translation>Число потоков (-t)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="334"/>
        <source>Number of seeds to start rev alginment (-N)</source>
        <translation>Число затравок, чтобы начать обратное выравнивание (-N)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="311"/>
        <source>Z-best (-z)</source>
        <translation>Лучшие (-z)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="393"/>
        <source>Prefer hard clipping in SAM output (-H)</source>
        <translation>Строгое отсечение в выходном SAM файле (-H)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="256"/>
        <source>Size of chunk of reads (-s)</source>
        <translation>Размер куска ридов (-s)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="90"/>
        <source>Score for a match (-a)</source>
        <translation>Веса совпадений (-a)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="113"/>
        <source>Mismatch penalty (-b)</source>
        <translation>Штраф за несовпадение (-b)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="64"/>
        <source>autodetect</source>
        <translation>autodetect</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="69"/>
        <source>bwtsw</source>
        <translation>bwtsw</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="74"/>
        <source>div</source>
        <translation>div</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="79"/>
        <source>is</source>
        <translation>is</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="136"/>
        <source>Gap open penalty (-q)</source>
        <translation>Штраф за открытие пробела (-q)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="159"/>
        <source>Gap extention penalty (-r)</source>
        <translation>Штраф за расширение пробела (-r)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="182"/>
        <source>Band width (-w)</source>
        <translation>Ширина полосы (-w)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="208"/>
        <source>Index algorithm (-a)</source>
        <translation>Алгоритм (-а)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="380"/>
        <source>Mask level (-c)</source>
        <translation>Уровень маски (-c)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="347"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Number of seed to trigger reverse alignment&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Число затравок, чтобы начать обратное выравнивание&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaSwSettings.ui" line="405"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;MS Shell Dlg 2&apos;; font-size:8.25pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot;-qt-paragraph-type:empty; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px; font-size:8pt;&quot;&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
</context>
<context>
    <name>CAP3SupportDialog</name>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="14"/>
        <source>Contig Assembly with CAP3</source>
        <translation>Сборка контигов с помощью CAP3</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="27"/>
        <source>Base</source>
        <translation>Основные</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="33"/>
        <source>Input files (long DNA reads to assembly)</source>
        <translation>Входные файлы (длинные ДНК риды)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="42"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Sequences in FASTA or FASTQ formats are supported. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;The quality scores for FASTA sequences can be provided in an additional file. The file must be located in the same folder as the original sequences and have the same name as FASTA file, but another extention: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;Ubuntu&apos;; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Допустимы последовательности в форматах FASTA или FASTQ. &lt;/p&gt;
&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Значения качества для последовательностей в формате FASTA будут записаны в соответствующий файл. Этот файл будет сохранен в ту же папку, что и оригинальные последовательности и будет назван как файл FASTA, но с другим расширением: &lt;span style=&quot; font-weight:600; font-style:italic;&quot;&gt;.qual&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="54"/>
        <source>Add</source>
        <translation>Добавить</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="61"/>
        <source>Remove</source>
        <translation>Удалить</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="68"/>
        <source>Remove All</source>
        <translation>Удалить все</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="90"/>
        <source>Result contig</source>
        <translation>Результирующий контиг</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="99"/>
        <source>Path to result contig alignment in ACE format.</source>
        <translation>Путь до результирующего контига в формате ACE.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="125"/>
        <source>Clipping for poor regions</source>
        <translation>Отсечение пустых регионов</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="131"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation>Размер среза (-c)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="148"/>
        <source>Clipping range (-y)</source>
        <translation>Диапазон обрезания (-y)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="171"/>
        <source>Length and percent identity of an overlap</source>
        <translation>Длина и процент идентичности перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="177"/>
        <source>Overlap length cutoff (-o)</source>
        <translation>Обрезная длина перекрытия (-o)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="197"/>
        <source>Overlap percent identity cutoff (-p)</source>
        <translation>Обрезной процент идентичности перекрытия (-p)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="220"/>
        <source>Quality difference score of an overlap</source>
        <translation>Оценка качества различия перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="226"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation>Базовый срез для различий (-b)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="256"/>
        <source>Max qscore sum at differences (-d)</source>
        <translation>Максимальная сумма оценок для различий (-d)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="266"/>
        <source>Other parameters</source>
        <translation>Дополнительные параметры</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="272"/>
        <source>Max number of word matches (-t)</source>
        <translation>Максимальное число совпадений (-t)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="292"/>
        <source>Band expansion size (-a)</source>
        <translation>Размер расширения диапазона (-a)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="318"/>
        <source>Max gap length in any overlap (-f)</source>
        <translation>Максимальная длина пробела в перекрытии (-f)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="338"/>
        <source>Assembly reverse reads (-r)</source>
        <translation>Сборка обратных прочтений (-r)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="364"/>
        <source>Similarity score of an overlap</source>
        <translation>Оценка схожести перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="370"/>
        <source>Mismatch score factor (-n)</source>
        <translation>Фактор оценки несовпадения (-n)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="390"/>
        <source>Overlap similarity score cutoff (-s)</source>
        <translation>Срез оценки схожести перекрытия (-s)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="410"/>
        <source>Gap penalty factor (-g)</source>
        <translation>Фактор штрафа за пробел (-g)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="417"/>
        <source>Match score factor (-m)</source>
        <translation>Фактор оценки совпадений (-m)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="109"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.ui" line="119"/>
        <source>Advanced</source>
        <translation>Дополнительные</translation>
    </message>
</context>
<context>
    <name>ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="26"/>
        <source>Align with Clustal Omega</source>
        <translation>Выравнивание с помощью ClustalO</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="34"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="42"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="52"/>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="69"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="59"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="81"/>
        <source>Iteration</source>
        <translation>Итерации</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="91"/>
        <source>Number of iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="137"/>
        <source>Max number guidetree iterations</source>
        <translation>Максимальное число итераций дерева</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="180"/>
        <source>Max number of HMM iterations</source>
        <translation>Максимальное число итераций HMM</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="217"/>
        <source>Miscellaneous</source>
        <translation>Другое</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="227"/>
        <source>Number of CPUs being used</source>
        <translation>Использовать процессоров</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.ui" line="265"/>
        <source>Set options automatically</source>
        <translation>Установить параметры автоматически</translation>
    </message>
</context>
<context>
    <name>ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="32"/>
        <source>Align with ClustalW</source>
        <translation>Выравнивание с помощью ClustalW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="46"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="57"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="67"/>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="91"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="74"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="103"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="113"/>
        <source>Gap opening penalty</source>
        <translation>Штраф за открытие пробела</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="159"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пробел</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="208"/>
        <source>Weight matrix</source>
        <translation>Весовая матрица</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="247"/>
        <source>IUB</source>
        <translation>IUB</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="252"/>
        <source>CLUSTALW</source>
        <translation>CLUSTALW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="328"/>
        <source>Max iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="300"/>
        <source>NONE</source>
        <translation>Нет</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="305"/>
        <source>TREE</source>
        <translation>Дерево</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="310"/>
        <source>ALIGNMENT</source>
        <translation>Выравнивание</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="264"/>
        <source>Iteration type</source>
        <translation>Тип итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="377"/>
        <source>Out sequences order</source>
        <translation>Порядок последовательностей</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="413"/>
        <source>Input</source>
        <translation>Входные</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="418"/>
        <source>Aligned</source>
        <translation>Выровненные</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="436"/>
        <source>Protein gap parameters</source>
        <translation>Параметры для белковых последовательностей</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="446"/>
        <source>Gap separation distance</source>
        <translation>Расстояние между пробелами</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="507"/>
        <source>Residue-specific gaps off</source>
        <translation>Откл. специфические для остатков пробелы</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="493"/>
        <source>Hydrophilic gaps off</source>
        <translation>Отключить гидрофильные пробелы</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.ui" line="500"/>
        <source>No end gap separation penalty</source>
        <translation>Не вводить штраф за разделение пробелов</translation>
    </message>
</context>
<context>
    <name>CustomToolConfigParser</name>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="58"/>
        <source>Invalid config file format: file %1 cann not be opened</source>
        <translation>Invalid config file format: file %1 cann not be opened</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="67"/>
        <source>Invalid config file format: custom tool description not found</source>
        <translation>Invalid config file format: custom tool description not found</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="68"/>
        <source>Invalid config file format: there are too many entities in the file</source>
        <translation>Invalid config file format: there are too many entities in the file</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="71"/>
        <source>Can&apos;t parse the config file</source>
        <translation>Can&apos;t parse the config file</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="74"/>
        <source>Can&apos;t parse config with version %1</source>
        <translation>Can&apos;t parse config with version %1</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="116"/>
        <source>Unknown element: &apos;%1&apos;, skipping</source>
        <translation>Unknown element: &apos;%1&apos;, skipping</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="160"/>
        <source>The tool id is not specified in the config file.</source>
        <translation>The tool id is not specified in the config file.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="161"/>
        <source>The tool id contains unexpected characters, the only letters, numbers, underlines and dashes are allowed.</source>
        <translation>The tool id contains unexpected characters, the only letters, numbers, underlines and dashes are allowed.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="162"/>
        <source>The custom tool&apos;s ID shouldn&apos;t start with &quot;USUPP_&quot;, this is a distinguishing feature of the supported tools.</source>
        <translation>The custom tool&apos;s ID shouldn&apos;t start with &quot;USUPP_&quot;, this is a distinguishing feature of the supported tools.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="163"/>
        <source>The custom tool&apos;s ID shouldn&apos;t start with &quot;UCUST_&quot;, this is a distinguishing feature of the supported tools.</source>
        <translation>The custom tool&apos;s ID shouldn&apos;t start with &quot;UCUST_&quot;, this is a distinguishing feature of the supported tools.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="164"/>
        <source>The tool name is not specified in the config file.</source>
        <translation>The tool name is not specified in the config file.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="166"/>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="168"/>
        <source>The imported custom tool &quot;%1&quot; does not have an executable file. Make sure to set up a valid executable file before you use the tool.</source>
        <translation>The imported custom tool &quot;%1&quot; does not have an executable file. Make sure to set up a valid executable file before you use the tool.</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/CustomToolConfigParser.cpp" line="172"/>
        <source>The executable file &quot;%1&quot; specified for the imported custom tool &quot;%2&quot; doesn&apos;t exist. Make sure to set up a valid executable file before you use the tool.</source>
        <translation>The executable file &quot;%1&quot; specified for the imported custom tool &quot;%2&quot; doesn&apos;t exist. Make sure to set up a valid executable file before you use the tool.</translation>
    </message>
</context>
<context>
    <name>ETSSettingsWidget</name>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="26"/>
        <source>Resource settings</source>
        <translation>Настройки ресурсов</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="131"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="104"/>
        <source>Download link</source>
        <translation>Скачать</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="151"/>
        <location filename="../src/ETSSettingsWidget.ui" line="226"/>
        <source>Name</source>
        <translation>Имя</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="156"/>
        <location filename="../src/ETSSettingsWidget.ui" line="231"/>
        <source>Path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="267"/>
        <source>Import</source>
        <translation>Испорт</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="277"/>
        <source>Delete</source>
        <translation>Удалить</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="379"/>
        <source>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;.SF NS Text&apos;; font-size:13pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Select an external tool to view more information about it.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
p, li { white-space: pre-wrap; }
&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:&apos;.SF NS Text&apos;; font-size:13pt; font-weight:400; font-style:normal;&quot;&gt;
&lt;p style=&quot; margin-top:12px; margin-bottom:12px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Выберите внешний инструмент, чтобы просмотреть дополнительную информацию о нем.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="399"/>
        <source>Binary path:</source>
        <translation>Путь:</translation>
    </message>
    <message>
        <location filename="../src/ETSSettingsWidget.ui" line="409"/>
        <source>Version info</source>
        <translation>Информация о версии</translation>
    </message>
</context>
<context>
    <name>FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="20"/>
        <source>Format Database</source>
        <translation>Форматирование базы данных</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="29"/>
        <source>Input data</source>
        <translation>Входные данные</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="38"/>
        <source>Select input file(s) for formatting database:</source>
        <translation>Выберите входной файл(ы) для форматирования базы данных:</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="59"/>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="93"/>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="220"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="68"/>
        <source>Or select directory with input files:</source>
        <translation>Или выберите директорию со входными файлами:</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="105"/>
        <source>File filter</source>
        <translation>Фильтр</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="119"/>
        <source>Include files filter</source>
        <translation>Включающий фильтр</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="129"/>
        <source>*.fa;*.fasta</source>
        <translation>*.fa;*.fasta</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="136"/>
        <source>Exclude files filter</source>
        <translation>Исключающий фильтр</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="146"/>
        <source>*.pal</source>
        <translation>*.pal</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="158"/>
        <source>Type of file(s)</source>
        <translation>Тип файла(ов)</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="167"/>
        <source>protein</source>
        <translation>белки</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="177"/>
        <source>nucleotide</source>
        <translation>нуклеотиды</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="190"/>
        <source>Output settings</source>
        <translation>Выходные настройки</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="205"/>
        <source>Select the path to save the database into:</source>
        <translation>Выберите путь для сохранения базы данных:</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="234"/>
        <source>Base name for BLAST files:</source>
        <translation>Базовое имя для файлов BLAST:</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.ui" line="244"/>
        <source>Title for database file:</source>
        <translation>Название для файла базы данных:</translation>
    </message>
</context>
<context>
    <name>HmmerBuildDialog</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="20"/>
        <source>HMM3 Build</source>
        <translation>Построение профиля HMM3</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="42"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="52"/>
        <source>Input alignment file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="62"/>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="79"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="69"/>
        <source>Build to profile</source>
        <translation>Профиль</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="104"/>
        <source>Construction strategies</source>
        <translation>Настройки построения</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="126"/>
        <source>assign cols w/ &gt;= symfrac residues as consensus</source>
        <translation>assign cols w/ &gt;= symfrac residues as consensus</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="129"/>
        <source>fast</source>
        <translation>Быстрая</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="142"/>
        <source>Sym fraction</source>
        <translation>Доля</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="164"/>
        <source>Manual construction</source>
        <translation>Ручное построение</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="167"/>
        <source>Hand (requires reference annotation)</source>
        <translation>Ручная</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="185"/>
        <source>Tag sequence as a fragment, if L &lt; x*&lt;L&gt;, where x is:</source>
        <translation>Обозначить последовательность как фрагмент, если L &lt; x*&lt;L&gt;, где x есть:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="226"/>
        <source>Relative weighting</source>
        <translation>Относительные веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="234"/>
        <source>Henikoff position-based weights</source>
        <translation>Веса Henikoff основанные на позициях</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="250"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation>Gerstein/Sonnhammer/Chothia веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="274"/>
        <source>Henikoff simple filter weights</source>
        <translation>Простой фильтр весов Henikoff</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="284"/>
        <source>Identity cutoff</source>
        <translation>Отсечки идентичности</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="309"/>
        <source>Skip relative weighting; set all to 1</source>
        <translation>Не использовать относительные веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="316"/>
        <source>Use weights given in alignment file</source>
        <translation>Использовать веса из файла выравнивания</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="339"/>
        <source>Effective weighting</source>
        <translation>Эффективные веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="347"/>
        <source>Adjust effective sequence number to achieve relative entropy target</source>
        <translation>Отрегулируйте эффективный порядковый номер, чтобы достичь цель энтропии</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="381"/>
        <source>Minimum relative entropy/position</source>
        <translation>Минимальнеая относительная энтрпия</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="433"/>
        <source>Sigma parameter</source>
        <translation>Сигма-параметр</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="466"/>
        <source>Use number of single linkage clusters as effective</source>
        <translation>Использовать число единичных кластеров как эффективное</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="476"/>
        <source>Fractional identity cutoff</source>
        <translation>Долевая отсечка идентичности</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="501"/>
        <source>Use number of sequences as effective</source>
        <translation>Использовать число последовательностей как эффективное</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="516"/>
        <source>Effective sequence number for all models to</source>
        <translation>Установить эффективное число последовательностей для всех моделей</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="554"/>
        <source>E-value calibration</source>
        <translation>Калибрация E-value</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="573"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation>Длина последовательностей для MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="609"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation>Число последовательностей для MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="642"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation>Длина последовательностей для Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="672"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation>Число последовательностей для MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="702"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation>Длина последовательностей для Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="732"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation>Число последовательностей для Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="762"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation>Масса для прямого экспоненциального хвоста mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="803"/>
        <source>Other</source>
        <translation>Другое</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="819"/>
        <source>Random generator seed</source>
        <translation>Затравка генератора случайных чисел</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.ui" line="862"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 - означает, что один раз будет использована случайная затравка</translation>
    </message>
</context>
<context>
    <name>HmmerSearchDialog</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="20"/>
        <source>HMM3 Search</source>
        <translation>Поиск с помощью HMMER3</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="54"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="65"/>
        <source>Query HMM file:</source>
        <translation>HMM профайл:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="75"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="92"/>
        <source>Reporting thresholds</source>
        <translation>Допустимые результаты</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="109"/>
        <source>Report domains with E-value less than</source>
        <translation>Сохранять результаты c величиной E-value менее</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="119"/>
        <source>1E+</source>
        <translation>1E+</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="146"/>
        <source>Report domains with score greater than</source>
        <translation>Отсечение по итоговой сумме</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="180"/>
        <source>Score threshold:</source>
        <translation>Порог итоговой суммы:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="229"/>
        <source>Use profile&apos;s GA gathering cutoffs</source>
        <translation>Использовать отсечения GA</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="265"/>
        <source>Use profile&apos;s NC noise cutoffs</source>
        <translation>Использовать отсечения NC</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="295"/>
        <source>Use profile&apos;s TC trusted cutoffs</source>
        <translation>Использовать отсечения TC</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="312"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation>Число значимых последовательностей для подсчета E-value домена</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="345"/>
        <source>Acceleration heuristics</source>
        <translation>Производительность</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="360"/>
        <source>Turn off composition bias filter</source>
        <translation>Выключить composition bias фильтр</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="373"/>
        <source>Turn off biased composition score corrections</source>
        <translation>Выключить diased composition коррекцию итоговой суммы</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="386"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation>Отключить все эвристические фильтры (более точно, менее производительно)</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="401"/>
        <source>MSV filter threshold:</source>
        <translation>Порог фильтрации MSV:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="434"/>
        <source>Viterbi filter threshold:</source>
        <translation>Порог фильтрации Viterbi:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="464"/>
        <source>Forward filter threshold:</source>
        <translation>Порог фильтрации Forward:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="500"/>
        <source>Other</source>
        <translation>Другое</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="522"/>
        <source>Random generator seed</source>
        <translation>Затравка генератора случайных чисел</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.ui" line="565"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 - означает, что один раз будет использована случайная затравка</translation>
    </message>
</context>
<context>
    <name>IlluminaClipAdditionalSettingsDialog</name>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipAdditionalSettingsDialog.ui" line="32"/>
        <source>Configure ILLUMINACLIP Optional Settings</source>
        <translation>Настройки ILLUMINACLIP</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipAdditionalSettingsDialog.ui" line="44"/>
        <source>Provide optional settings</source>
        <translation>Задать настройки</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipAdditionalSettingsDialog.ui" line="68"/>
        <source>Min adapter length</source>
        <translation>Мин длина адаптера</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipAdditionalSettingsDialog.ui" line="81"/>
        <source>Keep both reads</source>
        <translation>Оба рида</translation>
    </message>
</context>
<context>
    <name>IlluminaClipSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="46"/>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="61"/>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="68"/>
        <source>A FASTA file with the adapter sequences. Files for TruSeq2 (GAII machines), TruSeq3 (HiSeq and MiSeq machines) and Nextera kits for SE and PE reads are now available by default. The naming of the various sequences within the specified file determines how they are used.</source>
        <translation>FASTA  файл с адаптерами. Файлы для TruSeq2 (GAII машин), TruSeq3 (HiSeq и MiSeq машин) и Nextera наборов для SE и PE ридов доступны по умолчанию.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="49"/>
        <source>Adapter sequences</source>
        <translation>Адаптеры</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="71"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="87"/>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="97"/>
        <source>The maximum mismatch count in short sections which will still allow a full match to be performed.</source>
        <translation>Максимальное количество несоответствий в коротких участках, которое все равно позволит выполнить полное совпадение.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="90"/>
        <source>Seed mismatches</source>
        <translation>Несоответствия затравки</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="110"/>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="120"/>
        <source>A threshold for simple alignment mode. Values between 7 and 15 are recommended. A perfect match of a 12 base sequence will score just over 7, while 25 bases are needed to score 15.</source>
        <translation>Порог для простого режима выравнивания. Рекомендуются значения от 7 до 15. Идеальное совпадение из 12 базовых последовательностей принесет чуть больше 7 баллов, в то время как для получения 15 нужно 25 оснований.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="113"/>
        <source>Palindrome clip threshold</source>
        <translation>Порог для клипа палиндрома</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="139"/>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="149"/>
        <source>A threshold for palindrome alignment mode. For palindromic matches, a longer alignment is possible. Therefore the threshold can be in the range of 30. Even though this threshold is very high (requiring a match of almost 50 bases) Trimmomatic is still able to identify very, very short adapter fragments.</source>
        <translation>Порог для режима палинромиального выравнивания. Для палиндромиальных совпадений возможны более длинные выравнивания. Тем не менее порог может быть в пределах 30.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="142"/>
        <source>Simple clip threshold</source>
        <translation>Порог для простого клипа</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipSettingsWidget.ui" line="191"/>
        <source>Optional...</source>
        <translation>Опционально...</translation>
    </message>
</context>
<context>
    <name>ImportExternalToolDialog</name>
    <message>
        <location filename="../src/custom_tools/ImportExternalToolDialog.ui" line="14"/>
        <source>Import External Tool</source>
        <translation>Импорт внешнего инструмента</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/ImportExternalToolDialog.ui" line="25"/>
        <source>External tool config</source>
        <translation>Конфиг для инструмента</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/ImportExternalToolDialog.ui" line="40"/>
        <source>...</source>
        <translation>...</translation>
    </message>
</context>
<context>
    <name>LengthSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/util/LengthSettingsWidget.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/util/LengthSettingsWidget.ui" line="32"/>
        <source>Length</source>
        <translation>Длина</translation>
    </message>
</context>
<context>
    <name>MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="20"/>
        <source>Align with MAFFT</source>
        <translation>Выравнивание с помощью MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="28"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="46"/>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="85"/>
        <source>Gap opening penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="131"/>
        <source>Offset (works like gap extension penalty)</source>
        <translation>Смещение (работает как штраф за промежуточный пробел)</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.ui" line="174"/>
        <source>Maximum number of iterative refinement</source>
        <translation>Максимальное число итераций</translation>
    </message>
</context>
<context>
    <name>MaxInfoSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="32"/>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="42"/>
        <source>The read length which is likely to allow the location of the read within the target sequence. Extremely short reads, which can be placed into many different locations, provide little value. Typically, the length would be in the order of 40 bases, however, the value also depends on the size and complexity of the target sequence.</source>
        <translation>Длина рида, которая, вероятно, позволит определить местоположение рида в целевой последовательности. Чрезвычайно короткие риды, которые могут быть размещены во многих разных местах, обеспечивают небольшую ценность. Как правило, длина будет порядка 40 оснований, однако значение также зависит от размера и сложности целевой последовательности.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="35"/>
        <source>Target length</source>
        <translation>Целевая длина</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="58"/>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="68"/>
        <source>The balance between preserving as much read length as possible vs. removal of incorrect bases. A low value of this parameter (&lt;0.2) favours longer reads, while a high value (&gt;0.8) favours read correctness.</source>
        <translation>Баланс между сохранением как можно большей длины рида и удалением неверных оснований. Низкое значение этого параметра (&lt;0,2) способствует более длительному риду, а высокое значение (&gt; 0,8) способствует правильности рида.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoSettingsWidget.ui" line="61"/>
        <source>Strictness</source>
        <translation>Взыскательность</translation>
    </message>
</context>
<context>
    <name>MrBayesDialog</name>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="60"/>
        <source>Model</source>
        <translation>Настройки</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="78"/>
        <source>ModelType</source>
        <translation>ModelType</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="95"/>
        <source>Rate</source>
        <translation>Показатель</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="118"/>
        <source>Gamma</source>
        <translation>Гамма</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="151"/>
        <source>MCMC</source>
        <translation>MCMC</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="163"/>
        <source>Chain length</source>
        <translation>Длина цепи</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="189"/>
        <source>Subsampling frequence</source>
        <translation>Размер выборки</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="221"/>
        <source>Burn-in length</source>
        <translation>Длина прожига</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="250"/>
        <source>Heated chains</source>
        <translation>Подогретые цепи</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="279"/>
        <source>Heated chain temp</source>
        <translation>Температура подогрева цепей</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="321"/>
        <source>Random seed</source>
        <translation>Затравка</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialog.ui" line="329"/>
        <source>Display Options</source>
        <translation>Опции отображения</translation>
    </message>
</context>
<context>
    <name>PhmmerSearchDialog</name>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="20"/>
        <source>Phmmer Search</source>
        <translation>Поиск phmmer</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="36"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="50"/>
        <source>Query sequence file:</source>
        <translation>Файл последовательности:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="60"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="87"/>
        <source>Reporting thresholds</source>
        <translation>Допустимые результаты</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="103"/>
        <source>Report domains with E-value less than</source>
        <translation>Сохранять результаты c величиной E-value менее</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="113"/>
        <source>1E+</source>
        <translation>1E+</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="142"/>
        <source>Report domains with score greater than</source>
        <translation>Отсечение по итоговой сумме</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="166"/>
        <source>Number of significant sequences for domain E-value calculation</source>
        <translation>Число значимых последовательностей для подсчета E-value домена</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="201"/>
        <source>Scoring system</source>
        <translation>Оценка пропусков</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="217"/>
        <source>Gap open probability</source>
        <translation>Вероятность открытия пробела</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="247"/>
        <source>Gap extend probability</source>
        <translation>Вероятность продления пробела</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="285"/>
        <source>Acceleration</source>
        <translation>Производительность</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="297"/>
        <source>Turn off composition bias filter</source>
        <translation>Выключить composition bias фильтр</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="310"/>
        <source>Turn off biased composition score corrections</source>
        <translation>Выключить diased composition коррекцию итоговой суммы</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="323"/>
        <source>Turn all heuristic filters off (less speed, more power)</source>
        <translation>Отключить все эвристические фильтры (более точно, менее производительно)</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="338"/>
        <source>Viterbi filter threshold</source>
        <translation>Порог фильтрации Viterbi</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="368"/>
        <source>MSV filter threshold</source>
        <translation>Порог фильтрации MSV</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="401"/>
        <source>Forward filter threshold</source>
        <translation>Порог фильтрации Forward</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="437"/>
        <source>E-value calibration</source>
        <translation>Калибрация E-value</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="456"/>
        <source>Length of sequences for MSV Gumbel mu fit</source>
        <translation>Длина последовательностей для MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="492"/>
        <source>Number of sequences for MSV Gumbel mu fit</source>
        <translation>Число последовательностей для MSV Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="525"/>
        <source>Length of sequences for Viterbi Gumbel mu fit</source>
        <translation>Длина последовательностей для Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="555"/>
        <source>Number of sequences for Viterbi Gumbel mu fit</source>
        <translation>Число последовательностей для Viterbi Gumbel mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="585"/>
        <source>Length of sequences for Forward exp tail mu fit</source>
        <translation>Длина последовательностей для Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="615"/>
        <source>Number of sequences for Forward exp tail mu fit</source>
        <translation>Число последовательностей для Forward exp tail mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="645"/>
        <source>Tail mass for Forward exponential tail mu fit</source>
        <translation>Масса для прямого экспоненциального хвоста mu fit</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="686"/>
        <source>Other</source>
        <translation>Другое</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="702"/>
        <source>Random generator seed</source>
        <translation>Затравка генератора случайных чисел</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.ui" line="745"/>
        <source>* 0 means that one-time arbitrary seed will be used</source>
        <translation>* 0 - означает, что один раз будет использована случайная затравка</translation>
    </message>
</context>
<context>
    <name>PhyMLDialog</name>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="76"/>
        <source>Substitution model</source>
        <translation>Модель замен</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="69"/>
        <source>Equilibrium frequencies</source>
        <translation>Равновесные частоты</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="116"/>
        <source>optimized</source>
        <translation>оптимизированные</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="83"/>
        <source>empirical</source>
        <translation>эмпирические</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="106"/>
        <source>Transition / transversion ratio </source>
        <translation>Коэффициент перехода/трансверсии</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="231"/>
        <source>estimated </source>
        <translation>оценка</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="209"/>
        <source>Proportion of invariable sites</source>
        <translation>Доля неизменяемых сайтов</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="63"/>
        <source>Substitution Model</source>
        <translation>Модель замен</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="126"/>
        <source>Number of substitution rate categories</source>
        <translation>Число категорий коэффициента замещения</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="133"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="149"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="165"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="275"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="294"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="345"/>
        <source>Set a fixed value or get the maximum likelihood estimate.</source>
        <translation>Установить фиксированное значение или используйте максимальную оценку.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="136"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="152"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="348"/>
        <source>estimated</source>
        <translation>оценка</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="171"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="281"/>
        <location filename="../src/phyml/PhyMLDialog.ui" line="300"/>
        <source>fixed</source>
        <translation>фиксированное</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="313"/>
        <source>Gamma shape parameter</source>
        <translation>Параметр формы Гамма</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="362"/>
        <source>Branch Support</source>
        <translation>Поддержка ветви</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="374"/>
        <source>Use fast likelihood-based method</source>
        <translation>Использовать быстрый метод максимального правдоподобия</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="394"/>
        <source>Perform bootstrap</source>
        <translation>Выполнить начальную загрузку</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="404"/>
        <source>It is the number of bootstrap replicates.</source>
        <translation>Это число начальных загрузок.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="418"/>
        <source>Tree Searching</source>
        <translation>Поиск дерева</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="449"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="456"/>
        <source>Type of tree improvement</source>
        <translation>Тип улучшения дерева</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="479"/>
        <source>Set number of random starting tree</source>
        <translation>Установить число случайных начал дерева</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="506"/>
        <source>Optimise tree topology</source>
        <translation>Оптимизировать топологию дерева</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="516"/>
        <source>Optimise branch lengths</source>
        <translation>Оптимизировать длину ветвей</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="526"/>
        <source>Optimise substitution rate</source>
        <translation>Оптимизировать уровень замещения</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialog.ui" line="534"/>
        <source>Display Options</source>
        <translation>Опции отображения</translation>
    </message>
</context>
<context>
    <name>QCoreApplication</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="472"/>
        <source>&lt;html&gt;Select the type of input for SPAdes. URL(s) to the input files of the selected type(s) should be provided to the corresponding port(s) of the workflow element.&lt;p&gt;At least one library of the following types is required:&lt;ul&gt;&lt;li&gt;Illumina paired-end/high-quality mate-pairs/unpaired reads&lt;/li&gt;&lt;li&gt;IonTorrent paired-end/high-quality mate-pairs/unpaired reads&lt;/li&gt;&lt;li&gt;PacBio CCS reads (at least 5 reads coverage is recommended)&lt;/li&gt;&lt;/ul&gt;&lt;/p&gt;&lt;p&gt;It is strongly suggested to provide multiple paired-end and mate-pair libraries according to their insert size (from smallest to longest).&lt;/p&gt;&lt;p&gt;Additionally, one may input Oxford Nanopore reads, Sanger reads, contigs generated by other assembler(s), etc.Note that Illumina and IonTorrent libraries should not be assembled together. All other types of input data are compatible.&lt;/p&gt;&lt;p&gt;It is also possible to set up reads orientation (forward-reverse (fr), reverse-forward (rf), forward-forward (ff)) and specify whether paired reads are separate or interlaced.&lt;/p&gt;&lt;p&gt;Illumina, IonTorrent or PacBio CCS reads should be provided in FASTQ format.&lt;br&gt;Illumina or PacBio read may also be provided in FASTA format. Error correction should be skipped in this case (see the &quot;Running mode&quot; parameter).&lt;br&gt;Sanger, Oxford Nanopore and PacBio CLR reads can be provided in both formats since SPAdes does not run error correction for these types of data.&lt;/p&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt; Выберите тип ввода для SPAdes. URL-адреса входных файлов выбранного типа (-ов) должны быть указаны для соответствующего порта (-ов) элемента. &lt;P&gt; Требуется как минимум одна библиотека следующих типов: &lt;ul&gt; &lt;li&gt; Парные риды Illumina / высококачественные пары / непарные риды &lt;/ li&gt; &lt;li&gt; Парные риды IonTorrent / высококачественные пары / непарные риды &lt;/ li&gt; &lt;li&gt; риды PacBio CCS (рекомендуется покрутие не менее 5 ридов) &lt;/ li&gt; &lt;/ ul&gt; &lt;/ p&gt; &lt;p&gt; Настоятельно рекомендуется предоставить несколько библиотек парных и сопряженных ридов в соответствии с их размером вставки (от самого маленького до самого длинного). &lt;/ p&gt; &lt;p&gt; Кроме того, можно вводить риды Oxford Nanopore, риды Sanger, контиги, генерируемые другими ассемблерами, и т. д. Обратите внимание, что библиотеки Illumina и IonTorrent не следует собирать вместе. Все другие типы входных данных совместимы. &lt;/ P&gt; &lt;p&gt; Также можно настроить ориентацию рида (вперед-назад (fr), назад-вперед (rf), вперед-вперед (ff)) и указать, являются ли парные риды являются отдельными или чересстрочными. &lt;/ p&gt; &lt;p&gt; Риды Illumina, IonTorrent или PacBio CCS должны предоставляться в формате FASTQ. &lt;br&gt; Риды Illumina или PacBio также могут предоставляться в формате FASTA. В этом случае исправление ошибок должно быть пропущено (см. Параметр «Режим работы»). Риды Sanger, Oxford Nanopore и PacBio CLR могут предоставляться в обоих форматах, поскольку SPAdes не выполняет исправление ошибок для этих типов данных. &lt;/ р&gt; &lt;/ HTML&gt;</translation>
    </message>
</context>
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="137"/>
        <source>Path for temporary files</source>
        <translation>Путь к папке с временными файлами</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="138"/>
        <source>Path for temporary files not selected.</source>
        <translation>Путь для временных файлов не задан.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="139"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="173"/>
        <source>Genome sequence slot is not binded : &apos;%1&apos;</source>
        <translation>Слот последовательности не связан: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="176"/>
        <source>Transcripts slot is not binded : &apos;%1&apos;</source>
        <translation>Слот транскрипций не связан: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="189"/>
        <source>Input transcripts</source>
        <translation>Входные транскрипции</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="190"/>
        <source>Genomic sequence url</source>
        <translation>Путь до последовательности</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="190"/>
        <source>Genomic sequence url [FASTA]</source>
        <translation>Путь до последовательности [FASTA]</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="191"/>
        <source>Transcripts url</source>
        <translation>Путь до транскрипций</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="191"/>
        <source>Transcripts url [GTF]</source>
        <translation>Путь до транскрипций [GTF]</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="197"/>
        <source>Extracted sequences url</source>
        <translation>Путь до извлекаемых последовательностей</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="205"/>
        <source>Output sequences</source>
        <translation>Выходные последовательности</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="205"/>
        <source>The url to the output file with the extracted sequences.</source>
        <translation>Путь до выходного файла с извлеченными последовательностями.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="215"/>
        <source>Extract Transcript Sequences with gffread</source>
        <translation>Извлечение последовательностей транскриптов с помощью gffread</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="216"/>
        <source>Extract transcript sequences from the genomic sequence(s) with gffread.</source>
        <translation>извлекает последовательности транскриптов из геномных последовательностей при помощи gffread.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="91"/>
        <source>Installed</source>
        <translation>Установлено</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="92"/>
        <source>Not installed</source>
        <translation>Не установлено</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="93"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Download &lt;a href=&quot;http://ugene.net/download-all_html#en_data_analysis_tools&quot;&gt;&lt;span style=&quot; text-decoration: underline; color:#1866af;&quot;&gt;tools executables&lt;/span&gt;&lt;/a&gt; and configure the tools paths. &lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Скачать &lt;a href=&quot;http://ugene.net/ru/download-all_html#ru_data_analysis_tools&quot;&gt;&lt;span style=&quot; text-decoration: underline; color:#1866af;&quot;&gt;инструменты&lt;/span&gt;&lt;/a&gt; и сконфигурировать пути до файлов. &lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</translation>
    </message>
    <message>
        <location filename="../src/RnaSeqCommon.cpp" line="54"/>
        <source>Unexpected value &apos;%1&apos; of the library type</source>
        <translation>Unexpected value &apos;%1&apos; of the library type</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="816"/>
        <source>TopHat and Bowtie tool versions are undefined, this may cause some compatibility issues</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="821"/>
        <source>%1 tool&apos;s version is undefined, this may cause some compatibility issues</source>
        <translation>%1 версия инструмента не определена, это может вызвать некоторые проблемы совместимости</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="828"/>
        <source>Bowtie and TopHat tools have incompatible versions. Your TopHat&apos;s version is %1, Bowtie&apos;s one is %2. The following are considered to be compatible: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</source>
        <translation>Bowtie и TopHat инструменты имеют несовместимые версии. Версия TopHat %1, версия Bowtie %2. Для совместимости необходимо: Bowtie &lt; &quot;0.12.9&quot; and TopHat &lt;= &quot;2.0.8&quot; or Bowtie &gt;= &quot;0.12.9&quot; and TopHat &gt;= &quot;2.0.8.b&quot;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="865"/>
        <source>At least two samples are required</source>
        <translation>Необходимо как минимум два примера</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="872"/>
        <source>Duplicate sample name: </source>
        <translation>Повторяющееся имя примера: </translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="877"/>
        <source>No datasets in the sample: </source>
        <translation>В примере нет данных: </translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="58"/>
        <source>Sequence name from file</source>
        <translation>Имя последовательности в файле</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="59"/>
        <source>File name</source>
        <translation>Имя файла</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticDelegate.cpp" line="163"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;p&gt;Click the &quot;Add new step&quot; button and select a step. The following options are available:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;ILLUMINACLIP: Cut adapter and other illumina-specific sequences from the read.&lt;/li&gt;&lt;li&gt;SLIDINGWINDOW: Perform a sliding window trimming, cutting once the average quality within the window falls below a threshold.&lt;/li&gt;&lt;li&gt;LEADING: Cut bases off the start of a read, if below a threshold quality.&lt;/li&gt;&lt;li&gt;TRAILING: Cut bases off the end of a read, if below a threshold quality.&lt;/li&gt;&lt;li&gt;CROP: Cut the read to a specified length.&lt;/li&gt;&lt;li&gt;HEADCROP: Cut the specified number of bases from the start of the read.&lt;/li&gt;&lt;li&gt;MINLEN: Drop the read if it is below a specified length.&lt;/li&gt;&lt;li&gt;AVGQUAL: Drop the read if the average quality is below the specified level.&lt;/li&gt;&lt;li&gt;TOPHRED33: Convert quality scores to Phred-33.&lt;/li&gt;&lt;li&gt;TOPHRED64: Convert quality scores to Phred-64.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation>&lt;html&gt; &lt;head&gt; &lt;/ head&gt; &lt;body&gt; &lt;p&gt; Нажмите кнопку «Добавить новый шаг» и выберите шаг. Доступны следующие параметры: &lt;/ p&gt; &lt;ul&gt; &lt;li&gt; ILLUMINACLIP: вырезать адаптер и другие специфичные последовательности для рида. &lt;/ Li&gt; &lt;li&gt; SLIDINGWINDOW: выполнить обрезку скользящего окна, обрезав один раз среднее качество в пределах окна ниже порога. &lt;/ li&gt; &lt;li&gt; LEADING: обрезать базы от начала рида, если ниже порогового качества. &lt;/ li&gt; &lt;li&gt; TRAILING: обрезать базы от конца рида, если качество ниже порогового. &lt;/ li&gt; &lt;li&gt; CROP: обрезать рид до указанной длины. &lt;/ li&gt; &lt;li&gt; HEADCROP: вырезать указанное количество баз с начала рида. &lt;/ li&gt; &lt; li&gt; MINLEN: отбросить рид, если он меньше указанной длины. &lt;/ li&gt; &lt;li&gt; AVGQUAL: отбросить рид, если среднее качество ниже указанного уровня. &lt;/ li&gt; &lt;li&gt; TOPHRED33: преобразовать показатели качества в Phred-33. &lt;/ Li&gt; &lt;li&gt; TOPHRED64: преобразование показателей качества в Phred-64. &lt;/ Li&gt; &lt;/ ul&gt; &lt;/ body&gt; &lt;/ html&gt;</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticDelegate.cpp" line="178"/>
        <source>Add a step.</source>
        <translation>Добавить шаг.</translation>
    </message>
</context>
<context>
    <name>QualitySettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/util/QualitySettingsWidget.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/util/QualitySettingsWidget.ui" line="32"/>
        <source>Quality threshold</source>
        <translation>Порог качества</translation>
    </message>
</context>
<context>
    <name>SlidingWindowSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="14"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="32"/>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="42"/>
        <source>The number of bases to average across.</source>
        <translation>Число основания для усреднения.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="35"/>
        <source>Window size</source>
        <translation>Размер окна</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="49"/>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="59"/>
        <source>The average quality required.</source>
        <translation>Трубуемое среднее качество.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowSettingsWidget.ui" line="52"/>
        <source>Quality threshold</source>
        <translation>Порог качества</translation>
    </message>
</context>
<context>
    <name>SnpEffDatabaseDialog</name>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDialog.ui" line="14"/>
        <source>Select SnpEff Database</source>
        <translation>Выбор базы данных SnpEff</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDialog.ui" line="23"/>
        <source>Search...</source>
        <translation>Поиск...</translation>
    </message>
</context>
<context>
    <name>SpadesDelegate</name>
    <message>
        <location filename="../src/spades/SpadesDelegate.cpp" line="44"/>
        <source>Configure input type</source>
        <translation>Настрояка входного типа</translation>
    </message>
</context>
<context>
    <name>SpadesInputData</name>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="14"/>
        <source>Configure SPAdes Input Type</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="32"/>
        <source>Required input (at least one)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="53"/>
        <location filename="../src/spades/SpadesInputData.ui" line="337"/>
        <source>Illumina/Ion Torrent reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="97"/>
        <source>Sequencing platform</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="105"/>
        <source>Illumina</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="110"/>
        <source>Ion Torrent</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="141"/>
        <source>Paired-end reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="157"/>
        <location filename="../src/spades/SpadesInputData.ui" line="214"/>
        <location filename="../src/spades/SpadesInputData.ui" line="379"/>
        <source>fr</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="162"/>
        <location filename="../src/spades/SpadesInputData.ui" line="219"/>
        <location filename="../src/spades/SpadesInputData.ui" line="384"/>
        <source>rf</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="167"/>
        <location filename="../src/spades/SpadesInputData.ui" line="224"/>
        <location filename="../src/spades/SpadesInputData.ui" line="389"/>
        <source>ff</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="179"/>
        <location filename="../src/spades/SpadesInputData.ui" line="236"/>
        <location filename="../src/spades/SpadesInputData.ui" line="401"/>
        <source>Separate reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="184"/>
        <location filename="../src/spades/SpadesInputData.ui" line="241"/>
        <location filename="../src/spades/SpadesInputData.ui" line="406"/>
        <source>Interlaced reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="198"/>
        <source>High-quality mate-pairs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="263"/>
        <source>Unpaired reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="288"/>
        <source>PacBio CCS reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="316"/>
        <source>Additional input</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="363"/>
        <source>Mate-pairs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="438"/>
        <source>PacBio CLR reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="445"/>
        <source>Oxford Nanopore reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="452"/>
        <source>Sanger reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="459"/>
        <source>Trusted contigs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesInputData.ui" line="466"/>
        <source>Untrusted contigs</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SpadesPropertyDialog</name>
    <message>
        <location filename="../src/spades/SpadesDelegate.cpp" line="153"/>
        <source>At least one of the required input ports should be set in the &quot;Input data&quot; parameter.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesDelegate.cpp" line="169"/>
        <location filename="../src/spades/SpadesDelegate.cpp" line="177"/>
        <location filename="../src/spades/SpadesDelegate.cpp" line="199"/>
        <location filename="../src/spades/SpadesDelegate.cpp" line="229"/>
        <location filename="../src/spades/SpadesDelegate.cpp" line="305"/>
        <location filename="../src/spades/SpadesDelegate.cpp" line="309"/>
        <source>Incorrect parameters, can&apos;t parse</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>SpadesSettings</name>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="20"/>
        <source>Form</source>
        <translation>Форма</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="30"/>
        <source>Base Options</source>
        <translation>Основные параметры</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="45"/>
        <source>Running mode</source>
        <translation>Режим запуска</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="59"/>
        <source>Error Correction and Assembly</source>
        <translation>Исправление ошибок и сборка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="64"/>
        <source>Assembly only</source>
        <translation>Только сборка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="69"/>
        <source>Error correction only</source>
        <translation>Только исправление ошибок</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="80"/>
        <source>k-mer sizes (-k)</source>
        <translation>размеры k-mer (-k)</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="99"/>
        <source>Dataset type</source>
        <translation>Тип набора данных</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="120"/>
        <source>Multi Cell</source>
        <translation>Сложная ячейка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="125"/>
        <source>Single Cell</source>
        <translation>Одиночная ячейка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="142"/>
        <source>Memory limit GB (-m)</source>
        <translation>Предел памяти GB (-m)</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="158"/>
        <location filename="../src/spades/SpadesSettings.ui" line="190"/>
        <source>Number of threads</source>
        <translation>Число потоков</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesSettings.ui" line="174"/>
        <source>Number of threads (-t)</source>
        <translation>число потоков (-t)</translation>
    </message>
</context>
<context>
    <name>SpadesWorkerFactory</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="129"/>
        <source>unpaired reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="131"/>
        <source>PacBio CCS reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="133"/>
        <source>PacBio CLR reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="135"/>
        <source>Oxford Nanopore reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="137"/>
        <source>Sanger reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="139"/>
        <source>trusted contigs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="141"/>
        <source>untrusted contigs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="143"/>
        <source>paired-end reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="145"/>
        <source>mate-pairs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="147"/>
        <source>high-quality mate-pairs</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>StringtieGeneAbundanceReportWorkerFactory</name>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="60"/>
        <source>Input URL url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="61"/>
        <source>Input URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="67"/>
        <source>Input StringTie gene abundance file(s) url</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="68"/>
        <source>URL(s) to sorted gene abundance file(s), produced by StringTie.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="79"/>
        <source>Output file</source>
        <translation type="unfinished">Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="80"/>
        <source>Specify the name of the output tab-delimited text file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="89"/>
        <source>Auto</source>
        <translation type="unfinished">Авто</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="99"/>
        <source>StringTie Gene Abundance Report</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorkerFactory.cpp" line="100"/>
        <source>The element summarizes gene abundance output of StringTie and saves the result into a common tab-delimited text file. The first two columns of the file are &quot;Gene ID&quot; and &quot;Gene name&quot;. Each other column contains &quot;FPKM&quot; values for the genes from an input gene abundance file.&lt;br&gt;&lt;br&gt;Provide URL(s) to the StringTie gene abundance file(s) to the input port of the element.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="20"/>
        <source>Align with T-Coffee</source>
        <translation>Выравнивание с помощью T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="28"/>
        <source>Input and output</source>
        <translation>Входные и выходные данные</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="36"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="46"/>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="63"/>
        <source>...</source>
        <translation>...</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="53"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="75"/>
        <source>Advanced options</source>
        <translation>Дополнительные опции</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="83"/>
        <source>Gap opening penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="122"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пробел</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.ui" line="158"/>
        <source>Number of iterations</source>
        <translation>Количество итераций</translation>
    </message>
</context>
<context>
    <name>TrimmomaticPrompter</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPrompter.cpp" line="32"/>
        <source>Trim, crop and/or remove adapters for input Illumina FASTQ data.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="89"/>
        <source>Input FASTQ URL 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="90"/>
        <source>URL to a FASTQ file with SE reads or &quot;left&quot; PE reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="93"/>
        <source>Input FASTQ URL 2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="94"/>
        <source>URL to a FASTQ file with &quot;right&quot; PE reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="97"/>
        <source>Input FASTQ file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="98"/>
        <source>URL(s) to FASTQ file(s) with Illumina reads should be provided.&lt;br/&gt;&lt;br/&gt;In case of SE reads use the &quot;Input FASTQ URL 1&quot; slot only.&lt;br/&gt;&lt;br/&gt;In case of PE reads input &quot;left&quot; reads to &quot;Input FASTQ URL 1&quot;, &quot;right&quot; reads to &quot;Input FASTQ URL 2&quot;.&lt;br/&gt;&lt;br/&gt;See also the &quot;Input data&quot; parameter of the element.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="112"/>
        <source>Output FASTQ URL 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="113"/>
        <source>URL to a FASTQ file with SE reads or &quot;left&quot; PE reads, processed by Trimmomatic.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="116"/>
        <source>Output FASTQ URL 2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="117"/>
        <source>URL to a FASTQ file with &quot;right&quot; reads, processed by Trimmomatic.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="120"/>
        <source>Improved FASTQ file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="121"/>
        <source>The port outputs URLs to FASTQ files, produced by Trimmomatic.&lt;br/&gt;&lt;br/&gt;In case of SE reads for each input FASTQ file one output file is produced. The file URL is passed to the output slot &quot;Output FASTQ URL 1&quot;.&lt;br/&gt;&lt;br/&gt;In case of PE reads for each pair of input FASTQ files four output files are produced: for paired &quot;left&quot; reads, for unpaired &quot;left&quot; reads, for paired &quot;right&quot; reads, and for unpaired &quot;right&quot; reads. URLs of files with paired reads are passed to the output slots &quot;Output FASTQ URL 1&quot; and &quot;Output FASTQ URL 2&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="139"/>
        <source>Input data</source>
        <translation type="unfinished">Входные данные</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="140"/>
        <source>Set the type of the input reads: single-end (SE) or paired-end (PE).&lt;br/&gt;&lt;br/&gt;One or two slots of the input port are used depending on the value of the parameter. Pass URL(s) to data to these slots.&lt;br/&gt;&lt;br/&gt;Note that the paired-end mode will use additional information contained in paired reads to better find adapter or PCR primer fragments introduced by the library preparation process.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="148"/>
        <source>Trimming steps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="149"/>
        <source>Configure trimming steps that should be performed by Trimmomatic.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="152"/>
        <source>Output file</source>
        <translation type="unfinished">Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="153"/>
        <source>Specify the output file name.</source>
        <translation type="unfinished">Укажите имя выходного файла.</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="156"/>
        <source>Paired output file 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="157"/>
        <source>Specify the output file name for &quot;left&quot; reads that have paired &quot;right&quot; reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="160"/>
        <source>Paired output file 2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="161"/>
        <source>Specify the output file name for unpaired &quot;left&quot; reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="164"/>
        <source>Unpaired output file 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="165"/>
        <source>Specify the output file name for &quot;left&quot; reads that have no pair.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="168"/>
        <source>Unpaired output file 2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="169"/>
        <source>Specify the output file name for &quot;right&quot; reads that have no pair.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="172"/>
        <source>Generate detailed log</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="173"/>
        <source>Select &quot;True&quot; to generate a file with log of all read trimmings, indicating the following details (-trimlog): &lt;ul&gt;   &lt;li&gt;the read name&lt;/li&gt;   &lt;li&gt;the surviving sequence length&lt;/li&gt;   &lt;li&gt;the location of the first surviving base, aka. the amount trimmed from the start&lt;/li&gt;   &lt;li&gt;the location of the last surviving base in the original read&lt;/li&gt;   &lt;li&gt;the amount trimmed from the end&lt;/li&gt; &lt;/ul&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="184"/>
        <source>Log file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="185"/>
        <source>Specify a text file to keep detailed information about reads trimming.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="188"/>
        <source>Number of threads</source>
        <translation type="unfinished">Число потоков</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="189"/>
        <source>Use multiple threads (-threads).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="261"/>
        <source>Improve Reads with Trimmomatic</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorkerFactory.cpp" line="262"/>
        <source>Trimmomatic is a fast, multithreaded command line tool that can be used to trim and crop Illumina (FASTQ) data as well as to remove adapters.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>TrimmomaticPropertyDialog</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="20"/>
        <source>Configure Trimmomatic Steps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="38"/>
        <source>Steps</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="98"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Add new step&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="158"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Move selected step above&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="218"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Move selected step below&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="275"/>
        <source>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Remove selected step&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="309"/>
        <source>Step settings</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticPropertyDialog.ui" line="393"/>
        <source>Description</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AlignToReferenceBlastCmdlineTask</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="95"/>
        <source>Map Sanger reads to reference</source>
        <translation>Выравнивание данных секвенирования по Сэнгеру</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="111"/>
        <source>The &apos;%1&apos; reference file doesn&apos;t exist.</source>
        <translation>The &apos;%1&apos; reference file doesn&apos;t exist.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="116"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="119"/>
        <source>wrong reference format</source>
        <translation>wrong reference format</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="172"/>
        <source>&lt;u&gt;Filtered by quality (%1):&lt;/u&gt;</source>
        <translation>&lt;u&gt;Прочтения, отфильтрованные по качеству (%1):&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="199"/>
        <source>Failed to create output folder: %1.</source>
        <translation>Failed to create output folder: %1.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="227"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="230"/>
        <source>wrong output format</source>
        <translation>wrong output format</translation>
    </message>
</context>
<context>
    <name>U2::AlignToReferenceBlastDialog</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="259"/>
        <source>Map</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="260"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="266"/>
        <source>File name</source>
        <translation>Имя файла</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="267"/>
        <source>Sequence name from file</source>
        <translation>Имя последовательности в файле</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="285"/>
        <source>Select Output File...</source>
        <translation>Укажите выходной файл...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="298"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="304"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="322"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="344"/>
        <source>Error</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="298"/>
        <source>Reference sequence is not set.</source>
        <translation>Референсная последовательность не задана.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="304"/>
        <source>No reads provided.</source>
        <translation>Риды не заданы.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="322"/>
        <source>Output file is not set.</source>
        <translation>Выходной файл не указан.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="333"/>
        <source>Overwrite the file?</source>
        <translation>Переписать этот файл?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="334"/>
        <source>The result file already exists. Would you like to overwrite it?</source>
        <translation>Такой файл уже существует. Переписать его?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="344"/>
        <source>Unable to delete the file.</source>
        <translation>Невозможно удалить этот файл.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="356"/>
        <source>Open Reference Sequence</source>
        <translation>Выбрать референсную последовательность</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastDialog.cpp" line="367"/>
        <source>Select File(s) with Read(s)</source>
        <translation>Выбрать прочтения</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsIntersectAnnotationsByEntityTask</name>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="116"/>
        <source>Intersect annotations task</source>
        <translation>Intersect annotations task</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="173"/>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="193"/>
        <source>Failed to get IOAdapterFactory</source>
        <translation>Failed to get IOAdapterFactory</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="180"/>
        <source>Result document is NULL</source>
        <translation>Result document is NULL</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="190"/>
        <source>Failed to get BED format</source>
        <translation>Failed to get BED format</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsIntersectTask</name>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="65"/>
        <source>BedtoolsIntersect task</source>
        <translation>BedtoolsIntersect task</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="70"/>
        <source>No input A URL</source>
        <translation>No input A URL</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="71"/>
        <source>No input B URL(s)</source>
        <translation>No input B URL(s)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedtoolsIntersectTask.cpp" line="72"/>
        <source>Output URL is not set</source>
        <translation>Output URL is not set</translation>
    </message>
</context>
<context>
    <name>U2::BedtoolsSupport</name>
    <message>
        <location filename="../src/bedtools/BedtoolsSupport.cpp" line="56"/>
        <source>&lt;i&gt;Bedtools&lt;/i&gt;: flexible tools for genome arithmetic and DNA sequence analysis.</source>
        <translation>&lt;i&gt;Bedtools&lt;/i&gt;: гибкие инструменты для геномной арифметики и ДНК анализа.</translation>
    </message>
</context>
<context>
    <name>U2::BigWigSupport</name>
    <message>
        <location filename="../src/bigWigTools/BigWigSupport.cpp" line="54"/>
        <source>&lt;i&gt;bedGraphToBigWig&lt;/i&gt;: converts bedGrapth to bigWig.</source>
        <translation>&lt;i&gt;bedGraphToBigWig&lt;/i&gt;: конвертирует bedGrapth в bigWig.</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupport</name>
    <message>
        <source>The &lt;i&gt;blastall&lt;/i&gt; is the old program developed and distributed by the NCBI for running BLAST searches. The NCBI recommends that people start using the programs of the BLAST+ package instead.</source>
        <translation type="vanished">&lt;i&gt; blastall &lt;/i&gt;это программа разработанная NCBI для запуска BLAST поиска. NCBI рекомендует использовать более новый пакет BLAST+.</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportContext</name>
    <message>
        <source>Query with local BLAST...</source>
        <translation type="vanished">Поиск с помощью локальной базы данных BLAST...</translation>
    </message>
    <message>
        <source>Sequence object is NULL</source>
        <translation type="vanished">Sequence object is NULL</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportMultiTask</name>
    <message>
        <source>Source file</source>
        <translation type="vanished">Исходный файл</translation>
    </message>
    <message>
        <source>Used database</source>
        <translation type="vanished">Используемая база данных</translation>
    </message>
    <message>
        <source>No any results found</source>
        <translation type="vanished">Результатов не найдено</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportRunDialog</name>
    <message>
        <source>Wrong parameters for creating annotations</source>
        <translation type="vanished">Неверные параметры для создания аннотаций</translation>
    </message>
    <message>
        <source>Error</source>
        <translation type="vanished">Ошибка</translation>
    </message>
    <message>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation type="vanished">Невозможно создать аннотацию. Проверьте настройки</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllSupportTask</name>
    <message>
        <source>A problem occurred during doing BLAST. The sequence is no more available.</source>
        <translation type="vanished">A problem occurred during doing BLAST. The sequence is no more available.</translation>
    </message>
    <message>
        <source>Subdirectory for temporary files exists. Can not remove this folder.</source>
        <translation type="vanished">Subdirectory for temporary files exists. Can not remove this folder.</translation>
    </message>
    <message>
        <source>Can not create folder for temporary files.</source>
        <translation type="vanished">Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <source>Can not create fake NCBI ini file</source>
        <translation type="vanished">Can not create fake NCBI ini file</translation>
    </message>
    <message>
        <source>Output file not found</source>
        <translation type="vanished">Выходной файл не найден</translation>
    </message>
    <message>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation type="vanished">Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <source>Can not remove folder for temporary files.</source>
        <translation type="vanished">Can not remove folder for temporary files.</translation>
    </message>
    <message>
        <source>Incorrect number of fields in line: %1</source>
        <translation type="vanished">Incorrect number of fields in line: %1</translation>
    </message>
    <message>
        <source>Can&apos;t get location. Start position is absent in [%1]</source>
        <translation type="vanished">Can&apos;t get location. Start position is absent in [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get location. End position is absent in [%1]</source>
        <translation type="vanished">Can&apos;t get location. End position is absent in [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get hit start location from [%1]</source>
        <translation type="vanished">Can&apos;t get hit start location from [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get hit end location from [%1]</source>
        <translation type="vanished">Can&apos;t get hit end location from [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get align length from [%1]</source>
        <translation type="vanished">Can&apos;t get align length from [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get gaps from [%1]</source>
        <translation type="vanished">Can&apos;t get gaps from [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t get identity from [%1]</source>
        <translation type="vanished">Can&apos;t get identity from [%1]</translation>
    </message>
    <message>
        <source>Can&apos;t open output file</source>
        <translation type="vanished">Can&apos;t open output file</translation>
    </message>
    <message>
        <source>Can&apos;t get Hsp_query-from element</source>
        <translation type="vanished">Can&apos;t get Hsp_query-from element</translation>
    </message>
    <message>
        <source>Can&apos;t get Hsp_query-to element</source>
        <translation type="vanished">Can&apos;t get Hsp_query-to element</translation>
    </message>
    <message>
        <source>Unknown strand source setting</source>
        <translation type="vanished">Unknown strand source setting</translation>
    </message>
    <message>
        <source>Can&apos;t get location. %1[%2]</source>
        <translation type="vanished">Can&apos;t get location. %1[%2]</translation>
    </message>
    <message>
        <source>Can&apos;t get identity</source>
        <translation type="vanished">Can&apos;t get identity</translation>
    </message>
    <message>
        <source>Can&apos;t get gaps</source>
        <translation type="vanished">Can&apos;t get gaps</translation>
    </message>
    <message>
        <source>Can&apos;t get align length</source>
        <translation type="vanished">Can&apos;t get align length</translation>
    </message>
    <message>
        <source>Can&apos;t evaluate location</source>
        <translation type="vanished">Can&apos;t evaluate location</translation>
    </message>
</context>
<context>
    <name>U2::BlastAllWithExtFileSpecifySupportRunDialog</name>
    <message>
        <source>Select input file</source>
        <translation type="vanished">Выберите входной файл</translation>
    </message>
    <message>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation type="vanished">Этот файл имеет несовместимый формат для поиска BLAST+.</translation>
    </message>
    <message>
        <source>This file does not contain any sequence.</source>
        <translation type="vanished">Этот файл не содержит последовательностей.</translation>
    </message>
    <message>
        <source>Wrong input file</source>
        <translation type="vanished">Неверный входной файл</translation>
    </message>
    <message>
        <source>Wrong parameters for creating annotations</source>
        <translation type="vanished">Неверные параметры для создания аннотаций</translation>
    </message>
</context>
<context>
    <name>U2::BlastDBCmdDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="42"/>
        <source>Fetch</source>
        <translation>Получить</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="43"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdDialog.cpp" line="91"/>
        <source>Set a result FASTA file name</source>
        <translation>Выберите имя результирующего FASTA файла</translation>
    </message>
</context>
<context>
    <name>U2::BlastDBSelectorWidgetController</name>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidgetController.cpp" line="45"/>
        <source>Database path contains space characters.</source>
        <translation>Путь до базы данных содержит пробелы.</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidgetController.cpp" line="50"/>
        <source>Database name contains space characters.</source>
        <translation>Имя базы данных содержит пробелы.</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidgetController.cpp" line="77"/>
        <source>Select a database file</source>
        <translation>Выберите файл базы данных</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastDBSelectorWidgetController.cpp" line="111"/>
        <source>No alias or index file found for selected database.</source>
        <translation>Не найдено алиасов или индексного файла для выбранной базы данных.</translation>
    </message>
</context>
<context>
    <name>U2::BlastDbCmdSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="68"/>
        <source>The &lt;i&gt;BlastDBCmd&lt;/i&gt; fetches protein or nucleotide sequences from BLAST+ database based on a query.</source>
        <translation>&lt;i&gt;BlastDBCmd&lt;/i&gt; получает белковые или нуклеотидные последовательности из базы данных BLAST+ основанной на запросе.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="80"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Путь для BLAST+ %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastDBCmdSupport.cpp" line="81"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupport</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="199"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation>Путь для BLAST+ не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="200"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportCommonTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="68"/>
        <source>A problem occurred during doing BLAST+. The sequence is no more available.</source>
        <translation>A problem occurred during doing BLAST+. The sequence is no more available.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="81"/>
        <source>The task uses a temporary folder to process the data. The folder path is required not to have spaces. Please set up an appropriate path for the &quot;Temporary files&quot; parameter on the &quot;Directories&quot; tab of the UGENE Application Settings.</source>
        <translation>Задача использует временную директорию. Пожалуйста укажите путь до папки используя поле &quot;Временные файлы&quot; на вкладке &quot;Директории&quot; в основных настройках UGENE.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="90"/>
        <source>Subfolder for temporary files exists. Can not remove this folder.</source>
        <translation>Subfolder for temporary files exists. Can not remove this folder.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="95"/>
        <source>Can not create folder for temporary files.</source>
        <translation>Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="108"/>
        <source>Can not create fake NCBI ini file</source>
        <translation>Can not create fake NCBI ini file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="164"/>
        <source>Output file not found</source>
        <translation>Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="166"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="195"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="220"/>
        <source>There were no hits found for your BLAST search.</source>
        <translation>There were no hits found for your BLAST search.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="212"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="253"/>
        <source>Incorrect number of fields in line: %1</source>
        <translation>Incorrect number of fields in line: %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="258"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="263"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="418"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="425"/>
        <source>Can&apos;t get location</source>
        <translation>Can&apos;t get location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="275"/>
        <source>Can&apos;t evaluate location</source>
        <translation>Can&apos;t evaluate location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="281"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="286"/>
        <source>Can&apos;t get hit location</source>
        <translation>Can&apos;t get hit location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="302"/>
        <source>Can&apos;t evaluate hit location</source>
        <translation>Can&apos;t evaluate hit location</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="316"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="481"/>
        <source>Can&apos;t get align length</source>
        <translation>Can&apos;t get align length</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="321"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="473"/>
        <source>Can&apos;t get gaps</source>
        <translation>Can&apos;t get gaps</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="326"/>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="465"/>
        <source>Can&apos;t get identity</source>
        <translation>Can&apos;t get identity</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="355"/>
        <source>Can&apos;t open output file</source>
        <translation>Can&apos;t open output file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="359"/>
        <source>Can&apos;t read output file</source>
        <translation>Can&apos;t read output file</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="448"/>
        <source>Unknown strand source setting</source>
        <translation>Unknown strand source setting</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="454"/>
        <source>Can&apos;t get location. %1[%2]</source>
        <translation>Can&apos;t get location. %1[%2]</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportContext</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="272"/>
        <source>Fetch sequences by &apos;id&apos;</source>
        <translation>Получить последовательности по &apos;id&apos;</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="282"/>
        <source>Query with local BLAST+...</source>
        <translation>Поиск с помощью локальной базы данных BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="326"/>
        <source>Fetch sequences from local BLAST database</source>
        <translation>Извлечь последовательность из локальной базы данных BLAST</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="331"/>
        <source>Fetch sequences by &apos;id&apos; %1</source>
        <translation>Извлечь последовательности по &apos;id&apos; %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="349"/>
        <source>Path for BLAST+ tools not selected.</source>
        <translation>Путь для BLAST+ не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="350"/>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="425"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="398"/>
        <source>Sequence object is NULL</source>
        <translation>Sequence object is NULL</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupport.cpp" line="424"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Путь для BLAST+ %1 не установлен.</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportMultiTask</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="633"/>
        <source>Source file</source>
        <translation>Исходный файл</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="634"/>
        <source>Used database</source>
        <translation>Используемая база данных</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportCommonTask.cpp" line="636"/>
        <source>No any results found</source>
        <translation>Результатов не найдено</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="122"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="162"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation>Путь до &lt;i&gt;BLAST+ %1&lt;/i&gt; не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="195"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Неверные параметры для создания аннотаций</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="208"/>
        <source>Error</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="208"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Невозможно создать аннотацию. Проверьте настройки</translation>
    </message>
</context>
<context>
    <name>U2::BlastPlusWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="252"/>
        <source>Select input file</source>
        <translation>Выберите входной файл</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="305"/>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="351"/>
        <source>Wrong input file</source>
        <translation>Неверный входной файл</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="305"/>
        <source>This file has the incompatible format for the BLAST+ search.</source>
        <translation>Этот файл имеет несовместимый формат для поиска BLAST+.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="351"/>
        <source>This file does not contain sequences.</source>
        <translation>Этот файл не содержит последовательностей.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="401"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="436"/>
        <source>Path for &lt;i&gt;BLAST+ %1&lt;/i&gt; tool not selected.</source>
        <translation>Путь до &lt;i&gt;BLAST+ %1&lt;/i&gt; не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusSupportRunDialog.cpp" line="469"/>
        <source>Wrong parameters for creating annotations</source>
        <translation>Неверные параметры для создания аннотаций</translation>
    </message>
</context>
<context>
    <name>U2::BlastRunCommonDialog</name>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="54"/>
        <source>Restore to default</source>
        <translation>Значения по умолчанию</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="55"/>
        <source>Search</source>
        <translation>Искать</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="56"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="276"/>
        <source>Direct nucleotide alignment</source>
        <translation>Прямое нуклеотидное выравнивание</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="280"/>
        <source>Direct protein alignment</source>
        <translation>Прямое белковое выравнивание</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="284"/>
        <source>Direct protein alignment (on GPU)</source>
        <translation>Прямое белковое выравнивание (на GPU)</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="288"/>
        <source>Protein alignment, input nucleotide is translated input protein before the search</source>
        <translation>Белковое выравнивание, входные нуклеотиды транслируются в белки до поиска</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="292"/>
        <source>Protein alignment, nucleotide database is translated input protein before the search</source>
        <translation>Белковое выравнивание, нуклеотидная база данных транслируется в белковую до поиска</translation>
    </message>
    <message>
        <location filename="../src/utils/BlastRunCommonDialog.cpp" line="296"/>
        <source>Protein alignment, both input query and database are translated before the search</source>
        <translation>Белковое выравнивание, оба запроса и база данных транслируются до поиска</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2AlignTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="152"/>
        <source>Unknown short reads format %1</source>
        <translation>Неизвестный формат ридов %1</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2BuildIndexTask</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Task.cpp" line="47"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation>Reference file &quot;%1&quot; does not exist</translation>
    </message>
</context>
<context>
    <name>U2::Bowtie2Support</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="58"/>
        <source>&lt;i&gt;Bowtie 2 aligner&lt;/i&gt; takes a Bowtie 2 index and a set of sequencing read files and outputs a set of alignments.</source>
        <translation>&lt;i&gt;Выравнивание при помощи Bowtie 2 &lt;/i&gt; принимает индексный файл и набор ридов, в качестве результата работы выдает выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="66"/>
        <source>&lt;i&gt;Bowtie 2 build indexer&lt;/i&gt;  builds a Bowtie index from a set of DNA sequences. It outputs a set of 6 files with suffixes .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, and .rev.2.bt2. These files together constitute the index: they are all that is needed to align reads to that reference. The original sequence files are no longer used by &lt;i&gt;Bowtie 2&lt;/i&gt; once the index is built.</source>
        <translation>&lt;i&gt;Построение индекса при помощи Bowtie 2&lt;/i&gt; строит индекс из набора ДНК последовательностей. Результатом работы являются 6 файлов с суффиксами .1.bt2, .2.bt2, .3.bt2, .4.bt2, .rev.1.bt2, и .rev.2.bt2. Эти файлы вместе образуют индекс: они все необходимы для выравнивания ридов на референсную последовательность.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Support.cpp" line="79"/>
        <source>&lt;i&gt;Bowtie 2 index inspector&lt;/i&gt; extracts information from a Bowtie index about what kind of index it is and what reference sequence were used to build it.</source>
        <translation>&lt;i&gt;Проверка индекса, построенного Bowtie 2&lt;/i&gt; извлекает информацию из индекса о его типе и какая референсная последовательность была использована для его построения.</translation>
    </message>
</context>
<context>
    <name>U2::BowtieAssembleTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="160"/>
        <source>Reference index file &quot;%1&quot; does not exist</source>
        <translation>Reference index file &quot;%1&quot; does not exist</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="240"/>
        <source>Unknown short reads format %1</source>
        <translation>Unknown short reads format %1</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="313"/>
        <source>There is not enough memory on the computer!</source>
        <translation>There is not enough memory on the computer!</translation>
    </message>
</context>
<context>
    <name>U2::BowtieBuildIndexTask</name>
    <message>
        <location filename="../src/bowtie/BowtieTask.cpp" line="48"/>
        <source>Reference file &quot;%1&quot; does not exist</source>
        <translation>Reference file &quot;%1&quot; does not exist</translation>
    </message>
</context>
<context>
    <name>U2::BowtieSupport</name>
    <message>
        <location filename="../src/bowtie/BowtieSupport.cpp" line="60"/>
        <source>&lt;i&gt;Bowtie&lt;/i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end).</source>
        <translation>&lt;i&gt;Bowtie&lt;/i&gt; один из самых быстрых и эффективных инструментов для выравнивания данных секвенирования на референсную последовательность. Так, например, он способен выравнивать короткие (35 нукл.) последовательности ДНК на геном человека со скоростью более 25 миллионов последовательностей / час.</translation>
    </message>
</context>
<context>
    <name>U2::BwaAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="107"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="125"/>
        <source>Please, provide same number of files with downstream and upstream reads.</source>
        <translation>Укажите одинаковое количество файлов с парными ридами.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="197"/>
        <source>Align reads with BWA Multitask</source>
        <translation>Align reads with BWA Multitask</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="238"/>
        <source>Saming reads with BWA Multitask</source>
        <translation>Saming reads with BWA Multitask</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="290"/>
        <source>Log is incomplete</source>
        <translation>Log is incomplete</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="293"/>
        <source>Log is incorrect</source>
        <translation>Log is incorrect</translation>
    </message>
</context>
<context>
    <name>U2::BwaMemAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="316"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="335"/>
        <source>Please, provide same number of files with downstream and upstream reads.</source>
        <translation>Укажите одинаковое количество файлов с парными ридами.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="427"/>
        <source>Align reads with BWA-MEM Multitask</source>
        <translation>Align reads with BWA-MEM Multitask</translation>
    </message>
</context>
<context>
    <name>U2::BwaMemSettingsWidget</name>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="283"/>
        <source>NOTE: bwa mem accepts reads only in FASTA or FASTQ format. Reads should be compiled into a single file for each mate end.</source>
        <translation>NOTE: bwa mem принимает риды только в формате FASTA или FASTQ формате.</translation>
    </message>
</context>
<context>
    <name>U2::BwaSupport</name>
    <message>
        <location filename="../src/bwa/BwaSupport.cpp" line="48"/>
        <source>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; is an efficient program that aligns relatively short nucleotide sequences against a long reference sequence such as the human genome.</source>
        <translation>&lt;i&gt;Burrows-Wheeler Aligner (BWA)&lt;/i&gt; это один из самых популярных программных пакетов для выравнивания коротких последовательностей на референсный геном, например на геном человека.</translation>
    </message>
</context>
<context>
    <name>U2::BwaSwAlignTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="477"/>
        <source>Short reads are not provided</source>
        <translation>Short reads are not provided</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="486"/>
        <source>BWA SW can not align paired reads</source>
        <translation>BWA SW can not align paired reads</translation>
    </message>
</context>
<context>
    <name>U2::BwaSwSettingsWidget</name>
    <message>
        <location filename="../src/bwa/BwaSettingsWidget.cpp" line="216"/>
        <source>NOTE: bwa-sw performs alignment of long sequencing reads (Sanger or 454). It accepts reads only in FASTA or FASTQ format. Reads should be compiled into single file.</source>
        <translation>NOTE: bwa-sw выравнивает длинные риды. Принимает риды только в формате FASTA или FASTQ формате.</translation>
    </message>
</context>
<context>
    <name>U2::BwaTask</name>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="629"/>
        <source>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</source>
        <translation>Multiple read files are not supported by bwa-sw. Please combine your reads into single FASTA file.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaTask.cpp" line="636"/>
        <source>Please, provide same number of files with downstream and upstream reads.</source>
        <translation>Укажите одинаковое количество файлов с парными ридами.</translation>
    </message>
</context>
<context>
    <name>U2::CAP3Support</name>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="70"/>
        <source>&lt;i&gt;CAP3&lt;/i&gt; is a contig assembly program.                    &lt;br&gt;It allows one to assembly long DNA reads (up to 1000 bp).                    &lt;br&gt;Binaries can be downloaded from http://seq.cs.iastate.edu/cap3.html</source>
        <translation>&lt;i&gt;CAP3&lt;/i&gt;это программа для сборки контигов.                    &lt;br&gt;Она позволяет собирать длинные ДНК риды (более 1000 нуклеотидов).                    &lt;br&gt;Бинарный пакет может быть скачан отсюда http://seq.cs.iastate.edu/cap3.html</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="73"/>
        <source>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</source>
        <translation>&lt;br&gt;&lt;br&gt; Huang, X. and Madan, A.  (1999)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="74"/>
        <source>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</source>
        <translation>&lt;br&gt;CAP3: A DNA Sequence Assembly Program,</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="75"/>
        <source>&lt;br&gt;Genome Research, 9: 868-877.</source>
        <translation>&lt;br&gt;Genome Research, 9: 868-877.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="85"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Support.cpp" line="86"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportDialog</name>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="44"/>
        <source>Run</source>
        <translation>Запустить</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="94"/>
        <source>List of input files is empty!</source>
        <translation>Входные файлы не указаны!</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="105"/>
        <source>Result contig file name is not set!</source>
        <translation>Имя для выходного файла не указано!</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="110"/>
        <source>Destination file already exists.
To overwrite the file, press &apos;Replace&apos;.
To save under other name press &apos;Cancel&apos; and change name in &apos;Result contig&apos; field.</source>
        <translation>Файл уже существует.
Чтобы переписать файл нажмите &apos;Переписать&apos;.
Чтобы сохранить под другим именем нажмите &apos;Отмена&apos; и измените имя в поле &apos;Выходной контиг&apos;.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="113"/>
        <source>Replace</source>
        <translation>Переписать</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="141"/>
        <source>Add Sequences to Assembly</source>
        <translation>Добавить последовательности в сборку</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="85"/>
        <source>Set Result Contig File Name</source>
        <translation>Задать имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="45"/>
        <location filename="../src/cap3/CAP3SupportDialog.cpp" line="114"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::CAP3SupportTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="92"/>
        <source>Output file not found</source>
        <translation>Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="94"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="107"/>
        <source>Output file not found: copy from tmp dir failed.</source>
        <translation>Output file not found: copy from tmp dir failed.</translation>
    </message>
</context>
<context>
    <name>U2::CEASSupport</name>
    <message>
        <location filename="../src/ceas/CEASSupport.cpp" line="51"/>
        <source>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - helps to characterize genome-wide protein-DNA interaction patterns from ChIP-chip and ChIP-Seq of both sharp and broad binding factors. It provides statistics on ChIP enrichment at important genome features such as specific chromosome, promoters, gene bodies, or exons, and infers genes most likely to be regulated by a binding factor.</source>
        <translation>&lt;i&gt;CEAS&lt;/i&gt; - Cis-regulatory Element Annotation System - помогает охарактеризовать модели взаимодействия участков геном-белок-ДНК из ChIP-chip и ChIP-Seq факторов. Это предоставляет статистику о важных свойствах, например, хромосома, промоутеры, экзоны и т.д.</translation>
    </message>
</context>
<context>
    <name>U2::CEASSupportTask</name>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="199"/>
        <source>CEAS error: Can not find a required output file %1.</source>
        <translation>CEAS error: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASSupportTask.cpp" line="210"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupport</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="72"/>
        <source>&lt;i&gt;Clustal Omega&lt;/i&gt; is a free sequence alignment software for proteins.</source>
        <translation>&lt;i&gt;Clustal Omega&lt;/i&gt; это свободный инструмент для выравнивания белков.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="82"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="83"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportContext</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="137"/>
        <source>Align with ClustalO...</source>
        <translation>Выравнивание с помощью ClustalO...</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="162"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupport.cpp" line="163"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="48"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="49"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOSupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="97"/>
        <source>ClustalO alignment started</source>
        <translation>ClustalO alignment started</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="118"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="127"/>
        <source>Subfolder for temporary files exists. Can not remove this folder.</source>
        <translation>Subfolder for temporary files exists. Can not remove this folder.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="132"/>
        <source>Can not create folder for temporary files.</source>
        <translation>Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="179"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="181"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="189"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="272"/>
        <source>Failed to apply the result of ClustalO: alignment object is not available!</source>
        <translation>Failed to apply the result of ClustalO: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="277"/>
        <source>ClustalO alignment successfully finished</source>
        <translation>ClustalO alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="290"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="82"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="83"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="94"/>
        <source>Open an alignment file</source>
        <translation>Открыть выравнивание</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="107"/>
        <source>Save an multiple alignment file</source>
        <translation>Сохранить выравнивание</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="132"/>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="134"/>
        <source>Kalign with Align</source>
        <translation>Выравнивание с помощью Kalign</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="132"/>
        <source>Input file is not set!</source>
        <translation>Не указан входной файл!</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOSupportRunDialog.cpp" line="134"/>
        <source>Output file is not set!</source>
        <translation>Не указан выходной файл!</translation>
    </message>
</context>
<context>
    <name>U2::ClustalOWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalo/ClustalOSupportTask.cpp" line="326"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupport</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="72"/>
        <source>&lt;i&gt;ClustalW&lt;/i&gt; is a free sequence alignment software for DNA or proteins.</source>
        <translation>&lt;i&gt;ClustalW&lt;/i&gt;это свободный иснтрумент для выравнивания белков или ДНК последовательностей.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="82"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="83"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportContext</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="137"/>
        <source>Align with ClustalW...</source>
        <translation>Выравнивание с помощью ClustalW...</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="162"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupport.cpp" line="163"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="45"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="46"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWSupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="105"/>
        <source>Unsupported alphabet: %1</source>
        <translation>Unsupported alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="108"/>
        <source>ClustalW alignment started</source>
        <translation>ClustalW alignment started</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="129"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="207"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="209"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="217"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="292"/>
        <source>Failed to apply the result of ClustalW: alignment object is not available!</source>
        <translation>Failed to apply the result of ClustalW: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="297"/>
        <source>ClustalW alignment successfully finished</source>
        <translation>ClustalW alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="310"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="121"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="122"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="138"/>
        <source>Open an alignment file</source>
        <translation>Открыть выравнивание</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="151"/>
        <source>Save an multiple alignment file</source>
        <translation>Сохранить выравнивание</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="196"/>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="198"/>
        <source>Kalign with Align</source>
        <translation>Выравнивание с помощью Kalign</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="196"/>
        <source>Input file is not set!</source>
        <translation>Не указан входной файл!</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWSupportRunDialog.cpp" line="198"/>
        <source>Output file is not set!</source>
        <translation>Не указан выходной файл!</translation>
    </message>
</context>
<context>
    <name>U2::ClustalWWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/clustalw/ClustalWSupportTask.cpp" line="345"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::ConductGOSupport</name>
    <message>
        <location filename="../src/conduct_go/ConductGOSupport.cpp" line="46"/>
        <source>&lt;i&gt;Conduct GO&lt;/i&gt; - For a list of input genes, this tool uses R/BioC packages (GO, GOstats) to identify over represented GO terms.</source>
        <translation>&lt;i&gt;Conduct GO&lt;/i&gt; - Для набора входных генов этот инструмент использует пакеты R/BioC packages (GO, GOstats), чтобы определить представленные GO условия.</translation>
    </message>
</context>
<context>
    <name>U2::ConductGOTask</name>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="104"/>
        <source>Conduct GO warning: Can not find a required output file %1.</source>
        <translation>Conduct GO warning: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOTask.cpp" line="115"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotSupport</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotSupport.cpp" line="51"/>
        <source>&lt;i&gt;ConservationPlot&lt;/i&gt; - Draw conservation plot for many bed files.</source>
        <translation>&lt;i&gt;ConservationPlot&lt;/i&gt; - Рисует консервативный график для многих наборов файлов.</translation>
    </message>
</context>
<context>
    <name>U2::ConservationPlotTask</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="121"/>
        <source>An annotation table object wasn&apos;t found in the workflow data storage</source>
        <translation>An annotation table object wasn&apos;t found in the workflow data storage</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="164"/>
        <source>Conservation Plot error: Can not find a required output file %1.</source>
        <translation>Conservation Plot error: Can not find a required output file %1.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotTask.cpp" line="175"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
</context>
<context>
    <name>U2::ConvertAlignment2Stockholm</name>
    <message>
        <location filename="../src/hmmer/ConvertAlignment2StockholmTask.cpp" line="41"/>
        <source>Convert alignment to Stockholm format</source>
        <translation>Convert alignment to Stockholm format</translation>
    </message>
    <message>
        <location filename="../src/hmmer/ConvertAlignment2StockholmTask.cpp" line="102"/>
        <source>Cannot create a folder for temporary files.</source>
        <translation>Cannot create a folder for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/ConvertAlignment2StockholmTask.cpp" line="110"/>
        <source>File doesn&apos;t contain any multiple alignments.</source>
        <translation>File doesn&apos;t contain any multiple alignments.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/ConvertAlignment2StockholmTask.cpp" line="113"/>
        <source>File contains several multiple alignments. Only the first one is saved to the result file.</source>
        <translation>File contains several multiple alignments. Only the first one is saved to the result file.</translation>
    </message>
</context>
<context>
    <name>U2::CuffdiffSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="43"/>
        <source>Running Cuffdiff task</source>
        <translation>Running Cuffdiff task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="46"/>
        <source>Workflow data storage is NULL</source>
        <translation>Workflow data storage is NULL</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffSupportTask.cpp" line="83"/>
        <source>At least 2 sets of assemblies are required for Cuffdiff</source>
        <translation>At least 2 sets of assemblies are required for Cuffdiff</translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupport</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="67"/>
        <source>&lt;i&gt;Cuffcompare&lt;/i&gt; helps comparing assembled transcripts to a reference annotation, and also tracking transcripts across multiple experiments.</source>
        <translation>&lt;i&gt;Cuffcompare&lt;/i&gt; помогает сравнить собранные транскрипции с референсной аннотацией.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="83"/>
        <source>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;tests for differential expression and regulation in RNA-Seq samples.</source>
        <translation>&lt;i&gt;Cuffdiff&lt;/i&gt; &amp;nbsp;тестирование для дифференциальной экспрессии и регуляции в образцах RNA-Seq.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="98"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts and estimates their abundances.</source>
        <translation>&lt;i&gt;Cufflinks&lt;/i&gt; собирает транскрипции и оценивает их распространенность.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="113"/>
        <source>&lt;i&gt;Cuffmerge&lt;/i&gt; merges together several assemblies.</source>
        <translation>&lt;i&gt;Cuffmerge&lt;/i&gt; соединяет вместе несколько сборок.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupport.cpp" line="125"/>
        <source>&lt;i&gt;Gffread&lt;/i&gt; is used to verify or perform various operations on GFF files.</source>
        <translation>&lt;i&gt;Gffread&lt;/i&gt; используется для проверки и выполнения различных операций с GFF файлами.</translation>
    </message>
</context>
<context>
    <name>U2::CufflinksSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="56"/>
        <source>Running Cufflinks task</source>
        <translation>Running Cufflinks task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="124"/>
        <source>Unable to get an assembly object.</source>
        <translation>Unable to get an assembly object.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksSupportTask.cpp" line="232"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation>An internal error occurred during getting annotations from a %1 output file!</translation>
    </message>
</context>
<context>
    <name>U2::CuffmergeSupportTask</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="49"/>
        <source>Running Cuffmerge task</source>
        <translation>Running Cuffmerge task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="53"/>
        <source>Workflow data storage is NULL</source>
        <translation>Workflow data storage is NULL</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="54"/>
        <source>There are no annotations to process</source>
        <translation>There are no annotations to process</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="141"/>
        <source>Can not create a file: %1</source>
        <translation>Can not create a file: %1</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeSupportTask.cpp" line="193"/>
        <source>An internal error occurred during getting annotations from a %1 output file!</source>
        <translation>An internal error occurred during getting annotations from a %1 output file!</translation>
    </message>
</context>
<context>
    <name>U2::CutadaptSupport</name>
    <message>
        <location filename="../src/cutadapt/CutadaptSupport.cpp" line="48"/>
        <source>&lt;i&gt;cutadapt&lt;/i&gt; removes adapter sequences from high-throughput sequencing data. This is necessary when the reads are longer than the molecule that is sequenced, such as in microRNA data.</source>
        <translation>&lt;i&gt;cutadapt&lt;/i&gt; удаляет последовательности из данных секвенирования с высокой пропускной способностью. Это необходимо, когда риды больше, чем молекулы, которые секвенированы, например, в данных микроРНК.</translation>
    </message>
</context>
<context>
    <name>U2::ETSProjectViewItemsContoller</name>
    <message>
        <source>FormatDB...</source>
        <translation type="vanished">Создать БД BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="54"/>
        <source>BLAST+ make DB...</source>
        <translation>Создать БД BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="77"/>
        <source>BLAST</source>
        <translation>BLAST</translation>
    </message>
    <message>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="vanished">Путь для BLAST %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="92"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Путь для BLAST+ %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/ETSProjectViewItemsContoller.cpp" line="93"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolJustValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="53"/>
        <source>Tool&apos;s path is empty</source>
        <translation>Tool&apos;s path is empty</translation>
    </message>
    <message>
        <source>Tool&apos;s executable isn&apos;t exists</source>
        <translation type="vanished">Tool&apos;s executable isn&apos;t exists</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="59"/>
        <source>External tool is not found: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="86"/>
        <source>Scripting tool registry is NULL</source>
        <translation>Scripting tool registry is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="88"/>
        <source>Scripting tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation>Scripting tool &apos;%1&apos; isn&apos;t found in the registry</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="137"/>
        <source>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Tool does not start.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="169"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolManagerImpl</name>
    <message>
        <location filename="../src/ExternalToolManager.cpp" line="383"/>
        <source>A dependency tool isn&apos;t represented in the general tool list. Skip dependency &quot;%1&quot;</source>
        <translation>A dependency tool isn&apos;t represented in the general tool list. Skip dependency &quot;%1&quot;</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchAndValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="376"/>
        <source>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</source>
        <translation>Can not find expected message.&lt;br&gt;It is possible that the specified executable file &lt;i&gt;%1&lt;/i&gt; for %2 tool is invalid. You can change the path to the executable file in the external tool settings in the global preferences.</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="39"/>
        <source>&apos;%1&apos; external tool search task</source>
        <translation>&apos;%1&apos; external tool search task</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="47"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="128"/>
        <source>An external tool &apos;%1&apos; isn&apos;t found in the registry</source>
        <translation>An external tool &apos;%1&apos; isn&apos;t found in the registry</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="116"/>
        <source>Tool pointer is NULL</source>
        <translation>Tool pointer is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="124"/>
        <source>External tool module hasn&apos;t any dependencies: it hasn&apos;t master tool</source>
        <translation>External tool module hasn&apos;t any dependencies: it hasn&apos;t master tool</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="126"/>
        <source>External tool registry is NULL</source>
        <translation>External tool registry is NULL</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="132"/>
        <source>Tool&apos;s executable name is not set</source>
        <translation>Tool&apos;s executable name is not set</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportPlugin</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="173"/>
        <source>Search tools in PATH</source>
        <translation>Search tools in PATH</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="218"/>
        <source>External tool support</source>
        <translation>Поддержка внешних инструментов</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="218"/>
        <source>Runs other external tools</source>
        <translation>Модуль позволяет использовать внешние программы</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="285"/>
        <source>Align with ClustalW...</source>
        <translation>Выравнивание с помощью ClustalW...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="293"/>
        <source>Align with ClustalO...</source>
        <translation>Выравнивание с помощью ClustalO...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="301"/>
        <source>Align with MAFFT...</source>
        <translation>Выравнивание с помощью MAFFT...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="309"/>
        <source>Align with T-Coffee...</source>
        <translation>Выравнивание с помощью T-Coffee...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="459"/>
        <source>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) finds regions of local similarity between sequences. The program compares nucleotide or protein sequences to sequence databases and calculates the statistical significance of matches. BLAST can be used to infer functional and evolutionary relationships between sequences as well as help identify members of gene families.</source>
        <translation>The &lt;i&gt;Basic Local Alignment Search Tool&lt;/i&gt; (BLAST) ищет регионы локального сходства между последовательностями. Программа сравнивает нуклеотидные или белковые последовательности в базах данных последовательностей и вычисляет статистическую значимость совпадений. BLAST может использоваться для выявления функциональных и эволюционных взаимосвязей между последовательностями, а также для определения членов семейств гена.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="463"/>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="465"/>
        <source>&lt;i&gt;BLAST+&lt;/i&gt; is a new version of the BLAST package from the NCBI.</source>
        <translation>&lt;i&gt;BLAST+&lt;/i&gt; это новая версия пакета BLAST от NCBI.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="467"/>
        <source>&lt;i&gt;Bowtie&lt;i&gt; is an ultrafast, memory-efficient short read aligner. It aligns short DNA sequences (reads) to the human genome at a rate of over 25 million 35-bp reads per hour. Bowtie indexes the genome with a Burrows-Wheeler index to keep its memory footprint small: typically about 2.2 GB for the human genome (2.9 GB for paired-end). &lt;a href=&apos;http://qt-project.org/doc/qt-4.8/qtextbrowser.html#anchorClicked&apos;&gt;Link text&lt;/a&gt; </source>
        <translation>&lt;i&gt;Bowtie&lt;i&gt;один из самых быстрых и эффективных инструментов для выравнивания данных секвенирования на референсную последовательность. Так, например, он способен выравнивать короткие (35 нукл.) последовательности ДНК на геном человека со скоростью более 25 миллионов последовательностей / час. &lt;a href=&apos;http://qt-project.org/doc/qt-4.8/qtextbrowser.html#anchorClicked&apos;&gt;Link text&lt;/a&gt;</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="474"/>
        <source>&lt;i&gt;Cufflinks&lt;/i&gt; assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples. It accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. It also estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols. </source>
        <translation>&lt;i&gt;Cufflinks&lt;/i&gt; собирает транскрипции и оценивает их распространенность. </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="481"/>
        <source>&lt;i&gt;Bowtie 2&lt;/i&gt; is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. It is particularly good at aligning reads of about 50 up to 100s or 1000s of characters, and particularly good at aligning to relatively long (e.g. mammalian) genomes. &lt;br/&gt;&lt;br/&gt;It indexes the genome with an FM index to keep its memory footprint small: for the human genome, its memory footprint is typically around 3.2Gb. &lt;br/&gt;&lt;br/&gt;&lt;i&gt;Bowtie 2&lt;/i&gt; supports gapped, local, and paired-end alignment modes.</source>
        <translation>&lt;i&gt;Bowtie 2&lt;/i&gt; один из самых быстрых и эффективных инструментов для выравнивания данных секвенирования на большие референсные последовательности. Особенно эффективно применение инструмента для выравнивания ридов начиная от 50 и до сотен и тысяч символов, и больших геномов (например млекопитающих).</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="489"/>
        <source>&lt;i&gt;Cistrome&lt;/i&gt; is a UGENE version of Cistrome pipeline which also includes some tools useful for ChIP-seq analysisThis pipeline is aimed to provide the following analysis steps: peak calling and annotating, motif search and gene ontology.</source>
        <translation>&lt;i&gt;Cistrome&lt;/i&gt; это версия UGENE пайплайна Cistrome, который включает некоторые инструменты, используемые в ChIP-seq анализе.</translation>
    </message>
    <message>
        <source>BLAST make database...</source>
        <translation type="vanished">Создание базы данных BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="492"/>
        <source>BLAST+ make database...</source>
        <translation>Создание базы данных BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="496"/>
        <source>Map reads to reference...</source>
        <translation>Выравнивание прочтений на референсную последовательность...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="520"/>
        <source>Reads de novo assembly (with %1)...</source>
        <translation>Сборка прочтений de novo (при помощи %1)...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="507"/>
        <source>BLAST+ search...</source>
        <translation>Поиск с помощью BLAST+...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="511"/>
        <source>BLAST+ query database...</source>
        <translation>Запрос с помощью BLAST+...</translation>
    </message>
    <message>
        <source>BLAST search...</source>
        <translation type="vanished">Поиск с помощью BLAST...</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="35"/>
        <source>The subfolder for temporary files &apos;%1&apos; already exists. Can not remove this folder!</source>
        <translation>The subfolder for temporary files &apos;%1&apos; already exists. Can not remove this folder!</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportL10N.h" line="41"/>
        <source>Can not create a subfolder for temporary files &apos;%1&apos;!</source>
        <translation>Не удаётся создать папку для временных файлов &apos;%1&apos;!</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportService</name>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="711"/>
        <source>External tools support</source>
        <translation>Поддержка внешних инструментов</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportPlugin.cpp" line="711"/>
        <source>Provides support to run external tools from UGENE</source>
        <translation>Позволяет запускать внешние инструменты из UGENE</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageController</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="53"/>
        <source>External Tools</source>
        <translation>Внешние инструменты</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolSupportSettingsPageWidget</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="678"/>
        <source>Select an external tool to view more information about it.</source>
        <translation>Выберите инструмент, чтобы увидеть дополнительную информацию о нем.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="535"/>
        <source>No description</source>
        <translation>Нет описания</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="486"/>
        <source>External tool &apos;%1&apos; cannot be validated as it depends on other tools, some of which are not valid. The list of tools is the following: </source>
        <translation>Инструмент &apos;%1&apos; не может быть проверен, т.к. он зависит от других инструментов, которые не проверены. Это следующие инструменты: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="109"/>
        <source>Supported tools</source>
        <translation>Поддерживаемые программы</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="112"/>
        <source>Custom tools</source>
        <translation>Кастомные инструменты</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="115"/>
        <source>Additional information</source>
        <translation>Дополнительная информация</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="244"/>
        <source>Select configuration file to import</source>
        <translation>Выберите файл для импорта</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="265"/>
        <source>Can&apos;t remove custom external tool config file from the storage folder: %1</source>
        <translation>Can&apos;t remove custom external tool config file from the storage folder: %1</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="514"/>
        <source>&apos;%1&apos; is %2 module and it is not installed. Install it and restart UGENE or set another %2 with already installed &apos;%1&apos; module.</source>
        <translation>&apos;%1&apos; это %2 модуль и он не установлен. Установите его и перезапустите UGENE или выберите другой %2 с уже установленным модулем &apos;%1&apos;.</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="544"/>
        <source>&lt;br&gt;&lt;br&gt;Version: </source>
        <translation>&lt;br&gt;&lt;br&gt;Версия: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="548"/>
        <source>unknown</source>
        <translation>неизвестный</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="553"/>
        <source>&lt;br&gt;&lt;br&gt;Binary path: </source>
        <translation>&lt;br&gt;&lt;br&gt;Путь: </translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="723"/>
        <source>Choose Folder With Executables</source>
        <translation>Выберите папку с исполняемыми файлами</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="778"/>
        <source>Choose Folder With External Tools Pack</source>
        <translation>Выберите папку пакетом внешних инструментов</translation>
    </message>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="831"/>
        <source>Not a valid external tools folder</source>
        <translation>Нет подходящей папки внешних инструментов</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolUtils</name>
    <message>
        <location filename="../src/utils/ExternalToolUtils.cpp" line="57"/>
        <source>Paths for the following tools are not selected: %1.</source>
        <translation>Пути для следующих инструментов не указаны: %1.</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolUtils.cpp" line="58"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="42"/>
        <source>%1 validate task</source>
        <translation>%1 validate task</translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsInstallTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="422"/>
        <source>Installing external tools</source>
        <translation>Installing external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="427"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsSearchTask</name>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="137"/>
        <source>Searching external tools</source>
        <translation>Searching external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="155"/>
        <location filename="../src/utils/ExternalToolSearchTask.cpp" line="157"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::ExternalToolsValidateTask</name>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="393"/>
        <source>Checking external tools</source>
        <translation>Checking external tools</translation>
    </message>
    <message>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="411"/>
        <location filename="../src/utils/ExternalToolValidateTask.cpp" line="413"/>
        <source> failed: </source>
        <translation> failed: </translation>
    </message>
</context>
<context>
    <name>U2::FastQCParser</name>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="77"/>
        <source>FastQC: %1</source>
        <translation>FastQC: %1</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="79"/>
        <source>FastQC failed to process input file &apos;%1&apos;. Make sure each read takes exactly four lines.</source>
        <translation>FastQC failed to process input file &apos;%1&apos;. Make sure each read takes exactly four lines.</translation>
    </message>
</context>
<context>
    <name>U2::FastQCSupport</name>
    <message>
        <location filename="../src/fastqc/FastqcSupport.cpp" line="53"/>
        <source>&lt;i&gt;FastQC&lt;/i&gt;: A quality control tool for high throughput sequence data.</source>
        <translation>&lt;i&gt;FastQC&lt;/i&gt;: Инструмент контроля качества для данных о последовательностях с высокой пропускной способностью.</translation>
    </message>
</context>
<context>
    <name>U2::FastQCTask</name>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="120"/>
        <source>No input URL</source>
        <translation>No input URL</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="125"/>
        <source>The input file &apos;%1&apos; is empty.</source>
        <translation>The input file &apos;%1&apos; is empty.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="131"/>
        <source>Folder does not exist: %1</source>
        <translation>Folder does not exist: %1</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="148"/>
        <source>Result file does not exist: %1. See the log for details.</source>
        <translation>Result file does not exist: %1. See the log for details.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcTask.cpp" line="160"/>
        <source>Unable to move result file from temporary directory to desired location: %1.</source>
        <translation>Unable to move result file from temporary directory to desired location: %1.</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupport</name>
    <message>
        <source>The &lt;i&gt;formatdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by &lt;i&gt;blastall&lt;/i&gt;.</source>
        <translation type="vanished">The &lt;i&gt;formatdb&lt;/i&gt; создает белковые или нуклеотидные базы данных до того как эти базы будут использованы поиском &lt;i&gt;blastall&lt;/i&gt;.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupport.cpp" line="72"/>
        <location filename="../src/blast_plus/FormatDBSupport.cpp" line="87"/>
        <source>The &lt;i&gt;makeblastdb&lt;/i&gt; formats protein or nucleotide source databases before these databases can be searched by other BLAST+ tools.</source>
        <translation>The &lt;i&gt;makeblastdb&lt;/i&gt; создает белковые или нуклеотидные базы данных до того как эти базы будут использованы другими инструментами BLAST+.</translation>
    </message>
    <message>
        <source>Path for BLAST %1 tool not selected.</source>
        <translation type="vanished">Путь для BLAST %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupport.cpp" line="100"/>
        <source>Path for BLAST+ %1 tool not selected.</source>
        <translation>Путь для BLAST+ %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupport.cpp" line="101"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportRunDialog</name>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="43"/>
        <source>Format</source>
        <translation>Создать</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="44"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="74"/>
        <source>Select file(s)</source>
        <translation>Выберите файл(ы)</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="97"/>
        <source>Select a folder with input files</source>
        <translation>Выберите папку со входным файлом</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="108"/>
        <source>Select a folder to save database files</source>
        <translation>Выберите папку для сохранения базы данных</translation>
    </message>
    <message>
        <source>Input files paths contain space characters.</source>
        <translation type="vanished">Путь до входных файлов содержат пробелы.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="118"/>
        <location filename="../src/blast_plus/FormatDBSupportRunDialog.cpp" line="123"/>
        <source>Output database path contain space characters.</source>
        <translation>Путь до выходного файла содержит пробелы.</translation>
    </message>
</context>
<context>
    <name>U2::FormatDBSupportTask</name>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="56"/>
        <source>Run NCBI FormatDB task</source>
        <translation>Run NCBI FormatDB task</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="151"/>
        <source>Cannot create temp folder</source>
        <translation>Cannot create temp folder</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="168"/>
        <source>Trying to initialize Format DB task second time</source>
        <translation>Trying to initialize Format DB task second time</translation>
    </message>
    <message>
        <source>Input files paths contain space characters.</source>
        <translation type="vanished">Input files paths contain space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="179"/>
        <source>Output database path contain space characters.</source>
        <translation>Output database path contain space characters.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="103"/>
        <source>Blast database creation has been cancelled</source>
        <translation>Создание базы данных BLAST было отменено</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="93"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="110"/>
        <source>Blast database has been successfully created</source>
        <translation>База данных BLAST успешно создана</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="111"/>
        <source>Source sequences: </source>
        <translation>Исходные последовательности: </translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="119"/>
        <source>Database file path: %1</source>
        <translation>Путь до базы данных: %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="121"/>
        <source>Type: %1</source>
        <translation>Тип: %1</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="123"/>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="129"/>
        <source>Formatdb log file path: </source>
        <translation>Formatdb log file path: </translation>
    </message>
    <message>
        <location filename="../src/blast_plus/FormatDBSupportTask.cpp" line="127"/>
        <source>Blast database creation has been failed</source>
        <translation>Создание базы данных BLAST завершилось неудачно</translation>
    </message>
</context>
<context>
    <name>U2::GTest_CompareHmmFiles</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="337"/>
        <source>File #1 not set</source>
        <translation>File #1 not set</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="343"/>
        <source>File #2 not set</source>
        <translation>File #2 not set</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="403"/>
        <source>Error creating ioadapter for first file</source>
        <translation>Error creating ioadapter for first file</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="407"/>
        <source>Error opening 1 file</source>
        <translation>Error opening 1 file</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="414"/>
        <source>Error creating ioadapter for second file</source>
        <translation>Error creating ioadapter for second file</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="418"/>
        <source>Error opening second file</source>
        <translation>Error opening second file</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="443"/>
        <source>Names of aligments not matched</source>
        <translation>Names of aligments not matched</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="447"/>
        <source>Comparing files length not matched</source>
        <translation>Comparing files length not matched</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="454"/>
        <source>Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</source>
        <translation>Files parts not equal:&apos;%1&apos; and &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMM3SearchCompare</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchTaskTest.cpp" line="250"/>
        <source>Can&apos;t parse significance:%1</source>
        <translation>Can&apos;t parse significance:%1</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTaskTest.cpp" line="259"/>
        <source>Internal error (cannot parse float number from string &apos;%1&apos;)</source>
        <translation>Internal error (cannot parse float number from string &apos;%1&apos;)</translation>
    </message>
</context>
<context>
    <name>U2::GTest_UHMMER3Build</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="262"/>
        <source>No input file given</source>
        <translation>No input file given</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTaskTest.cpp" line="268"/>
        <source>No output file given</source>
        <translation>No output file given</translation>
    </message>
</context>
<context>
    <name>U2::Gene2PeakFormatLoader</name>
    <message>
        <location filename="../src/peak2gene/Gene2PeakFormatLoader.cpp" line="102"/>
        <source>Incorrect start position at line %1: &apos;%2&apos;</source>
        <translation>Incorrect start position at line %1: &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Gene2PeakFormatLoader.cpp" line="108"/>
        <source>Incorrect end position at line %1: &apos;%2&apos;</source>
        <translation>Incorrect end position at line %1: &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Gene2PeakFormatLoader.cpp" line="126"/>
        <source>Incorrect NA value at line %1: &apos;%2&apos;</source>
        <translation>Incorrect NA value at line %1: &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::GffreadSupportTask</name>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="34"/>
        <source>Running Gffread task</source>
        <translation>Running Gffread task</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="57"/>
        <source>Unknown file format: %1</source>
        <translation>Unknown file format: %1</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/GffreadSupportTask.cpp" line="66"/>
        <source>The file format is not [%1]: %2</source>
        <translation>The file format is not [%1]: %2</translation>
    </message>
</context>
<context>
    <name>U2::HmmerAdvContext</name>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="328"/>
        <source>Find HMM signals with HMMER3...</source>
        <translation>Поиск HMM сигналов при помощи HMMER3...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="344"/>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="352"/>
        <source>Error</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="352"/>
        <source>Sequences larger 2Gb are not supported on 32-bit architecture.</source>
        <translation>Последовательности более 2 Gb не поддерживаюся на 32-битных системах.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="344"/>
        <source>No sequence in focus found</source>
        <translation>No sequence in focus found</translation>
    </message>
</context>
<context>
    <name>U2::HmmerBuildDialog</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="69"/>
        <source>Build</source>
        <translation>Построить</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="70"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="84"/>
        <source>Select hmm file to create</source>
        <translation>Выберите файл HMM профайла</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="125"/>
        <source>Select multiple alignment file</source>
        <translation>Выберите файл со множественным выравниванием</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="189"/>
        <source>input file is empty</source>
        <translation>Входной файл не выбран</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="192"/>
        <source>output hmm file is empty</source>
        <translation>Выходной HMM файл не выбран</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildDialog.cpp" line="201"/>
        <source>Error: bad arguments!</source>
        <translation>Ошибка: неверные входные параметры!</translation>
    </message>
</context>
<context>
    <name>U2::HmmerBuildFromFileTask</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildFromFileTask.cpp" line="34"/>
        <source>Build HMMER profile from file</source>
        <translation>Build HMMER profile from file</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildFromFileTask.cpp" line="39"/>
        <source>Msa URL is empty</source>
        <translation>Msa URL is empty</translation>
    </message>
</context>
<context>
    <name>U2::HmmerBuildFromMsaTask</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildFromMsaTask.cpp" line="40"/>
        <source>Build HMMER profile from msa</source>
        <translation>Build HMMER profile from msa</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildFromMsaTask.cpp" line="46"/>
        <source>HMM profile URL is empty</source>
        <translation>HMM profile URL is empty</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildFromMsaTask.cpp" line="105"/>
        <source>Cannot create a folder for temporary files.</source>
        <translation>Cannot create a folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::HmmerBuildTask</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="55"/>
        <source>Source alignment</source>
        <translation>Входное множественное выравнивание</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="57"/>
        <source>Profile name</source>
        <translation>HMM профайл</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="59"/>
        <source>Options:</source>
        <translation>Опции:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="60"/>
        <source>Model construction strategies</source>
        <translation>Стратегии построения</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="73"/>
        <source>Relative model construction strategies</source>
        <translation>Относительные веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="76"/>
        <source>Gerstein/Sonnhammer/Chothia tree weights</source>
        <translation>Gerstein/Sonnhammer/Chothia веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="79"/>
        <source>Henikoff simple filter weights</source>
        <translation>Простой фильтр весов Henikoff</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="82"/>
        <source>Henikoff position-based weights</source>
        <translation>Веса Henikoff основанные на позициях</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="85"/>
        <source>No relative weighting; set all to 1</source>
        <translation>Не использовать относительные веса; все равны 1</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="88"/>
        <source>Weights given in MSA file</source>
        <translation>Использовать веса из файла выравнивания</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="95"/>
        <source>Effective sequence weighting strategies</source>
        <translation>Эффективные веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="98"/>
        <source>adjust effective sequence number to achieve relative entropy target</source>
        <translation>Отрегулируйте эффективный порядковый номер, чтобы достичь цель энтропии</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="101"/>
        <source>effective sequence number is number of single linkage clusters</source>
        <translation>Использовать число единичных кластеров как эффективное</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="104"/>
        <source>no effective sequence number weighting: just use number of sequences</source>
        <translation>Использовать число последовательностей в качестве эффективного</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="107"/>
        <source>set effective sequence number for all models to: %1</source>
        <translation>установить эффективное число последовательностей для всех моделей: %1</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="115"/>
        <source>Task finished with error: &apos;%1&apos;</source>
        <translation>Задание закончилось с ошибкой: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="142"/>
        <source>Unknown model construction strategy</source>
        <translation>Unknown model construction strategy</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="163"/>
        <source>Unknown relative sequence weighting strategy</source>
        <translation>Unknown relative sequence weighting strategy</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildTask.cpp" line="185"/>
        <source>Unknown effective sequence weighting strategy</source>
        <translation>Unknown effective sequence weighting strategy</translation>
    </message>
</context>
<context>
    <name>U2::HmmerMsaEditorContext</name>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="286"/>
        <source>Build HMMER3 profile</source>
        <translation>Построить HMMER3 профиль</translation>
    </message>
</context>
<context>
    <name>U2::HmmerParseSearchResultsTask</name>
    <message>
        <location filename="../src/hmmer/HmmerParseSearchResultsTask.cpp" line="80"/>
        <source>Can&apos;t parse line %1</source>
        <translation>Can&apos;t parse line %1</translation>
    </message>
</context>
<context>
    <name>U2::HmmerSearchDialog</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="69"/>
        <source>Run</source>
        <translation>Запустить</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="70"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="173"/>
        <source>HMM profile is not set</source>
        <translation>HMM профиль не установлен</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="179"/>
        <source>Settings are invalid</source>
        <translation>Некорректные настройки</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="190"/>
        <source>Error</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="190"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Невозможно создать аннотацию. Проверьте настройки</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="198"/>
        <source>Error: bad arguments!</source>
        <translation>Ошибка: плохие аргументы!</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="249"/>
        <source>HMM profile</source>
        <translation>HMM профиль</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchDialog.cpp" line="251"/>
        <source>Select query HMM profile</source>
        <translation>Select query HMM profile</translation>
    </message>
</context>
<context>
    <name>U2::HmmerSearchTask</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="50"/>
        <source>HMMER search</source>
        <translation>HMMER поиск</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="104"/>
        <source>HMM profile used: </source>
        <translation>Использовать HMM профиль:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="107"/>
        <source>Task was not finished</source>
        <translation>Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="113"/>
        <source>Result annotation table: </source>
        <translation>Таблица аннотаций:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="115"/>
        <source>Result annotation group: </source>
        <translation>Группа аннотаций:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="116"/>
        <source>Result annotation name: </source>
        <translation>Имя аннотации:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="118"/>
        <source>Results count: </source>
        <translation>Количество:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="145"/>
        <source>Cannot create a folder for temporary files.</source>
        <translation>Cannot create a folder for temporary files.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="193"/>
        <source>Unknown option controlling model-specific thresholding</source>
        <translation>Unknown option controlling model-specific thresholding</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchTask.cpp" line="243"/>
        <source>Warning: it is not recommended to use HMMER2 models with HMMER3</source>
        <translation>Warning: it is not recommended to use HMMER2 models with HMMER3</translation>
    </message>
</context>
<context>
    <name>U2::HmmerSupport</name>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="139"/>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="166"/>
        <source>Error!</source>
        <translation>Ошибка!</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="139"/>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="166"/>
        <source>Target sequence not selected: no opened annotated dna view</source>
        <translation>Таргетная последовательность не выбрана: невозможно отобразить</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="195"/>
        <source>&lt;i&gt;HMMER build&lt;/i&gt; constructs HMM profiles from multiple sequence alignments.</source>
        <translation>&lt;i&gt;Построение HMMER профиля&lt;/i&gt; строит HMM профили по множественным выравниваниям.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="199"/>
        <source>Build HMM3 profile...</source>
        <translation>Построить HMM3 профиль...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="215"/>
        <source>&lt;i&gt;HMMER search&lt;/i&gt; searches profile(s) against a sequence database.</source>
        <translation>&lt;i&gt;HMMER поиск&lt;/i&gt; ищет профиль(и) по базе данных последоавтельности.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="219"/>
        <source>Search with HMMER3...</source>
        <translation>Поиск при помощи HMMER3...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="235"/>
        <source>&lt;i&gt;PHMMER search&lt;/i&gt; searches a protein sequence against a protein database.</source>
        <translation>&lt;i&gt;PHMMER поиск&lt;/i&gt; ищет протеиновые последовательности по протеиновой базе данных.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="239"/>
        <source>Search with phmmer...</source>
        <translation>Поиск при помощи phmmer...</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="250"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSupport.cpp" line="251"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::ImportCustomToolsTask</name>
    <message>
        <location filename="../src/custom_tools/ImportCustomToolsTask.cpp" line="48"/>
        <source>Import custom external tools configuration</source>
        <translation>Импорт конфигурации внешних инструментов</translation>
    </message>
</context>
<context>
    <name>U2::ImportExternalToolDialog</name>
    <message>
        <location filename="../src/custom_tools/ImportExternalToolDialog.cpp" line="53"/>
        <source>Select configuration file to import</source>
        <translation>Выберите файл для импорта</translation>
    </message>
</context>
<context>
    <name>U2::JavaSupport</name>
    <message>
        <location filename="../src/java/JavaSupport.cpp" line="55"/>
        <source>Java Platform lets you develop and deploy Java applications on desktops and servers.&lt;br&gt;&lt;i&gt;(Requires Java 1.7 or higher)&lt;/i&gt;.&lt;br&gt;Java can be freely downloaded on the official web-site: https://www.java.com/en/download/</source>
        <translation>Java позволяет разрабатывать приложения Java на компьютерах и серверах &lt;br&gt; &lt;i&gt; (Требуется Java 1.7 или выше) &lt;/i&gt; &lt;br&gt; Java можно свободно скачать на официальном веб-сайте:. HTTPS: / /www.java.com/en/download/</translation>
    </message>
</context>
<context>
    <name>U2::LoadCustomExternalToolsTask</name>
    <message>
        <location filename="../src/custom_tools/LoadCustomExternalToolsTask.cpp" line="38"/>
        <source>Load custom external tools</source>
        <translation>Загрузить внешние инструменты</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::AlignToReferenceBlastPrompter</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="85"/>
        <source>Input sequence</source>
        <translation>Входная последовательность</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="85"/>
        <source>Input sequence.</source>
        <translation>Входная последовательность.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="86"/>
        <source>Aligned data</source>
        <translation>Выровненные данные</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="86"/>
        <source>Aligned data.</source>
        <translation>Выровненные данные.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="95"/>
        <source>Multiple Chromatogram Alignment URL</source>
        <translation>Путь до множественного выравнивания хроматограмм</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="96"/>
        <source>Location of a result file with a Multiple Chromatogram Alignment.</source>
        <translation>Путь до файла с множественным выравниванием хроматограмм.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="103"/>
        <source>Reference URL</source>
        <translation>Референс</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="103"/>
        <source>A URL to the file with a reference sequence.</source>
        <translation>Путь до файла с референсной последовательностью.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="106"/>
        <source>Result alignment URL</source>
        <translation>Результат</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="106"/>
        <source>An URL to write the result alignment.</source>
        <translation>Путь для записи результата.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="109"/>
        <source>Mapping min similarity</source>
        <translation>Минимальное сходство при выравнивании</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="109"/>
        <source>Reads, whose similarity with the reference is less than the stated value, will be ignored.</source>
        <translation>Риды, чьё сходство с референсом меньше чем стартовое значение будут проигнорированы.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="112"/>
        <source>Read name in result alignment</source>
        <translation>Имя прочтения в выравнивании</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="112"/>
        <source>Reads in the result alignment can be named either by names of the sequences in the input files or by the input files names. For example, if the sequences have the same name, set this value to &quot;File name&quot; to be able to distinguish the reads in the result alignment.</source>
        <translation>Риды в выходном выравнивании могут быть названы по имени последовательностей во входном файле или по именам входных файлов. Например, если последовательности имеют одинаковые имена установите этот параметр в значение &quot;Имя файла&quot;, чтобы отличить риды в выходном выравнивании.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="156"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="159"/>
        <source>Aligns each sequence from &lt;u&gt;%1&lt;/u&gt; to the reference sequence from &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation>Выравнивает каждую последовательность из &lt;u&gt;%1&lt;/u&gt; на референсную последовательность из &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::AlignToReferenceBlastTask</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="250"/>
        <source>Map to reference</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="322"/>
        <source>Details</source>
        <translation>Детальная информация</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="323"/>
        <source>Reference sequence:</source>
        <translation>Референсная последовательность:</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="324"/>
        <source>Mapped reads (%1):</source>
        <translation>Выровненные прочтения (%1):</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="330"/>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="337"/>
        <source>similarity</source>
        <translation>сходство</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="335"/>
        <source>Filtered by low similarity (%1):</source>
        <translation>Прочтения, отфильтрованные из-за недостаточного сходства (%1):</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::AlignToReferenceBlastWorker</name>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="133"/>
        <source>Map to Reference</source>
        <translation>Выровнять на референс</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="133"/>
        <source>Align input sequences (e.g. Sanger reads) to the reference sequence.</source>
        <translation>Выровнять входные последовательности (например Сэнгеровские риды) на референсную последовательность.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/AlignToReferenceBlastWorker.cpp" line="223"/>
        <source>The result file was not produced</source>
        <translation>The result file was not produced</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::AvgQualStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/AvgQualStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;AVGQUAL&lt;/h4&gt;&lt;p&gt;This step drops a read if the average quality is below the specified level.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Quality threshold&lt;/b&gt;: the minimum average quality required to keep a read.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/AvgQualStep.cpp" line="47"/>
        <source>The minimum average quality required to keep a read.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BaseShortReadsAlignerWorker</name>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="250"/>
        <source>Not enough upstream reads datasets</source>
        <translation>Недостаточно upstream ридов</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="253"/>
        <source>Not enough downstream reads datasets</source>
        <translation>Недостаточно downstream ридов</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="315"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="316"/>
        <source>Folder to save output files.</source>
        <translation>Папка для сохранения выходных файлов.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="299"/>
        <source>Reference genome</source>
        <translation>Референсный геном</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="300"/>
        <source>Path to indexed reference genome.</source>
        <translation>Путь до индексированного референсного генома.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="319"/>
        <source>Library</source>
        <translation>Библиотека</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="320"/>
        <source>Is this library mate-paired?</source>
        <translation>Это парная библиотека?</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="323"/>
        <source>Filter unpaired reads</source>
        <translation>Фильтровать непарные риды</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="324"/>
        <source>Should the reads be checked for incomplete pairs?</source>
        <translation>Проверять риды для неполных пар?</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="311"/>
        <source>Output file name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="292"/>
        <source>Reference input type</source>
        <translation>Тип референса</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="293"/>
        <source>Select &quot;Sequence&quot; to input a reference genome as a sequence file. &lt;br/&gt;Note that any sequence file format, supported by UGENE, is allowed (FASTA, GenBank, etc.). &lt;br/&gt;The index will be generated automatically in this case. &lt;br/&gt;Select &quot;Index&quot; to input already generated index files, specific for the tool.</source>
        <translation>Выберите &quot;Последовательность&quot; чтобы ввести референсный геном как файл. &lt;br/&gt;Любой формат файла, который поддерживается UGENE допустим (FASTA, GenBank, и т.д.). &lt;br/&gt;Индекс будет сгенерирован автоматически. &lt;br/&gt;Выберите &quot;Индекс&quot; чтобы ввести уже готовый файл с индексом.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="304"/>
        <source>The folder with the index for the reference sequence.</source>
        <translation>Папка с индексом для референсной последоательности.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="308"/>
        <source>The basename of the index for the reference sequence.</source>
        <translation>Базовое имя для индекса референсной последовательности.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="312"/>
        <source>Base name of the output file. &apos;out.sam&apos; by default</source>
        <translation>Базовое имя для выходных файлов. По умолчанию &apos;out.sam&apos;</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="352"/>
        <source>Sequence</source>
        <translation>Последовательность</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="353"/>
        <source>Index</source>
        <translation>Индекс</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="378"/>
        <source>URL of a file with reads</source>
        <translation>Путь до файла с ридами</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="379"/>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="389"/>
        <source>Input reads to be aligned.</source>
        <translation>Входные риды будут выровнены.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="381"/>
        <source>URL of a file with mate reads</source>
        <translation>Путь до файла с ридами</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="382"/>
        <source>Input mate reads to be aligned.</source>
        <translation>Входные риды будут выровнены.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="388"/>
        <source>Input data</source>
        <translation>Входные данные</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="392"/>
        <source>Reverse FASTQ file</source>
        <translation>Обратный FASTQ файл</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="393"/>
        <source>Reverse paired reads to be aligned.</source>
        <translation>Обратные парные риды будут выровнены.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="405"/>
        <source>Assembly URL</source>
        <translation>Путь до сборки</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="406"/>
        <source>Output assembly URL.</source>
        <translation>Путь выходной сборки.</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="409"/>
        <source>Output data</source>
        <translation>Выходные данные</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="410"/>
        <source>Output assembly files.</source>
        <translation>Выходные файлы сборки.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigPrompter</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="78"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="79"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="81"/>
        <source>Converts bedGraph files to bigWig %1 with bedGraphToBigWig.</source>
        <translation>Конвертирует файлы bedGraph в bigWig %1 при помощи bedGraphToBigWig.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedGraphToBigWigWorker</name>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="98"/>
        <source>Convert bedGraph Files to bigWig</source>
        <translation>Преобразование файлов bedGraph в bigWig</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="98"/>
        <source>Convert bedGraph Files to bigWig.</source>
        <translation>Преобразование файлов bedGraph в bigWig.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="102"/>
        <source>BedGrapgh files</source>
        <translation>Файлы BedGrapgh</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="102"/>
        <source>Set of bedGraph files</source>
        <translation>Укажите bedGraph файлы</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="103"/>
        <source>BigWig files</source>
        <translation>BigWig файлы</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="116"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="116"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="120"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="120"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="122"/>
        <source>Output name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="122"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first BAM file with an extension.</source>
        <translation>Имя выходного файла. По умолчанию будет взято имя первого файла с соответствующим расширением.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="124"/>
        <source>Block size</source>
        <translation>Размер блока</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="124"/>
        <source>Number of items to bundle in r-tree (-blockSize).</source>
        <translation>Количество пунктов для объединения r-tree (-blockSize).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="126"/>
        <source>Items per slot</source>
        <translation>Пунктов на слот</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="126"/>
        <source>Number of data points bundled at lowest level (-itemsPerSlot).</source>
        <translation>Количество данных в комплекте на низком уровне (-itemsPerSlot).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="128"/>
        <source>Uncompressed</source>
        <translation>Не сжимать</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="128"/>
        <source>If set, do not use compression (-unc).</source>
        <translation>Если установлено, не используется сжатие (-unc).</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="130"/>
        <source>Genome</source>
        <translation>Геном</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="130"/>
        <source>File with genome length.</source>
        <translation>Файл с геномом.</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="160"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="158"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/bigWigTools/BedGraphToBigWigWorker.cpp" line="159"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedtoolsIntersectPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="969"/>
        <source>Intersect annotations from &lt;u&gt;%1&lt;/u&gt; (&lt;b&gt;set A&lt;/b&gt;) with annotations from &lt;u&gt;%2&lt;/u&gt; (&lt;b&gt;set B&lt;/b&gt;). Report </source>
        <translation>Пересечение аннотаций из &lt;u&gt;%1&lt;/u&gt; (&lt;b&gt;набор A&lt;/b&gt;) с аннотациями из &lt;u&gt;%2&lt;/u&gt; (&lt;b&gt;набор B&lt;/b&gt;). Отчет </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BedtoolsIntersectWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="859"/>
        <source>Data not found by %1 id</source>
        <translation>Data not found by %1 id</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="863"/>
        <source>Can not get annotation table object</source>
        <translation>Can not get annotation table object</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="877"/>
        <source>Annotations A</source>
        <translation>Аннотации A</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="882"/>
        <source>Annotations B</source>
        <translation>Аннотации B</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="887"/>
        <source>Annotations</source>
        <translation>Аннотации</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="887"/>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="903"/>
        <source>Result annotations</source>
        <translation>Выходные аннотации</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="896"/>
        <source>Minimum overlap</source>
        <translation>Минимальное перекрытие</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="897"/>
        <source>Minimum overlap required as a fraction of an annotation from set A.&lt;br/&gt;By default, even 1 bp overlap between annotations from set A and set B is taken into account. Yet sometimes you may want to restrict reported overlaps to cases where the annotations in B overlaps at least X% (e.g. 50%) of the A annotation. </source>
        <translation>Минимальное пересечение, требуемое как фракция из набора A.&lt;br/&gt;По умолчанию, это свыше 1 нуклеотида пересечений между аннотациями из набора А и аннотациями набора В.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="904"/>
        <source>Select one of the following:&lt;ul&gt;&lt;li&gt;&lt;i&gt;Shared interval&lt;/i&gt; to report intervals shared between overlapped annotations from set A and set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Overlapped annotations from A&lt;/i&gt; to report annotations from set A that have an overlap with annotations from set B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Non-overlapped annotations from A&lt;/i&gt; to report annotations from set A that have NO overlap with annotations from set B.&lt;/li&gt;&lt;/ul&gt;</source>
        <translation>Выберите одно из следующего:&lt;ul&gt;&lt;li&gt;&lt;i&gt;Общий интервал&lt;/i&gt; включить в отчет общие интервалы пересекающихся аннотаций набора А и набора В.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Пересекающиеся аннотации из A&lt;/i&gt; включить в отчетаннотации из набора A, которые имеют пересечения с аннотациями набора B.&lt;/li&gt;&lt;li&gt;&lt;i&gt;Не пересекающиеся аннотации из набора А&lt;/i&gt; включить в отчет аннотации из набора A, которые не пересекаются с аннотациями из набора B.&lt;/li&gt;&lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="916"/>
        <source>Unique overlaps</source>
        <translation>Уникальное перекрытие</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="917"/>
        <source>If the parameter value is &quot;True&quot;, write original A entry once if any overlaps found in B. In other words, just report the fact at least one overlap was found in B.&lt;br/&gt;The minimum overlap number is ignored in this case.&lt;br/&gt;&lt;br/&gt;If the parameter value is &quot;False&quot;, the A annotation is reported for every overlap found.</source>
        <translation>Если значение параметра &quot;Истина&quot;, запишется оригинальная запись А один раз, если есть какие-либо совпадения в В. Если значение параметра &quot;Ложь&quot;, аннотация А сообщает о всех совпадениях.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="953"/>
        <source>Intersect Annotations</source>
        <translation>Пересечение аннотаций</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="954"/>
        <source>Intersects two sets of annotations denoted as A and B.</source>
        <translation>Перечение двух наборов аннотаций, обозначенных как А и В.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllPrompter</name>
    <message>
        <source>unset</source>
        <translation type="vanished">не указан</translation>
    </message>
    <message>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation type="vanished"> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <source>For sequence %1 find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation type="vanished">Для последовательности %1 ищет аннотации в базе данных &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastAllWorker</name>
    <message>
        <source>Best hits limit</source>
        <translation type="vanished">Предел лучших попыток</translation>
    </message>
    <message>
        <source>Number of best hits from a region to keep. 0 turns it off. If used, 100 is recommended.</source>
        <translation type="vanished">Количество лучших попыток для сохранения. 0 выключает опцию. При использовании рекомендуется значение 100.</translation>
    </message>
    <message>
        <source>Input sequence</source>
        <translation type="vanished">Входная последовательность</translation>
    </message>
    <message>
        <source>Sequence for which annotations is searched.</source>
        <translation type="vanished">Последовательность для поиска аннотаций.</translation>
    </message>
    <message>
        <source>Annotations</source>
        <translation type="vanished">Аннотации</translation>
    </message>
    <message>
        <source>Found annotations.</source>
        <translation type="vanished">Найденные аннотации.</translation>
    </message>
    <message>
        <source>Search type</source>
        <translation type="vanished">Тип поиска</translation>
    </message>
    <message>
        <source>Select type of BLAST searches.</source>
        <translation type="vanished">Выберите тип BLAST поиска.</translation>
    </message>
    <message>
        <source>Database Path</source>
        <translation type="vanished">Путь до базы данных</translation>
    </message>
    <message>
        <source>Path with database files.</source>
        <translation type="vanished">Путь до файлов базы данных.</translation>
    </message>
    <message>
        <source>Database Name</source>
        <translation type="vanished">Имя базы данных</translation>
    </message>
    <message>
        <source>Base name for BLAST DB files.</source>
        <translation type="vanished">Базовое имя для файлов базы данных BLAST.</translation>
    </message>
    <message>
        <source>Expected value</source>
        <translation type="vanished">Ожидаемое значение</translation>
    </message>
    <message>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation type="vanished">Этот параметр определяет статистическую значимость порога совпадений.</translation>
    </message>
    <message>
        <source>Annotate as</source>
        <translation type="vanished">Аннотировать как</translation>
    </message>
    <message>
        <source>Name for annotations.</source>
        <translation type="vanished">Имя для аннотаций.</translation>
    </message>
    <message>
        <source>Tool Path</source>
        <translation type="vanished">Путь</translation>
    </message>
    <message>
        <source>External tool path.</source>
        <translation type="vanished">Путь до внешнего инструмента.</translation>
    </message>
    <message>
        <source>Temporary folder</source>
        <translation type="vanished">Временная папка</translation>
    </message>
    <message>
        <source>Folder for temporary files.</source>
        <translation type="vanished">Папка для временных файлов.</translation>
    </message>
    <message>
        <source>BLAST output</source>
        <translation type="vanished">Выходной файл</translation>
    </message>
    <message>
        <source>Location of BLAST output file.</source>
        <translation type="vanished">Расположение выходного файла.</translation>
    </message>
    <message>
        <source>BLAST output type</source>
        <translation type="vanished">Тип выходного файла</translation>
    </message>
    <message>
        <source>Type of BLAST output file.</source>
        <translation type="vanished">Имп выходного файла.</translation>
    </message>
    <message>
        <source>Gapped alignment</source>
        <translation type="vanished">Выравнивание с пробелами</translation>
    </message>
    <message>
        <source>Perform gapped alignment.</source>
        <translation type="vanished">Выполняет выравнивание с пробелами.</translation>
    </message>
    <message>
        <source>Gap costs</source>
        <translation type="vanished">Вес пробела</translation>
    </message>
    <message>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation type="vanished">Стоимость создания и расширения пробела в выравнивании.</translation>
    </message>
    <message>
        <source>Match scores</source>
        <translation type="vanished">Веса совпадений</translation>
    </message>
    <message>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation type="vanished">Вознаграждения и наказания за совпадения и несовпадения.</translation>
    </message>
    <message>
        <source>Local BLAST search</source>
        <translation type="vanished">Локальный поиск c помощью BLAST</translation>
    </message>
    <message>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation type="vanished">Ищет аннотации для ДНК последовательности в локальной базе данных.</translation>
    </message>
    <message>
        <source>Empty sequence supplied to BLAST</source>
        <translation type="vanished">Empty sequence supplied to BLAST</translation>
    </message>
    <message>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation type="vanished">Selected BLAST search with nucleotide input sequence</translation>
    </message>
    <message>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation type="vanished">Selected BLAST search with amino acid input sequence</translation>
    </message>
    <message>
        <source>Not selected BLAST output file</source>
        <translation type="vanished">Not selected BLAST output file</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusPrompter</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="248"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="249"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="250"/>
        <source>For sequence &lt;u&gt;%1&lt;/u&gt; find annotations in database &lt;u&gt;%2&lt;/u&gt;.</source>
        <translation>Для последовательности %1 ищет аннотации в базе данных &lt;u&gt;%2&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BlastPlusWorker</name>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="89"/>
        <source>Input sequence</source>
        <translation>Входная последовательность</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="89"/>
        <source>Sequence for which annotations is searched.</source>
        <translation>Последовательность для поиска аннотаций.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="90"/>
        <source>Annotations</source>
        <translation>Аннотации</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="90"/>
        <source>Found annotations.</source>
        <translation>Найденные аннотации.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="99"/>
        <source>Search type</source>
        <translation>Тип поиска</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="100"/>
        <source>Database Path</source>
        <translation>Путь до базы данных</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="101"/>
        <source>Database Name</source>
        <translation>Имя базы данных</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="102"/>
        <source>Expected value</source>
        <translation>Ожидаемое значение</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="102"/>
        <source>This setting specifies the statistical significance threshold for reporting matches against database sequences.</source>
        <translation>Этот параметр определяет статистическую значимость порога совпадений.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="105"/>
        <source>Annotate as</source>
        <translation>Аннотировать как</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="106"/>
        <source>Tool Path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="107"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="108"/>
        <source>BLAST output</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="108"/>
        <source>Location of BLAST output file.</source>
        <translation>Расположение выходного файла.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="109"/>
        <source>BLAST output type</source>
        <translation>Тип выходного файла</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="109"/>
        <source>Type of BLAST output file.</source>
        <translation>Имп выходного файла.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="110"/>
        <source>Gapped alignment</source>
        <translation>Выравнивание с пробелами</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="99"/>
        <source>Select type of BLAST+ searches.</source>
        <translation>Выберите тип BLAST+ поиска.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="62"/>
        <source>Culling limit</source>
        <translation>Предел отбраковки</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="66"/>
        <source>If the query range of a hit is enveloped by that of at least this many higher-scoring hits, delete the hit</source>
        <translation>Если диапазон запроса попытки оограничен тем, что много попыток с высокой оценкой, удалить попытку</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="100"/>
        <source>Path with database files.</source>
        <translation>Путь до файлов базы данных.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="101"/>
        <source>Base name for BLAST+ DB files.</source>
        <translation>Базовое имя для файлов базы данных BLAST+.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="103"/>
        <source>Composition-based statistics</source>
        <translation>Статистика состава</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="103"/>
        <source>Composition-based statistics.</source>
        <translation>Статистика состава.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="105"/>
        <source>Name for annotations.</source>
        <translation>Имя для аннотаций.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="106"/>
        <source>External tool path.</source>
        <translation>Путь до внешнего инструмента.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="107"/>
        <source>Folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="110"/>
        <source>Perform gapped alignment.</source>
        <translation>Выполняет выравнивание с пробелами.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="112"/>
        <source>Gap costs</source>
        <translation>Вес пробела</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="112"/>
        <source>Cost to create and extend a gap in an alignment.</source>
        <translation>Стоимость создания и расширения пробела в выравнивании.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="113"/>
        <source>Match scores</source>
        <translation>Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="113"/>
        <source>Reward and penalty for matching and mismatching bases.</source>
        <translation>Вознаграждения и наказания за совпадения и несовпадения.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="150"/>
        <source>Local BLAST+ Search</source>
        <translation>Локальный поиск c помощью BLAST+</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="150"/>
        <source>Finds annotations for DNA sequence in local database.</source>
        <translation>Ищет аннотации для ДНК последовательности в локальной базе данных.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="321"/>
        <source>Empty sequence supplied to BLAST</source>
        <translation>Empty sequence supplied to BLAST</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="331"/>
        <source>Selected BLAST search with nucleotide input sequence</source>
        <translation>Selected BLAST search with nucleotide input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="335"/>
        <source>Selected BLAST search with amino acid input sequence</source>
        <translation>Selected BLAST search with amino acid input sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/BlastPlusWorker.cpp" line="342"/>
        <source>Not selected BLAST output file</source>
        <translation>Not selected BLAST output file</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Bowtie2Worker</name>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="110"/>
        <source>Bowtie index folder</source>
        <translation>Папка для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="110"/>
        <source>Bowtie index basename</source>
        <translation>Базовое имя для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="113"/>
        <source>Mode</source>
        <translation>Режим</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="114"/>
        <source>When the -n option is specified (which is the default), bowtie determines which alignments                               
are valid according to the following policy, which is similar to Maq&apos;s default policy.                               
In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               
through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</source>
        <translation>Когда задана опция -n, bowtie определяет, какие выравнивания
действительны в соответствии со следующей политикой, которая похожа на политику Maq по умолчанию.
В -v режиме, выравнивание может иметь не больше чем V несоответствий, где V может быть число от 0. Опция -v является взаимоисключающим с опцией -n.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="120"/>
        <source>Number of mismatches</source>
        <translation>Число несоответствий</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="121"/>
        <source>Sets the number of mismatches to allowed in a seed alignment. Can be set to 0 or 1.                               
Setting this higher makes alignment slower (often much slower) but increases sensitivity.</source>
        <translation>Задает количество несовпадений с разрешенным в выравнивании. Может быть 0 или 1. Повышение этого параметра делает выранивание более медленным (чаще всего очень медленным) но повышает точность.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="125"/>
        <source>Seed length (--L)</source>
        <translation>Длина затравки (--L)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="126"/>
        <source>Sets the length of the seed substrings to align. Smaller values make alignment slower but more senstive.</source>
        <translation>Задает длину затравок для выравнивания. Меньшие значения замедляют выравнивание, но повышают точность.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="129"/>
        <source>Add columns to allow gaps (--dpad)</source>
        <translation>Добавить столбцы, чтобы допустить пробелы (--dpad)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="130"/>
        <source>&quot;Pads&quot; dynamic programming problems by specified number of columns on either side to allow gaps.</source>
        <translation>Добавляет столбцы, чтобы допустить пробелы.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="133"/>
        <source>Disallow gaps (--gbar)</source>
        <translation>Запретить пробелы (--gbar)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="134"/>
        <source>Disallow gaps within specified number of positions of the beginning or end of the read.</source>
        <translation>Запрещает пробелы в указанных позициях в начале или в конце рида.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="137"/>
        <source>Seed (--seed)</source>
        <translation>Затравка (--seed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="138"/>
        <source>Use specified value as the seed for pseudo-random number generator.</source>
        <translation>Использует указанное значение в качестве затравки для псевдо-случайного генератора.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="141"/>
        <source>Threads</source>
        <translation>Потоки</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="142"/>
        <source>Launch specified number of parallel search threads. Threads will run on separate processors/cores and synchronize                               
when parsing reads and outputting alignments. Searching for alignments is highly parallel, and speedup is close to linear.</source>
        <translation>Запуск определенного количество параллельных потоков поиска. Потоки будут работать на отдельных процессорах/ ядрах и синхронизоваться при разборе ридов и выводе выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="146"/>
        <source>No unpaired alignments (--no-mixed)</source>
        <translation>Только парное выравнивание (--no-mixed)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="147"/>
        <source>If Bowtie2 cannot find a paired-end alignment for a pair, by default it will go on to look for unpaired alignments                               
for the constituent mates. This is called &quot;mixed mode.&quot; To disable mixed mode, set this option. Bowtie2 runs a little                               
faster in the mixed mode, but will only consider alignment status of pairs per se, not individual mates.</source>
        <translation>Если Bowtie2 не сможет найти пару для выполнения парного выравнивания по умолчанию будет сделано непарное выравнивание. Это называется &amp;quot;смешанным режимом.&amp;quot; Чтобы отключить смешанный режим, установите эту опцию. Bowtie2 запустится немного быстрее в смешанном режиме чем для парного выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="152"/>
        <source>No discordant alignments (--no-discordant)</source>
        <translation>Только согласованные выравнивания (--no-discordant)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="153"/>
        <source>By default, Bowtie2 looks for discordant alignments if it cannot find any concordant alignments. A discordant alignment                               
is an alignment where both mates align uniquely, but that does not satisfy the paired-end constraints. This option disables that behavior.</source>
        <translation>По умолчанию, Bowtie2 ищет несогласованные выравнивания, если он не может найти какие-либо согласованные выравнивания. Несогласованное выравнивание это выравнивание, в котором оба рида выровнены однозначно, но они не удовлетворяют условию парности. Это опция отключает это поведение.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="157"/>
        <source>No forward orientation (--nofw)</source>
        <translation>Не прямая ориентация (--nofw)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="158"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>Если установлена опция --nofw, bowtie  не будет пытаться выровнять прямую цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="161"/>
        <source>No reverse-complement orientation (--norc)</source>
        <translation>Не обратно-комплементарная ориентация (--norc)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="162"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>Если установлена опция --norc, bowtie не будет пытаться выровнять обратно-комплементарную цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="165"/>
        <source>No overlapping mates (--no-overlap)</source>
        <translation>Без перекрытий (--no-overlap)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="166"/>
        <source>If one mate alignment overlaps the other at all, consider that to be non-concordant. Default: mates can overlap in                                 
a concordant alignment.</source>
        <translation>Если один участник выравнивания перекрывается с другим, то это несогласованное выравнивание. По умолчанию: участники могут перекрываться в согласованном выравнивании.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="170"/>
        <source>No mates containing one another (--no-contain)</source>
        <translation>Без вложений (--no-contain)</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="171"/>
        <source>If one mate alignment contains the other, consider that to be non-concordant. Default: a mate can contain the                                
other in a concordant alignment.</source>
        <translation>Если один участник выравнивания содержит другой, то это несогласованное выравнивание. По умолчанию одна часть выравнивания может содержать другую в согласованном выравнивании.</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="221"/>
        <source>Map Reads with Bowtie2</source>
        <translation>Выравнивание с помощью Bowtie2</translation>
    </message>
    <message>
        <location filename="../src/bowtie2/Bowtie2Worker.cpp" line="222"/>
        <source>Bowtie2  is a program for mapping short DNA sequence reads to a long reference sequence. In addition to to the Burrows-Wheeler transform, Bowtie2 uses an FM-index (similar to a suffix array) to keep its memory footprint small.&lt;br/&gt;&lt;br/&gt;Bowtie2 is more suited to finding longer, gapped alignments than the first version Bowtie, it supports ambiguous characters in the reference (e.g. &apos;N&apos;), etc.&lt;br/&gt;&lt;br/&gt;Provide URL(s) to FASTA or FASTQ file(s) with NGS reads to the input port of the element, set up the reference sequence in the parameters. The result is saved to the specified SAM file, URL to the file is passed to the output port.</source>
        <translation>Выполняет выравнивание коротких ридов при помощи Bowtie2.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BowtieWorker</name>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="117"/>
        <source>Bowtie index folder</source>
        <translation>Папка для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="117"/>
        <source>Bowtie index basename</source>
        <translation>Базовое имя для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="137"/>
        <source>Mode</source>
        <translation>Режим</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="138"/>
        <source>When the -n option is specified (which is the default), bowtie determines which alignments                               are valid according to the following policy, which is similar to Maq&apos;s default policy.                               In -v mode, alignments may have no more than V mismatches, where V may be a number from 0                               through 3 set using the -v option. Quality values are ignored. The -v option is mutually exclusive with the -n option.</source>
        <translation>Когда задана опция -n, bowtie определяет, какие выравнивания
действительны в соответствии со следующей политикой, которая похожа на политику Maq по умолчанию.
В -v режиме, выравнивание может иметь не больше чем V несоответствий, где V может быть число от 0. Опция -v является взаимоисключающим с опцией -n.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="144"/>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="148"/>
        <source>Mismatches number</source>
        <translation>Число несоответствий</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="145"/>
        <source>Mismatches number.</source>
        <translation>Число несоответствий.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="149"/>
        <source>Maximum permitted total of quality values at all mismatched read positions throughout the entire alignment,                              not just in the seed. The default is 70. Like Maq, bowtie rounds quality values to the nearest 10 and saturates at 30;                              rounding can be disabled with --nomaqround.</source>
        <translation>Максимальн допустимое общее значение качества всех несогласованных позиций ридов на протяжении всего выравнивания.
По умолчанию 70.
Подобно Maq, значения качества близки к 10 и максимальны к значению 30, округление может быть отключено при помощи --nomaqround.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="154"/>
        <source>Seed length</source>
        <translation>Длина затравки</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="155"/>
        <source>The seed length; i.e., the number of bases on the high-quality end of the read to which the                               -n ceiling applies. The lowest permitted setting is 5 and the default is 28. bowtie is faster for larger values of -l.</source>
        <translation>&amp;quot;Длина затравки&amp;quot; т.е., число оснований в высококачественном конце рида, для которого применяется -n.
Наименьшее допустимое значение 5, по умолчанию 28. Большие значения -l увеличивают скорость.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="161"/>
        <source>No forward orientation</source>
        <translation>Не прямая ориентация</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="162"/>
        <source>If --nofw is specified, bowtie will not attempt to align against the forward reference strand.</source>
        <translation>Если установлена опция --nofw, bowtie  не будет пытаться выровнять прямую цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="165"/>
        <source>No reverse-complement orientation</source>
        <translation>Не обратно-комплементарная ориентация</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="166"/>
        <source>If --norc is specified, bowtie will not attempt to align against the reverse-complement reference strand.</source>
        <translation>Если установлена опция --norc, bowtie не будет пытаться выровнять обратно-комплементарную цепь.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="169"/>
        <source>Maximum of backtracks</source>
        <translation>Максимальное число отступов</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="170"/>
        <source>The maximum insert size for valid paired-end alignments. E.g. if -X 100 is specified and a paired-end alignment                               consists of two 20-bp alignments in the proper orientation with a 60-bp gap between them, that alignment is                               considered valid (as long as -I is also satisfied). A 61-bp gap would not be valid in that case.                                If trimming options -3 or -5 are also used, the -X constraint is applied with respect to the untrimmed mates,                               not the trimmed mates. Default: 250.</source>
        <translation>Максимальный размер вставки для выравнивания paired-end.
Если -X 100 и выравнивание состоит из участков по 20 нуклеотидов и расстоянием между ними в 60 нуклеотидов, 
такое выравнивание считается действительным. 
В таком случае 61 пропуск не будет считаться допустимым.
Если также используются опции обрезки -3 или -5, 
 -X ограничение применяется к необрезанным участкам.
По умолчанию: 250.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="177"/>
        <source>Try as hard</source>
        <translation>Упорно пытаться</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="178"/>
        <source>Try as hard as possible to find valid alignments when they exist, including paired-end alignments.                                This is equivalent to specifying very high values for the --maxbts and --pairtries options. This mode is generally much                                slower than the default settings, but can be useful for certain problems. This mode is slower when (a) the reference is                                very repetitive, (b) the reads are low quality, or (c) not many reads have valid alignments.</source>
        <translation>Пытаться пока это возможно найти подходящее выравнивание, когда они есть, включая paired-end выравнивания.
Это равносильно очень высоким значениям параметров --maxbts и --pairtries.
Этот режим более медленный, чем настройки по умолчанию,
но может быть применим для некоторых случаев.
Этот режим медленнее, когда (a) референсная последовательность очень сильно состоит из повторяющихся участков,
(b) риды низкого качества,
или (c) не очень много ридов имеют подходящее выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="184"/>
        <source>Best hits</source>
        <translation>Лучшие попытки</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="185"/>
        <source>The number of megabytes of memory a given thread is given to store path descriptors in --best mode. Best-first                                search must keep track of many paths at once to ensure it is always extending the path with the lowest cumulative cost.                                Bowtie tries to minimize the memory impact of the descriptors, but they can still grow very large in some cases.                                If you receive an error message saying that chunk memory has been exhausted in --best mode,                                try adjusting this parameter up to dedicate more memory to the descriptors. Default: 64.</source>
        <translation>Количество мегабайт памяти для данного потока задается для хранения дескрипторов пути в --best режиме.
Поиск по первому наилучшему совпадению должен отслеживать множество путей одновременно, убедитесь, что расширение пути происходит с наименьшей стоимостью.
Bowtie пытается свести к минимуму воздействие памяти дескрипторов, но они по-прежнему могут сильно возрастать в некоторых случаях.
Если вы получаете сообщение об ошибке, говорящее, что участок памяти исчерпан в --best режиме,
попробуйте отрегулировать этот параметр и выделить больше памяти для дескрипторов.
Значение по умолчанию: 64.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="192"/>
        <source>No Maq rounding</source>
        <translation>Без округления Maq</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="193"/>
        <source>Maq accepts quality values in the Phred quality scale, but internally rounds values to the nearest 10,              with a maximum of 30. By default, bowtie also rounds this way. --nomaqround prevents this rounding in bowtie.</source>
        <translation>Maq принимает значения качества в масштабе качества Phred,
но внутренние значения близки к 10,
с максимумом около 30.
По умолчанию, bowtie также вокруг этих значений. --nomaqround препятствует этому округлению в bowtie.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="197"/>
        <source>Seed</source>
        <translation>Затравка</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="198"/>
        <source>Use &lt;int&gt; as the seed for pseudo-random number generator.</source>
        <translation>Используйте &lt;int&gt; в качестве затравки для псевдо-случайного генератора.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="201"/>
        <source>Best alignments</source>
        <translation>Лучшие выравнивания</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="202"/>
        <source>Make Bowtie guarantee that reported singleton alignments are best in terms of stratum                              (i.e. number of mismatches, or mismatches in the seed in the case of -n mode) and in terms of                              the quality values at the mismatched position(s). bowtie is somewhat slower when --best is specified.</source>
        <translation>Создать Bowtie гарантию того, что одиночные выравнивания являются &quot;лучшими&quot; в плане слоя
(т.е. число несовпадений, или несоответствия в затравке в -n режиме)
и с точки зрения значений качества на несогласованной позиции (ях).
bowtie медленнее, когда указан режим --best.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="209"/>
        <source>All alignments</source>
        <translation>Все выравнивания</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="210"/>
        <source>Report all valid alignments per read or pair.</source>
        <translation>Сообщить о всех допустимых выравниваниях для рида или пары.</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="213"/>
        <source>Colorspace</source>
        <translation>Цвета</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="214"/>
        <source>When -C is specified, read sequences are treated as colors. Colors may be encoded either as numbers                               (0=blue, 1=green, 2=orange, 3=red) or as characters A/C/G/T (A=blue, C=green, G=orange, T=red).</source>
        <translation>Когда -С указано, последовательности рассматриваются как цвета. Цвета могут быть закодированы либо как номера (0 = синий, 1 = зеленый, 2 = оранжевый, 3 = красный) или символов A / C / G / T (A = синий, С = зеленый, G = оранжевый, Т = красный).</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="253"/>
        <source>Map Reads with Bowtie</source>
        <translation>Выравнивание с помощью Bowtie</translation>
    </message>
    <message>
        <location filename="../src/bowtie/BowtieWorker.cpp" line="254"/>
        <source>Bowtie is a program for mapping short DNA sequence reads to a long reference sequence. It uses Burrows-Wheeler techniques extended with quality-aware backtracking algorithm that permits mismatches.&lt;br/&gt;&lt;br/&gt;Provide URL(s) to FASTA or FASTQ file(s) with NGS reads to the input port of the element, set up the reference sequence in the parameters. The result is saved to the specified SAM file, URL to the file is passed to the output port.</source>
        <translation>Выполняет выравнивание коротких ридов при помощи Bowtie.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BwaMemWorker</name>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="154"/>
        <source>Number of threads</source>
        <translation>Число потоков</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="155"/>
        <source>Number of threads (-t).</source>
        <translation>число потоков (-t).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="158"/>
        <source>Min seed length</source>
        <translation>Минимальная длина затравки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="159"/>
        <source>Path to indexed reference genome (-k).</source>
        <translation>Путь до индексированного референсного генома (-k).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="162"/>
        <source>Index algorithm</source>
        <translation>Алгоритм</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="163"/>
        <source>Index algorithm (-a).</source>
        <translation>Алгоритм (-а).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="166"/>
        <source>Band width</source>
        <translation>Ширина полосы</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="167"/>
        <source>Band width for banded alignment (-w).</source>
        <translation>Ширина полосы для выравнивания (-w).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="170"/>
        <source>Dropoff</source>
        <translation>Снижение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="171"/>
        <source>Off-diagonal X-dropoff (-d).</source>
        <translation>Снижение (-d).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="174"/>
        <source>Internal seed length</source>
        <translation>Внутренняя длина затравки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="175"/>
        <source>Look for internal seeds inside a seed longer than {-k} (-r).</source>
        <translation>Поиск внутренние затравки внутри затравки длиннее чем {-k} * (-r).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="178"/>
        <source>Skip seed threshold</source>
        <translation>Порог затравки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="179"/>
        <source>Skip seeds with more than INT occurrences (-c).</source>
        <translation>Пропускает затравки с более чем с указанным количеством вхождений (-c).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="182"/>
        <source>Drop chain threshold</source>
        <translation>Порог снижения цепи</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="183"/>
        <source>Drop chains shorter than FLOAT fraction of the longest overlapping chain (-D).</source>
        <translation>Снижение цепи короче FLOAT фракции самого длинного перекрытия цепи (-D).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="186"/>
        <source>Rounds of mate rescues</source>
        <translation>Округление за освобождение участника</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="187"/>
        <source>Perform at most INT rounds of mate rescues for each read (-m).</source>
        <translation>Выполняет для большинства INT округлений освобождений участников для каждого рида (-m).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="190"/>
        <source>Skip mate rescue</source>
        <translation>Пропустить освобождение участника</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="191"/>
        <source>Skip mate rescue (-S).</source>
        <translation>Пропустить освобождение участника (-S).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="194"/>
        <source>Skip pairing</source>
        <translation>Пропустить спаривание</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="195"/>
        <source>Skip pairing; mate rescue performed unless -S also in use (-P).</source>
        <translation>Пропустить спаривание; участник освобождается кроме случая когда -S также используется (-P).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="198"/>
        <source>Matching score</source>
        <translation>Веса совпадений</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="202"/>
        <source>Mismatch penalty</source>
        <translation>Штраф за несовпадение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="274"/>
        <source>Map Reads with BWA-MEM</source>
        <translation>Выравнивание с помощью BWA MEM</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="275"/>
        <source>Burrows-Wheeler Alignment (BWA) is a program for mapping short DNA sequence reads to a long reference sequence. This element runs &quot;BWA-MEM&quot; algorithm of the program. The algorithm is designed for sequence reads from 70bp to 1Mbp.&lt;br/&gt;&lt;br/&gt;Provide URL(s) to FASTA or FASTQ file(s) with NGS reads to the input port of the element, set up the reference sequence in the parameters. The result is saved to the specified SAM file, URL to the file is passed to the output port.</source>
        <translation>Выполняет выравнивание коротких ридов при помощи BWA MEM.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="199"/>
        <source>Score for a sequence match (-A).</source>
        <translation>Вес совпадения (-A).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="151"/>
        <source>BWA index folder</source>
        <translation>Папка для индекса BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="151"/>
        <source>BWA index basename</source>
        <translation>Базовое имя для индекса BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="203"/>
        <source>Penalty for a mismatch (-B).</source>
        <translation>Штраф за несовпадение (-B).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="206"/>
        <source>Gap open penalty</source>
        <translation>Штраф за открытие пробела</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="207"/>
        <source>Gap open penalty (-O).</source>
        <translation>Штраф за открытие пробела (-O).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="210"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пробел</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="211"/>
        <source>Gap extension penalty; a gap of size k cost {-O} (-E).</source>
        <translation>Штраф за расширение пробела; пробел размера k стоит {-O} (-E).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="214"/>
        <source>Penalty for clipping</source>
        <translation>Штраф за отсечение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="215"/>
        <source>Penalty for clipping (-L).</source>
        <translation>Штраф за отсечение (-L).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="218"/>
        <source>Penalty unpaired</source>
        <translation>Штраф за непарные риды</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="219"/>
        <source>Penalty for an unpaired read pair (-U).</source>
        <translation>Штраф за непарные риды (-U).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="222"/>
        <source>Score threshold</source>
        <translation>Порог оценки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaMemWorker.cpp" line="223"/>
        <source>Minimum score to output (-T).</source>
        <translation>Минимальная оценка для вывода (-T).</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::BwaWorker</name>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="129"/>
        <source>BWA index folder</source>
        <translation>Папка для индекса BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="129"/>
        <source>BWA index basename</source>
        <translation>Базовое имя для индекса BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="132"/>
        <source>Use missing prob</source>
        <translation>Использовать долю потери</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="133"/>
        <source>Use missing prob instead maximum edit distance.</source>
        <translation>Использовать долю потери вместо максимального расстояния редактирования.</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="136"/>
        <source>Max #diff</source>
        <translation>Максимальное расстояние редактирования</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="137"/>
        <source>Max #diff (-n)</source>
        <translation>Максимальное расстояние редактирования (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="140"/>
        <source>Missing prob</source>
        <translation>Доля потери</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="141"/>
        <source>Missing prob (-n)</source>
        <translation>Доля потери (-n)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="144"/>
        <source>Seed length</source>
        <translation>Длина затравки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="145"/>
        <source>Seed length (-l).</source>
        <translation>Длина затравки (-l).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="148"/>
        <source>Max gap opens</source>
        <translation>Максимальное число пробелов</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="149"/>
        <source>Max gap opens (-o).</source>
        <translation>Максимальное число пробелов (-o).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="152"/>
        <source>Index algorithm</source>
        <translation>Алгоритм</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="153"/>
        <source>Index algorithm (-a).</source>
        <translation>Алгоритм (-а).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="156"/>
        <source>Best hits</source>
        <translation>Лучшие попытки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="157"/>
        <source>Best hits (-R).</source>
        <translation>Лучшие попытки (-R).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="160"/>
        <source>Long-scaled gap penalty for long deletions</source>
        <translation>Долгосрочный штраф за длительный пробел</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="161"/>
        <source>Long-scaled gap penalty for long deletions (-L)</source>
        <translation>Долгосрочный штраф за длительный пробел (-L)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="164"/>
        <source>Non iterative mode</source>
        <translation>Неитеративный режим</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="165"/>
        <source>Non iterative mode (-N).</source>
        <translation>Неитеративный режим (-N).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="168"/>
        <location filename="../src/bwa/BwaWorker.cpp" line="169"/>
        <source>Enable long gaps</source>
        <translation>Допустимы длинные пробелы</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="172"/>
        <source>Max gap extensions</source>
        <translation>Максимальное расширение пробела</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="173"/>
        <source>Max gap extensions (-e)</source>
        <translation>Максимальное расширение пробела (-e)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="176"/>
        <source>Indel offset</source>
        <translation>Смещение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="177"/>
        <source>Indel offset (-i)</source>
        <translation>Смещение (-i)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="180"/>
        <source>Max long deletions extensions</source>
        <translation>Максимальная длина сокрашения пропуска</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="181"/>
        <source>Max long deletions extensions(-d)</source>
        <translation>Максимальная длина сокрашения пропуска (-d)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="184"/>
        <source>Max queue entries</source>
        <translation>Максимальное число записей в очереди</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="185"/>
        <source>Max queue entries (-m)</source>
        <translation>Максимальное число записей в очереди (-m)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="188"/>
        <source>Barcode length</source>
        <translation>Длина штрих-кода</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="189"/>
        <source>Barcode length (-B)</source>
        <translation>Длина штрих-кода (-B)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="192"/>
        <source>Threads</source>
        <translation>Потоки</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="193"/>
        <source>Threads (-t)</source>
        <translation>Потоки (-t)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="196"/>
        <source>Max seed differencies</source>
        <translation>Максимальное различие затравок</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="197"/>
        <source>Max seed differencies (-k)</source>
        <translation>Максимальное различие затравок (-k)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="200"/>
        <source>Mismatch penalty</source>
        <translation>Штраф за несовпадение</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="201"/>
        <source>Mismatch penalty (-M)</source>
        <translation>Штраф за несовпадение (-M)</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="204"/>
        <source>Gap open penalty</source>
        <translation>Штраф за открытие пробела</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="205"/>
        <source>Gap open penalty (-O).</source>
        <translation>Штраф за открытие пробела (-O).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="208"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пробел</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="209"/>
        <source>Gap extension penalty; a gap of size k cost (-E).</source>
        <translation>Штраф за расширение пробела; пробел размера k стоит (-E).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="212"/>
        <source>Quolity threshold</source>
        <translation>Порог качества</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="213"/>
        <source>Quolity threshold (-q).</source>
        <translation>Порог качества (-q).</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="281"/>
        <source>Map Reads with BWA</source>
        <translation>Выравнивание с помощью BWA</translation>
    </message>
    <message>
        <location filename="../src/bwa/BwaWorker.cpp" line="282"/>
        <source>Burrows-Wheeler Alignment (BWA) is a program for mapping short DNA sequence reads to a long reference sequence. This element runs &quot;BWA-backtrack&quot; algorithm of the program. The algorithm is designed for Illumina sequence reads up to 100bp.&lt;br/&gt;&lt;br/&gt;Provide URL(s) to FASTA or FASTQ file(s) with NGS reads to the input port of the element, set up the reference sequence in the parameters. The result is saved to the specified SAM file, URL to the file is passed to the output port.</source>
        <translation>Выполняет выравнивание коротких ридов при помощи BWA.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Prompter</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="378"/>
        <source> from %1</source>
        <translation> из %1</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CAP3Worker</name>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="89"/>
        <source>Input sequences</source>
        <translation>Входные последовательности</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="90"/>
        <source>DNA sequences that need to be assembled</source>
        <translation>ДНК последовательности для сборки</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="94"/>
        <source>Input URL(s)</source>
        <translation>Входные пути</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="95"/>
        <source>URL(s) of input file(s) in FASTA format with DNA sequences that need to be assembled</source>
        <translation>Пути входных файлов в формате FASTA</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="106"/>
        <source>Assembly Sequences with CAP3</source>
        <translation>Выравнивание с помощью CAP3</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="111"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="112"/>
        <source>Write assembly results to this output file in ACE format.</source>
        <translation>Записать выходной файл в этот выходной файл в формате ACE.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="115"/>
        <source>Quality cutoff for clipping</source>
        <translation>Размер среза</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="116"/>
        <source>Base quality cutoff for clipping (-c)</source>
        <translation>Размер среза (-c)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="119"/>
        <source>Clipping range</source>
        <translation>Диапазон обрезания</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="120"/>
        <source>One of the parameters to control clipping of a poor end region of a read (-y)&lt;p&gt;If there are quality values, CAP3 computes two positions qualpos5 and qualpos3 of read f such that the region of read f from position qualpos5 to position qualpos3 consists mostly of quality values greater than &lt;i&gt;Quality cutoff for clipping&lt;/i&gt;.&lt;/p&gt; &lt;p&gt;If there are no quality values, then qualpos5 is set to 1 and qualpos3 is set the length of read f. &lt;p&gt;The range for the left clipping position of read f is from 1 to qualpos5 + &lt;i&gt;Clipping range&lt;/i&gt;. &lt;br/&gt;The range for the right clipping position of read f is from qualpos3 - &lt;i&gt;Clipping range&lt;/i&gt; to the end of read f.&lt;/p&gt;</source>
        <translation>Один из параметров для управления обрезкой пустых концевых областей рида (-у) &lt;р&gt; Если есть значения качества, CAP3 вычисляет две позиции qualpos5 и qualpos3 рида f, такую, что регион f из положения qualpos5 в положение qualpos3 состоит в основном из значений качества больше чем &lt;i&gt; Размер среза &lt;/i&gt;. &lt;/p&gt; &lt;p&gt;Если нет значений качества, то qualpos5 установлен в 1 и qualpos3 установливает длину рида f.&lt;p&gt; Диапазон левой позиции f составляет от 1 до qualpos5 + &lt;i&gt; Диапазон среза &lt;/i&gt;. &lt;br/&gt;Диапазон правой позиции f от qualpos3 - &lt;i&gt; Диапазон среза &lt;/i&gt; до конца рида f .&lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="133"/>
        <source>Quality cutoff for differences</source>
        <translation>Срез для различий</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="134"/>
        <source>Base quality cutoff for differences (-b)</source>
        <translation>Базовый срез для различий (-b)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="137"/>
        <source>Maximum difference score</source>
        <translation>Максимальное оценка различия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="138"/>
        <source>Max qscore sum at differences (-d)&lt;p&gt;If an overlap contains lots of differences at bases of high quality, then the overlap is removed.&lt;/p&gt;&lt;p&gt;The difference score is calculated as follows. If the overlap contains a difference at bases of quality values q1 and q2, then the score at the difference is max(0, min(q1, q2) - b), where b is &lt;i&gt;Quality cutoff for differences&lt;/i&gt;. The difference score of an overlap is the sum of scores at each difference.&lt;/p&gt;</source>
        <translation>Максимальная сумма оценок для различий (-d)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="149"/>
        <source>Match score factor</source>
        <translation>Фактор оценки совпадений</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="150"/>
        <source>Match score factor (-m) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Фактор оценки совпадений (-m) - один из параметров, который влияет на оценку схожести перекрытия.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="155"/>
        <source>Mismatch score factor</source>
        <translation>Фактор оценки несовпадения</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="156"/>
        <source>Mismatch score factor (-n) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Фактор оценки несовпадения (-n) - один из параметров, который влияет на оценку схожести перекрытия.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="161"/>
        <source>Gap penalty factor</source>
        <translation>Фактор штрафа за пропуск</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="162"/>
        <source>Gap penalty factor (-g) is one of the parameters that affects similarity score of an overlap. See &lt;i&gt;Overlap similarity score cutoff&lt;/i&gt; description for details.</source>
        <translation>Фактор штрафа за пробел (-g) - один из параметров, который влияет на оценку схожести перекрытия.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="167"/>
        <source>Overlap similarity score cutoff</source>
        <translation>Срез оценки схожести перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="168"/>
        <source>If the similarity score of an overlap is less than the overlap similarity score cutoff (-s), then the overlap is removed.&lt;p&gt;The similarity score of an overlapping alignment is defined using base quality values as follows.&lt;p&gt;A match at bases of quality values q1 and q2 is given a score of m * min(q1,q2), where m is &lt;i&gt;Match score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A mismatch at bases of quality values q1 and q2 is given a score of n * min(q1,q2), where n is &lt;i&gt;Mismatch score factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;A base of quality value q1 in a gap is given a score of -g * min(q1,q2), where q2 is the quality value of the base in the other sequence right before the gap and g is &lt;i&gt;Gap penalty factor&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;The score of a gap is the sum of scores of each base in the gap minus a gap open penalty.&lt;/p&gt;&lt;p&gt;The similarity score of an overlapping alignment is the sum of scores of each match, each mismatch, and each gap. &lt;/p&gt;</source>
        <translation>Если оценка схожести перекрытия меньше чем срез оценки схожести перекрытия (-s), то перекрытие удаляется.&lt;p&gt;Оценка схожести перекрытия выравнивания определяется используя базовые значения качества, например, как приведены ниже.&lt;p&gt;Совпадения с базовым значением q1 и q2 даны как оценка m * min(q1,q2), где m это &lt;i&gt;Фактор оценки совпадений&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;Несовпадения с базовым значением q1 и q2 даны как оценка n * min(q1,q2), где n это &lt;i&gt;Фактор оценки несовпадения&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;Базовое значение качества q1 в пропуске дано как оценка -g * min(q1,q2), где q2 это значение качества в другой последовательности прямо перед пропуском и g это &lt;i&gt;Фактор штрафа за пропуск&lt;/i&gt;.&lt;/p&gt;&lt;p&gt;Оценка за пропуск это сумма всех оценок в пропуске минус штраф за открытие пропуска.&lt;/p&gt;&lt;p&gt;Оценка схожести перекрытия это сумма оценок каждого совпадения, каждого несовпадения и каждого пропуска. &lt;/p&gt;</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="186"/>
        <source>Overlap length cutoff</source>
        <translation>Обрезная длина перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="187"/>
        <source>An overlap is taken into account only if the length of the overlap in bp is no less than the specified value (parameter -o of CAP3).</source>
        <translation>Перекрытие учитывается, только если длина перекрытия в нуклеотидах не менее чем указанное значение (параметр -o).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="191"/>
        <source>Overlap percent identity cutoff</source>
        <translation>Обрезной процент идентичности перекрытия</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="192"/>
        <source>An overlap is taken into account only if the percent identity of the overlap is no less than the specified value (parameter -p of CAP3).</source>
        <translation>Перекрытие учитывается, только если длина перекрытия в процентах не менее чем указанное значение (параметр -o).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="196"/>
        <source>Max number of word matches</source>
        <translation>Максимальное число совпадений</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="197"/>
        <source>This parameter allows one to trade off the efficiency of the program for its accuracy (parameter -t of CAP3).&lt;p&gt;For a read f, CAP3 computes overlaps between read f and other reads by considering short word matches between read f and other reads. A word match is examined to see if it can be extended into a long overlap. If read f has overlaps with many other reads, then read f has many short word matches with many other reads.&lt;/p&gt;&lt;p&gt;This parameter gives an upper limit, for any word, on the number of word matches between read f and other reads that are considered by CAP3.&lt;/p&gt;&lt;p&gt;Using a large value for this parameter allows CAP3 to consider more word matches between read f and other reads, which can find more overlaps for read f, but slows down the program.&lt;/p&gt;&lt;p&gt;Using a small value for this parameter has the opposite effect.&lt;/p&gt;</source>
        <translation>Максимальное число совпадений (параметр -t)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="212"/>
        <source>Band expansion size</source>
        <translation>Размер расширения диапазона</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="213"/>
        <source>CAP3 determines a minimum band of diagonals for an overlapping alignment between two sequence reads. The band is expanded by a number of bases specified by this value (parameter -a of CAP3).</source>
        <translation>CAP3 определяет минимальное группу диагоналей для перекрытия выравнивания между двумя последовательностями ридов. Группа расширяется числом оснований, указанных на это значение (Параметр -a САР 3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="218"/>
        <source>Max gap length in an overlap</source>
        <translation>Максимальная длина пропуска в перекрытии</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="219"/>
        <source>The maximum length of gaps allowed in any overlap (-f)&lt;p&gt;I.e. overlaps with longer gaps are rejected.&lt;/p&gt;&lt;p&gt;Note that a small value for this parameter may cause the program to remove true overlaps and to produce incorrect results.&lt;/p&gt;&lt;p&gt;The parameter may be used to split reads from alternative splicing forms into separate contigs.&lt;/p&gt;</source>
        <translation>Максимальная длина пробела допустимая в перекрытии (-f)</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="227"/>
        <source>Assembly reverse reads</source>
        <translation>Сборка обратных ридов</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="228"/>
        <source>Specifies whether to consider reads in reverse orientation for assembly (originally, parameter -r of CAP3).</source>
        <translation>Указывает, следует ли рассматривать риды обратной ориентации для сборки (первоначально параметр -r САР 3).</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="232"/>
        <source>CAP3 tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="233"/>
        <source>The path to the CAP3 external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента CAP3 в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="236"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="237"/>
        <source>The folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3Worker.cpp" line="379"/>
        <source>Aligns long DNA reads from &lt;u&gt;%1&lt;/u&gt; with CAP3.</source>
        <translation>Выполняет выравнивание длинных ридов из &lt;u&gt;%1&lt;/u&gt; при помощи CAP3.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportPrompter</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="370"/>
        <source>Creates summary statistics on ChIP enrichment and saves it to %1.</source>
        <translation>Создает сводную статистику по обогащению ChIP и сохраняет в%1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CEASReportWorker</name>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="214"/>
        <source>Peak regions</source>
        <translation>Пиковые регионы</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="215"/>
        <source>Peak regions.</source>
        <translation>Пиковые регионы.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="217"/>
        <source>Enrichment signal</source>
        <translation>Сигнал обогащения</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="218"/>
        <source>Enrichment signal.</source>
        <translation>Сигнал обогащения.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="223"/>
        <source>CEAS data</source>
        <translation>Данные CEAS</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="224"/>
        <source>Peak regions and enrichment signal profile for generating CEAS report.</source>
        <translation>Пиковые регионы и профиль сигнала обогщения для генерации отчета CEAS.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="233"/>
        <source>Output report file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="234"/>
        <source>Path to the report output file. Result for CEAS analysis.</source>
        <translation>Путь до выходного файла. Результат анализа CEAS.</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="239"/>
        <source>Output annotations file</source>
        <translation>Выходной файл с аннотациями</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="240"/>
        <source>Name of tab-delimited output text file, containing a row of annotations for every RefSeq gene. (file is not generated if no peak location data are supplied).</source>
        <translation>Название для выходного текстового файла, содержащего строки аннотации для каждого гена RefSeq. (файл не генерируется, если нет пиков).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="243"/>
        <source>Gene annotations table</source>
        <translation>Таблица аннотаций генов</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="244"/>
        <source>Path to gene annotation table (e.g. a refGene table in sqlite3 db format (--gt).</source>
        <translation>Путь до таблицы аннотаций (--gt).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="247"/>
        <source>Span size</source>
        <translation>Размер промежутка</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="248"/>
        <source>Span from TSS and TTS in the gene-centered annotation (base pairs). ChIP regions within this range from TSS and TTS are considered when calculating the coverage rates in promoter and downstream (--span).</source>
        <translation>Промежуток от TSS и TTS в центральной аннотации. Регионы ChIP в этом регионе от TSS и TTS учитываются при вычислении покрытия в промоутере (--span).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="253"/>
        <source>Wiggle profiling resolution</source>
        <translation>Разрешение профилирования</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="254"/>
        <source>Wiggle profiling resolution. WARNING: Value smaller than the wig interval (resolution) may cause aliasing error (--pf-res).</source>
        <translation>Разрешение профилирования. Внимание: значения, меньшие чем разрешение могут вызвать ошибки сглаживания (--pf-res).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="258"/>
        <source>Promoter/downstream interval</source>
        <translation>Интервал промоутера</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="259"/>
        <source>Promoter/downstream intervals for ChIP region annotation are three values or a single value can be given. If a single value is given, it will be segmented into three equal fractions (e.g. 3000 is equivalent to 1000,2000,3000) (--rel-dist).</source>
        <translation>Интервалы промоутера для регионов ChIP это три значения или одиночное значение. Если задано одиночное значение, то оно будет разделено на три равные фракции (например, 3000 эквивалентно 1000,2000,3000) (--rel-dist).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="265"/>
        <source>BiPromoter ranges</source>
        <translation>Интервалы двунаправленных промоторов</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="266"/>
        <source>Bidirectional-promoter sizes for ChIP region annotation. It&apos;s two values or a single value can be given. If a single value is given, it will be segmented into two equal fractions (e.g. 5000 is equivalent to 2500,5000) (--bisizes).</source>
        <translation>Размеры двунаправленных промоторов для ChIP регионов аннотаций. Это два значения или одиночное значение. Если задано одиночное значение, то оно будет разделено на две равные фракции (например, 5000 эквивалентно 2500,5000) (--sizes).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="271"/>
        <source>Relative distance</source>
        <translation>Относительное расстояние</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="272"/>
        <source>Relative distance to TSS/TTS in WIGGLE file. profiling. (--rel-dist)</source>
        <translation>Относительное расстояние для TSS/TTS в WIGGLE файле. (--rel-dist)</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="275"/>
        <source>Gene group files</source>
        <translation>Файлы групп генов</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="276"/>
        <source>Gene groups of particular interest in wig profiling. Each gene group file must have gene names in the 1st column. The file names are separated by commas (--gn-groups).</source>
        <translation>Групп генов, представляющие особый интерес в профилировании. Каждый файл группы генов должны иметь имена генов в 1-м столбце. Имена файлов разделяются запятыми (--gn-groups).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="280"/>
        <source>Gene group names</source>
        <translation>Имена групп генов</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="281"/>
        <source>&lt;i&gt;Set this parameter empty for using default values.&lt;/i&gt;&lt;br&gt;The names of the gene groups from &quot;Gene group files&quot; parameter. These names appear in the legends of the wig profiling plots.&lt;br&gt; Values range: comma-separated list of strings. Default value: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</source>
        <translation>&lt;i&gt;Оставьте этот параметр пустым, чтобы использовать значения по умолчанию.&lt;/i&gt;&lt;br&gt;Имена групп генов из параметра &quot;Файлы групп генов&quot;. Эти имена появятся в легенде графика профилирования.&lt;br&gt; Интервал значений: строки, разделенные запятыми. Значение по умолчанию: &apos;Group 1, Group 2,...Group n&apos; (--gn-group-names).</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="320"/>
        <source>Report file</source>
        <translation>Отчет</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="327"/>
        <source>Annotations file</source>
        <translation>Файл с аннотациями</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="350"/>
        <source>Create CEAS Report</source>
        <translation>Создание отчета CEAS</translation>
    </message>
    <message>
        <location filename="../src/ceas/CEASReportWorker.cpp" line="351"/>
        <source>Provides summary statistics on ChIP enrichment in important genomic regions such as individual chromosomes, promoters, gene bodies or exons, and infers the genes most likely to be regulated by the binding factor under study.</source>
        <translation>Обеспечивает сводные статистические данные по обогащению ChIP важных областей генома, таких как отдельные хромосомы, промотеры, гены или экзонов, и указывает гены, наиболее подверженные регуляции связывающим фактором.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOPrompter</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="142"/>
        <source> from %1</source>
        <translation> из %1</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="143"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</source>
        <translation>Выравнивает каждое MSA из &lt;u&gt;%1&lt;/u&gt; при помощи &quot;&lt;u&gt;ClustalO&lt;/u&gt;&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalOWorker</name>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="65"/>
        <source>Input MSA</source>
        <translation>Входное MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="65"/>
        <source>Input MSA to process.</source>
        <translation>Входное выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="66"/>
        <source>ClustalO result MSA</source>
        <translation>Выходное MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="66"/>
        <source>The result of the ClustalO alignment.</source>
        <translation>Выходное выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="75"/>
        <source>Number of iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="75"/>
        <source>Number of (combined guide-tree/HMM) iterations.</source>
        <translation>Число итераций.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="76"/>
        <source>Number of guidetree iterations</source>
        <translation>Число итераций направляющих деревьев</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="76"/>
        <source>Maximum number guidetree iterations.</source>
        <translation>Максимальное число итераций направляющих деревьев.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="77"/>
        <source>Number of HMM iterations</source>
        <translation>Число итераций HMM</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="77"/>
        <source>Maximum number of HMM iterations.</source>
        <translation>Максимальное число итераций HMM.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="78"/>
        <source>Set auto options</source>
        <translation>Установить параметры автоматически</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="78"/>
        <source>Set options automatically (might overwrite some of your options).</source>
        <translation>Установить параметры автоматически.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="80"/>
        <source>Tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="80"/>
        <source>Path to the ClustalO tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation>Путь до ClustalO.&lt;p&gt;Значение по умолчанию может быть указано в диалоге настроек UGENE.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="83"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="83"/>
        <source>Folder to store temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="93"/>
        <source>Align with ClustalO</source>
        <translation>Выравнивание с помощью ClustalO</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="93"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalO.&lt;p&gt;ClustalO is a general purpose multiple sequence alignment program for proteins.Visit &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; to learn more about it.</source>
        <translation>Выравнивает множественные выравнивания (MSA) при помощи ClustalO.&lt;p&gt;ClustalO это программа для выравнивания белков. Посмотрите &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/omega&lt;/a&gt; для большей информации.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="188"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to ClustalO.</translation>
    </message>
    <message>
        <location filename="../src/clustalo/ClustalOWorker.cpp" line="220"/>
        <source>Aligned %1 with ClustalO</source>
        <translation>Aligned %1 with ClustalO</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWPrompter</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="190"/>
        <source> from %1</source>
        <translation> из %1</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="191"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</source>
        <translation>Выравнивает каждое MSA из &lt;u&gt;%1&lt;/u&gt; при помощи &quot;&lt;u&gt;ClustalW&lt;/u&gt;&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ClustalWWorker</name>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="69"/>
        <source>Input MSA</source>
        <translation>Входное MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="69"/>
        <source>Input MSA to process.</source>
        <translation>Входное выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="70"/>
        <source>ClustalW result MSA</source>
        <translation>Выходное MSA</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="70"/>
        <source>The result of the ClustalW alignment.</source>
        <translation>Выходное выравнивание.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="79"/>
        <source>Gap open penalty</source>
        <translation>Штраф за открывающий пропуск</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="79"/>
        <source>The penalty for opening a gap.</source>
        <translation>Штраф за открытие пробела.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="80"/>
        <source>Gap extension penalty</source>
        <translation>Штраф за промежуточный пробел</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="80"/>
        <source>The penalty for extending a gap.</source>
        <translation>Штраф за расширение пробел.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="81"/>
        <source>Gap distance</source>
        <translation>Расстояние пробела</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="81"/>
        <source>The gap separation penalty. Tries to decrease the chances of gaps being too close to each other.</source>
        <translation>Штраф за разделение пробела. Пытается уменьшить изменение пробелов располагая их близко друг к другу.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="82"/>
        <source>End gaps</source>
        <translation>Убрать пробел</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="82"/>
        <source>The penalty for closing a gap.</source>
        <translation>Штраф за закрытие пробела.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="83"/>
        <source>Residue-specific gaps off</source>
        <translation>Откл. спецефические для остатков пробелы</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="83"/>
        <source>Residue-specific penalties are amino specific gap penalties that reduce or increase the gap opening penalties at each position in the alignment.</source>
        <translation>Отключение спецефические для остатков пробелы - штрафы за специфические белковые остатки которые увеличивают или уменьшают штрафы за открытие пробела в каждой позиции.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="84"/>
        <source>Hydrophilic gaps off</source>
        <translation>Отключить гидрофильные пробелы</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="84"/>
        <source>Hydrophilic gap penalties are used to increase the chances of a gap within a run (5 or more residues) of hydrophilic amino acids.</source>
        <translation>Штрафы за гидрофильные пробелы используются для увеличения изменения пробела гидрофильных амино кислот.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="85"/>
        <source>Iteration type</source>
        <translation>Тип итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="85"/>
        <source>Alignment improvement iteration type. Can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;None - No iteration;&lt;/li&gt;                            &lt;li&gt;Tree - Iteration at each step of alignment process;&lt;/li&gt;                            &lt;li&gt;Alignment - Iteration only on final alignment.&lt;/li&gt;                            &lt;/ul&gt;</source>
        <translation>Тип итераций. Возможные значения: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;Нет - нет итерации;&lt;/li&gt;                            &lt;li&gt;Дерево - итерация на каждом шаге выравнивания;&lt;/li&gt;                            &lt;li&gt;Выравнивание - итерация финального выравнивания.&lt;/li&gt;                            &lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="91"/>
        <source>Number of iterations</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="91"/>
        <source>The maximum number of iterations to perform.</source>
        <translation>Максимальное число итераций.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="92"/>
        <source>Weight matrix</source>
        <translation>Весовая матрица</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="105"/>
        <source>Tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="105"/>
        <source>Path to the ClustalW tool.&lt;p&gt;The default path can be set in the UGENE application settings.</source>
        <translation>Путь до ClustalW.&lt;p&gt;Значение по умолчанию может быть указано в диалоге настроек UGENE.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="108"/>
        <source>Folder to store temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="122"/>
        <source>Aligns multiple sequence alignments (MSAs) supplied with ClustalW.&lt;p&gt;ClustalW is a general purpose multiple sequence alignment program for DNA or proteins.Visit &lt;a href=&quot;http://www.clustal.org/&quot;&gt;http://www.clustal.org/&lt;/a&gt; to learn more about it.</source>
        <translation>Выравнивает множественные выравнивания (MSA) при помощи ClustalO.&lt;p&gt;ClustalO это программа для выравнивания белков или ДНК последовательностей. Посмотрите &lt;a href=&quot;http://www.clustal.org/omega&quot;&gt;http://www.clustal.org/&lt;/a&gt; для большей информации.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="263"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to ClustalW.</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="108"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="92"/>
        <source>For proteins it is a scoring table which describes the similarity of each amino acid to each other and can take values: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - Appear to be the best available for carrying out database similarity (homology searches).                            &lt;li&gt;PAM - Have been extremely widely used since the late &apos;70s.                            &lt;li&gt;Gonnet - Derived using almost the same procedure as the PAM bit are much more up to date.                            &lt;li&gt;ID - Identitical matrix gives a score of 1.0 to two identical amino acids and a score of zero otherwise.                            &lt;/ul&gt;                            &lt;p&gt;                            For DNA it is the scores assigned to matches and mismatches. The values available are:                            &lt;ul&gt;                            &lt;li&gt;IUB - The default scoring matrix used by BESTFIT for the comparison of nucleic acid sequences. All matches score 1.9; all mismatches for IUB symbols score 0.                           &lt;li&gt;ClustalW - The previous system used by Clustal W, in which matches score 1.0 and mismatches score 0. All matches for IUB symbols also score 0.                            &lt;/ul&gt;</source>
        <translation>Для белков это таблица подсчета оценок, которая описывает сходство каждой аминокислоты друг к другу, и может принимать значения: &lt;p&gt;                            &lt;ul&gt;                            &lt;li&gt;BLOSUM - для сравнения баз данных (поиск гомологий).                            &lt;li&gt;PAM - широко используется с конца 70х.                            &lt;li&gt;Gonnet - похож на PAM, начал использоваться немного позже.                            &lt;li&gt;ID - идентификационная матрица дает оценку равную 1 для двух одинаковых амино кислот и нулю в остальных случаях.                            &lt;/ul&gt;                            &lt;p&gt;                            Для ДНК это оценки для совпадений и несовпадений. Доступные значения:                            &lt;ul&gt;                            &lt;li&gt;IUB - дефолтная матрица, используемая BESTFIT для сравнения нуклеотидных последовательностей. Все оценки совпадений 1.9; все оценки несовпадений для IUB равны 0.                           &lt;li&gt;ClustalW - прошлая система используемая  Clustal W, в которой оценка несовпадений равна 1.0, а несопадений 0. Все совпадения для IUB тоже равны 0.                            &lt;/ul&gt;</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="122"/>
        <source>Align with ClustalW</source>
        <translation>Выравнивание с помощью ClustalW</translation>
    </message>
    <message>
        <location filename="../src/clustalw/ClustalWWorker.cpp" line="292"/>
        <source>Aligned %1 with ClustalW</source>
        <translation>Aligned %1 with ClustalW</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOPrompter</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="208"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="213"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as gene list to conduct GO.</source>
        <translation>Использует аннотации из &lt;u&gt;%1&lt;/u&gt; в качестве списка генов для conduct GO.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="215"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; folder</source>
        <translation>Записывает все выходные файлы в &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConductGOWorker</name>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="136"/>
        <source>Target genes</source>
        <translation>Контрольные гены</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="137"/>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="142"/>
        <source>URL to file with genes to identify over represented GO terms.</source>
        <translation>URL to file with genes to identify over represented GO terms.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="141"/>
        <source>Conduct GO data</source>
        <translation>Данные Conduct GO</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="150"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="151"/>
        <source>The folder to store Conduct GO results.</source>
        <translation>Папка для сохранения выходных Conduct GO файлов.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="153"/>
        <source>Title</source>
        <translation>Имя</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="154"/>
        <source>Title is used to name the output files - so make it meaningful.</source>
        <translation>имя для выходных файлов.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="156"/>
        <source>Gene Universe</source>
        <translation>Множество генов</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="157"/>
        <source>Select a gene universe.</source>
        <translation>выберите множество генов.</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="188"/>
        <source>Conduct GO</source>
        <translation>Conduct GO</translation>
    </message>
    <message>
        <location filename="../src/conduct_go/ConductGOWorker.cpp" line="189"/>
        <source>Given a list of genes, using Bioconductor (GO, GOstats) and  DAVID at NIH.</source>
        <translation>этот элемент выдает список генов, используя Bioconductor (GO, GOstats) и  DAVID в NIH.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotPrompter</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="278"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="284"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions for conservation plot.</source>
        <translation>Использует аннотации из &lt;u&gt;%1&lt;/u&gt; в качестве пиков для консервативного графика.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="285"/>
        <source> Conservations scores from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation>Консервативные оценки из &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="286"/>
        <source> Outputs the result to &lt;u&gt;%1&lt;/u&gt;</source>
        <translation>Записывает все выходные файлы в &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ConservationPlotWorker</name>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="179"/>
        <source>Input regions</source>
        <translation>Входные регионы</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="180"/>
        <source>Regions (centered at peak summits for better performance).</source>
        <translation>Регионы.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="184"/>
        <source>conservation_plot data</source>
        <translation>данные conservation_plot</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="185"/>
        <source>Regions to plot the PhastCons scores profiles.</source>
        <translation>Регионы для построения графика оценок PhastCons.</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="193"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="194"/>
        <source>File to store phastcons results (BMP).</source>
        <translation>Файл для сохранения результатов (BMP).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="196"/>
        <source>Title</source>
        <translation>Название</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="197"/>
        <source>Title of the figure (--title).</source>
        <translation>Название графика (--title).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="199"/>
        <source>Label</source>
        <translation>Ярлык</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="200"/>
        <source>Label of data in the figure (--bed-label).</source>
        <translation>Ярлык данных на графике (--bed-label).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="202"/>
        <source>Assembly version</source>
        <translation>Версия сборки</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="203"/>
        <source>The folder to store phastcons scores (--phasdb).</source>
        <translation>Папка для сохранения оценок (--phasdb).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="205"/>
        <source>Window width</source>
        <translation>Ширина окна</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="206"/>
        <source>Window width centered at middle of regions (-w).</source>
        <translation>Ширина окна (-w).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="208"/>
        <source>Height</source>
        <translation>Высота</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="209"/>
        <source>Height of plot (--height).</source>
        <translation>Высота графика (--height).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="211"/>
        <source>Width</source>
        <translation>Ширина</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="212"/>
        <source>Width of plot (--width).</source>
        <translation>Ширина графика (--width).</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="236"/>
        <source>Conservation plot file</source>
        <translation>Файл консервативного графика</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="258"/>
        <source>Build Conservation Plot</source>
        <translation>Построение графика уровня консервативности последовательности</translation>
    </message>
    <message>
        <location filename="../src/conservation_plot/ConservationPlotWorker.cpp" line="259"/>
        <source>Plots the PhastCons scores profiles.</source>
        <translation>Строит график оценок PhastCons.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CropStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/CropStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;CROP&lt;/h4&gt;&lt;p&gt;This step removes bases regardless of quality from the end of the read, so that the read has maximally the specified length after this step has been performed. Steps performed after CROP might of course further shorten the read.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Length&lt;/b&gt;: the number of bases to keep, from the start of the read.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/CropStep.cpp" line="49"/>
        <source>The number of bases to keep, from the start of the read.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffdiffWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="93"/>
        <source>Test for Diff. Expression with Cuffdiff</source>
        <translation>Поиск транскриптов с дифференциальной экспрессией с помощью Cuffdiff</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="94"/>
        <source>Cuffdiff takes a transcript file as input, along with two or more fragment alignments (e.g. in SAM format) for two or more samples. It produces a number of output files that contain test results for changes in expression at the level of transcripts, primary transcripts, and genes. It also tracks changes in the relative abundance of transcripts sharing a common transcription start site, and in the relative abundances of the primary transcripts of each gene. Tracking the former allows one to see changes in splicing, and the latter lets one see changes in relative promoter use within a gene.</source>
        <translation>Cuffdiff принимает на вход файл с транскриптами и два или более файла с выравниванием фрагментов (например, в формате SAM) для разных образцов. На выходе Cuffdiff предоставляет ряд файлов, содержащих результаты анализа изменений уровня экспрессии транскриптов, основных транскриптов и генов. Кроме того, Cuffdiff отмечает изменение относительного уровня транскриптов с общей точкой начала транскрипции и транскриптов одного гена. В первом случае это позволяет зафиксировать изменения сплайсинга, а вот втором — использование разных участков промоторной области гена.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="107"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="108"/>
        <source>The base name of output folder. It could be modified with a suffix.</source>
        <translation>Базовое имя для выходной папки.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="111"/>
        <source>Time series analysis</source>
        <translation>Анализ временных рядов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="112"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, instructs Cuffdiff to analyze the provided samples as a time series, rather than testing for differences between all pairs of samples. Samples should be provided in increasing time order.</source>
        <translation>При выборе этой опции Cuffdiff будет анализировать предоставленные образцы как временные ряды, а не проверки различий между всеми парами образцов. Образцы должны быть представлены в порядке возрастания времени.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="118"/>
        <source>Upper quartile norm</source>
        <translation>Квартильная нормализация</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="119"/>
        <source>If set to &lt;i&gt;True&lt;/i&gt;, normalizes by the upper quartile of the number of fragments mapping to individual loci instead of the total number of sequenced fragments. This can improve robustness of differential expression calls for less abundant genes and transcripts.</source>
        <translation>если выбран этот параметр, будет применена нормализация по верхнему квартилю числа фрагментов, выровненных на каждый локус, вместо общего числа отсеквенированных фрагментов. Это может увеличить надежность расчёта дифференциальной экспрессии для генов и транскриптов с невысокой экспрессией.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="125"/>
        <source>Hits norm</source>
        <translation>Нормализация по числу выровненных фрагментов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="126"/>
        <source>Instructs how to count all fragments. &lt;i&gt;Total&lt;/i&gt; specifies to count all fragments, including those not compatible with any reference transcript, towards the number of mapped fragments used in the FPKM denominator. &lt;i&gt;Compatible&lt;/i&gt; specifies to use only compatible fragments. Selecting &lt;i&gt;Compatible&lt;/i&gt; is generally recommended in Cuffdiff to reduce certain types of bias caused by differential amounts of ribosomal reads which can create the impression of falsely differentially expressed genes.</source>
        <translation>указывает, как считать все фрагменты. “Все”: рассчитывать отношение числах всех фрагментов, в том числе тех, которые не совместимы ни с одним референсным транскриптов, к числу выровненных фрагментов, в знаменателе формулы FPKM. “Совместимые”: использовать только совместимые фрагменты. Рекомендуется выбирать опцию “совместимые”, чтобы уменьшить определенные типы смещения, вызванные вариацией в количестве рибосомных ридов, которые могут создать впечатление ложной дифференциальной экспрессии генов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="135"/>
        <source>Frag bias correct</source>
        <translation>Корректировка смещения</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="136"/>
        <source>Providing the sequences your reads were mapped to instructs Cuffdiff to run bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation>при указании файла с  последовательностями в формате fasta Cufflinks с несколькими файлами запустит алгоритм обнаружения и коррекции смещения, который может значительно улучшить точность оценки обилия транскриптов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="141"/>
        <source>Multi read correct</source>
        <translation>Корректировка множественно выровненных ридов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="142"/>
        <source>Do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation>указывает Cufflinks применить процедуру первичной оценки для более качественного присвоения веса ридам, выровенным на разные участки генома.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="146"/>
        <source>Library type</source>
        <translation>Тип библиотеки</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="147"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>Задает протокол RNA-Seq.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="150"/>
        <source>Mask file</source>
        <translation>Файл маски</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="151"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts or other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation>игнорирует все риды, которые могли произойти от транскриптов в этом файле. Рекомендуется включить в этот файл любые аннотированные последовательности рРНК, транскрипты митохондриального происхождения и другие транскрипты с высокой экспрессией, которые вы хотите игнорировать в своем анализе. В связи с различной эффективностью методов обогащения мРНК и уменьшения количества рРНК такая маскировка обычно улучшает общее качество оценки обильности транскриптов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="159"/>
        <source>Min alignment count</source>
        <translation>Минимальное количество выравниваний</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="160"/>
        <source>The minimum number of alignments in a locus for needed to conduct significance testing on changes in that locus observed between samples. If no testing is performed, changes in the locus are deemed not significant, and the locus&apos; observed changes don&apos;t contribute to correction for multiple testing.</source>
        <translation>минимальное число выравниваний для локуса для тестирования значимости изменений в этом локусе, наблюдаемом между образцами. Если тестирование не проводится, изменения не считаются существенными и не влияют на процедуру поправки на множественные сравнения.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="167"/>
        <source>FDR</source>
        <translation>FDR</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="168"/>
        <source>The allowed false discovery rate used in testing.</source>
        <translation>верхний порог ложноположительных результатов (FDR) для анализа.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="171"/>
        <source>Max MLE iterations</source>
        <translation>Количество итераций MLE</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="172"/>
        <source>Sets the number of iterations allowed during maximum likelihood estimation of abundances.</source>
        <translation>устанавливает количество итераций, допустимое при оценки обилия транскрипта методом максимального правдоподобия.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="176"/>
        <source>Emit count tables</source>
        <translation>Выдавать таблицы расчетов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="177"/>
        <source>Include information about the fragment counts, fragment count variances, and fitted variance model into the report.</source>
        <translation>включить в отчет информацию  о количестве фрагментов, разбросе этой величины и подобранные параметры дисперсионного анализа.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="181"/>
        <source>Cuffdiff tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="182"/>
        <source>The path to the Cuffdiff external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента Cuffdiff в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="185"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="186"/>
        <source>The folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="206"/>
        <source>Assembly</source>
        <translation>Сборка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="207"/>
        <source>RNA-Seq reads assemblies</source>
        <translation>Сборки ридов RNA-Seq</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="209"/>
        <source>Annotations</source>
        <translation>Аннотации</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="210"/>
        <source>Transcript annotations</source>
        <translation>Транскрипционные аннотации</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="212"/>
        <source>Sample</source>
        <translation>Пример</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="213"/>
        <source>Sample name of assembly file</source>
        <translation>Пример имени для файла сборки</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="238"/>
        <source>Total</source>
        <translation>Всего</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="239"/>
        <source>Compatible</source>
        <translation>Совместимо</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="295"/>
        <source>Searches for significant changes in transcript expression, splicing and promoter use.</source>
        <translation>Поиск существенных изменений в выражении транскрипции.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CufflinksWorker</name>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="98"/>
        <source>Input reads</source>
        <translation>Входные риды</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="99"/>
        <source>Input RNA-Seq read alignments.</source>
        <translation>Входные риды RNA-Seq.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="102"/>
        <source>Output annotations</source>
        <translation>Выходные аннотации</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="103"/>
        <source>Assembled isoforms, estimated isoform-level expression values, and estimated gene-level expression values, produced by Cufflinks.</source>
        <translation>Собранные изоформы, оцененные значения экспрессии изоформ и оценочные значения экспрессии генов производенные Cufflinks.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="116"/>
        <source>Isoform-level expression values</source>
        <translation>Значения экспрессии изоформ</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="117"/>
        <source>A set of annotated regions</source>
        <translation>Набор аннотированных регионов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="134"/>
        <source>Assembly Transcripts with Cufflinks</source>
        <translation>Сборка транскриптов с помощью Cufflinks</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="135"/>
        <source>Cufflinks accepts aligned RNA-Seq reads and assembles the alignments into a parsimonious set of transcripts. Cufflinks then estimates the relative abundances of these transcripts based on how many reads support each one, taking into account biases in library preparation protocols.</source>
        <translation>Cufflinks принимает выровненные риды RNA-Seq и собирает выравнивание в набор транскрипций. После этого Cufflinks оценивает относительное содержание этих транскриптов на основе того, сколько ридов поддерживает каждый из них, с учетом возможных систематических ошибок разных протоколов подготовки библиотек.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="143"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="144"/>
        <source>The base name of output folder. It could be modified with a suffix.</source>
        <translation>базовое имя для выходной папки.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="147"/>
        <source>Reference annotation</source>
        <translation>Референсная аннотация</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="148"/>
        <source>Tells Cufflinks to use the supplied reference annotation to estimate isoform expression. Cufflinks will not assemble novel transcripts and the program will ignore alignments not structurally compatible with any reference transcript.</source>
        <translation>указывает Cufflinks использовать прилагаемую референсную аннотацию для оценки экспрессии изоформ. Cufflinks не будет собирать новые транскрипты и программа будет игнорировать выравнивания, по структуре не совместимые с каким-либо референсным транскриптом.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="154"/>
        <source>RABT annotation</source>
        <translation>Аннотация RABT</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="155"/>
        <source>Tells Cufflinks to use the supplied reference annotation to guide Reference Annotation Based Transcript (RABT) assembly. Reference transcripts will be tiled with faux-reads to provide additional information in assembly. Output will include all reference transcripts as well as any novel genes and isoforms that are assembled.</source>
        <translation>указывает Cufflinks использовать прилагаемую референсную аннотацию для сборки транскриптома по референсу (Reference Annotation Based Transcript (RABT)). Референсные транскрипции наряду со сгенерированными ридами будут использованы как дополнительная информация в сборке. В итоговый результат войдут как референсные транскрипты, так и гены и изоформы из новой сборки.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="162"/>
        <source>Library type</source>
        <translation>Тип библиотеки</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="163"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>задает протокол секвенирования РНК.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="166"/>
        <source>Mask file</source>
        <translation>Файл маски</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="167"/>
        <source>Ignore all reads that could have come from transcripts in this file. It is recommended to include any annotated rRNA, mitochondrial transcripts or other abundant transcripts you wish to ignore in your analysis in this file. Due to variable efficiency of mRNA enrichment methods and rRNA depletion kits, masking these transcripts often improves the overall robustness of transcript abundance estimates.</source>
        <translation>игнорирует все риды, которые могли произойти от транскрипций в этом файле.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="175"/>
        <source>Multi-read correct</source>
        <translation>Регулировка множественных ридов</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="176"/>
        <source>Tells Cufflinks to do an initial estimation procedure to more accurately weight reads mapping to multiple locations in the genome.</source>
        <translation>начальная процедура оценки для более точного отображения веса ридов в нескольких местах в геноме.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="180"/>
        <source>Min isoform fraction</source>
        <translation>Минимальное содержание изоформ</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="181"/>
        <source>After calculating isoform abundance for a gene, Cufflinks filters out transcripts that it believes are very low abundance, because isoforms expressed at extremely low levels often cannot reliably be assembled, and may even be artifacts of incompletely spliced precursors of processed transcripts. This parameter is also used to filter out introns that have far fewer spliced alignments supporting them.</source>
        <translation>после расчета уровня изоформ конкретного гена Cufflinks удалит транскрипты с очень маленькой обильностью, потому что изоформы с таким низким уровнем экспрессии часто не могут быть надежно собраны. Этот параметр также используется для удаления из анализа предполагаемых интронов в том случае, если их существование поддерживает очень мало сплайсированных выравниваний.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="189"/>
        <source>Frag bias correct</source>
        <translation>Корректировка смещения</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="190"/>
        <source>Providing Cufflinks with a multifasta file via this option instructs it to run the bias detection and correction algorithm which can significantly improve accuracy of transcript abundance estimates.</source>
        <translation>при указании файла с последовательностями в формате fasta Cufflinks с несколькими файлами запустит алгоритм обнаружения и коррекции смещения, который может значительно улучшить точность оценки обилия транскриптов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="195"/>
        <source>Pre-mRNA fraction</source>
        <translation>Содержание пре-мРНК</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="196"/>
        <source>Some RNA-Seq protocols produce a significant amount of reads that originate from incompletely spliced transcripts, and these reads can confound the assembly of fully spliced mRNAs. Cufflinks uses this parameter to filter out alignments that lie within the intronic intervals implied by the spliced alignments. The minimum depth of coverage in the intronic region covered by the alignment is divided by the number of spliced reads, and if the result is lower than this parameter value, the intronic alignments are ignored.</source>
        <translation>некоторые протоколы секвенирования РНК производят значительное количество ридов, которые происходят из не полностью сплайсированных транскриптов, и эти риды могут затруднить сборку послностью сплайсированных последовательностей мРНК. Cufflinks использует этот параметр, чтобы отфильтровать выравнивания внутри интронов, границы которых определяются по результатам выравнивания сплайсированных последовательностей. Минимальное покрытие в области интронов, присутствующее в выравнивании, делится на число сплайсированных ридов, и если частное от деления меньше значения параметра, выравнивания в области интронов будут исключены из анализа.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="206"/>
        <source>Cufflinks tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="207"/>
        <source>The path to the Cufflinks external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента Cufflinks в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="210"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="211"/>
        <source>The folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="333"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation>Incorrect value of the library type parameter for Cufflinks!</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CuffmergeWorker</name>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="62"/>
        <source>Merge Assemblies with Cuffmerge</source>
        <translation>Соединение сборок с помощью Cuffmerge</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="63"/>
        <source>Cuffmerge merges together several assemblies. It also handles running Cuffcompare for you, and automatically filters a number of transfrags that are probably artifacts. If you have a reference file available, you can provide it to Cuffmerge in order to gracefully merge input (e.g. novel) isoforms and known isoforms and maximize overall assembly quality.</source>
        <translation>Cuffmerge соединяет вместе несколько сборок, а также обеспечивает запуска Cuffcompare и автоматически отбрасывает ряд объединенных фрагментов, наиболее вероятно являющихся артефактами. При наличии файла с референсом вы можете указать его, и Cuffmerge объединит известные и новые изоформы, повысив качество итоговой сборки.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="73"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="74"/>
        <source>The base name of output folder. It could be modified with a suffix.</source>
        <translation>Базовое имя для выходной папки.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="77"/>
        <source>Reference annotation</source>
        <translation>Референсная аннотация</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="78"/>
        <source>Merge the input assemblies together with this reference annotation.</source>
        <translation>объединить входные сборки с этой референсной аннотацией.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="82"/>
        <source>Reference sequence</source>
        <translation>Референсная последовательность</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="83"/>
        <source>The genomic DNA sequences for the reference. It is used to assist in classifying transfrags and excluding artifacts (e.g. repeats). For example, transcripts consisting mostly of lower-case bases are classified as repeats.</source>
        <translation>референсная последовательность ДНК. Она используется для классификации химерных фрагментов и исключения артефактов (например, повторов). Например, транскрипты, состоящие в основном из строчных символов будут признаны повторами.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="89"/>
        <source>Minimum isoform fraction</source>
        <translation>Минимальное содержание изоформы</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="90"/>
        <source>Discard isoforms with abundance below this.</source>
        <translation>не рассматривать изоформы с обилием ниже указанного уровня.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="93"/>
        <source>Cuffcompare tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="94"/>
        <source>The path to the Cuffcompare external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента Cuffcompare в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="97"/>
        <source>Cuffmerge tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="98"/>
        <source>The path to the Cuffmerge external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента Cuffmerge в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="101"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="102"/>
        <source>The folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="116"/>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="119"/>
        <source>Set of annotations</source>
        <translation>Набор аннотаций</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="117"/>
        <source>Annotations for merging</source>
        <translation>Аннотации для соединения</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="120"/>
        <source>Merged annotations</source>
        <translation>Соединенные аннотации</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CuffmergeWorker.cpp" line="179"/>
        <source>Merges together supplied assemblies.</source>
        <translation>Соединяет вместе несколько сборок.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqPrompter</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="75"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="76"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="78"/>
        <source>Removes adapter sequences %1.</source>
        <translation>Удаляет переходные последовательности %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqTask</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="230"/>
        <source>Can not copy the result file to: %1</source>
        <translation>Can not copy the result file to: %1</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="265"/>
        <source>Unknown file format: </source>
        <translation>Unknown file format: </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::CutAdaptFastqWorker</name>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="96"/>
        <source>Cut Adapter</source>
        <translation>Удаление адаптеров</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="96"/>
        <source>Removes adapter sequences</source>
        <translation>Удаляет последовательности адаптеров</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="100"/>
        <source>Input File</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="100"/>
        <source>Set of FASTQ reads files</source>
        <translation>Набор ридов вормате FASTQ</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="101"/>
        <source>Output File</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="101"/>
        <source>Output FASTQ file(s)</source>
        <translation>Выходные FASTQ файлы</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="114"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="114"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="118"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="118"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="120"/>
        <source>Output file name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="120"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extension.</source>
        <translation>Имя выходного файла. По умолчанию будет использовано имя первого файла с соответствующим расширением.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="122"/>
        <source>FASTA file with 3&apos; adapters</source>
        <translation>FASTA файл с адаптерами 3&apos;</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="122"/>
        <source>A FASTA file with one or multiple sequences of adapter that were ligated to the 3&apos; end. The adapter itself and anything that follows is trimmed. If the adapter sequence ends with the &apos;$&apos; character, the adapter is anchored to the end of the read and only found if it is a suffix of the read.</source>
        <translation>FASTA файл с одним или несколькими адаптерами 3&apos;. Сам адаптер и все, что следует за ним обрезано. Если последовательность адаптера заканчивается символом &apos;$&apos;, адаптер прикрепляется к концу рида.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="128"/>
        <source>FASTA file with 5&apos; adapters</source>
        <translation>FASTA файл с адаптерами 5&apos;</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="128"/>
        <source>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end. If the adapter sequence starts with the character &apos;^&apos;, the adapter is &apos;anchored&apos;. An anchored adapter must appear in its entirety at the 5&apos; end of the read (it is a prefix of the read). A non-anchored adapter may appear partially at the 5&apos; end, or it may occur within the read. If it is found within a read, the sequence preceding the adapter is also trimmed. In all cases, the adapter itself is trimmed.</source>
        <translation>FASTA файл с одним или несколькими адаптерами. Если адаптер начинается с символа &apos;^&apos;, адаптер называется«якорь». Если будет найден такой адаптер в риде, последовательность предшествующая адаптеру также обрезается.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="138"/>
        <source>FASTA file with 5&apos; and 3&apos; adapters</source>
        <translation>FASTA файл с 5&apos; и 3&apos; концевыми адаптерами</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="138"/>
        <source>A FASTA file with one or multiple sequences of adapters that were ligated to the 5&apos; end or 3&apos; end.</source>
        <translation>FASTA файл с одним или несколькими адаптерами 5&apos; или 3&apos; концов.</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="166"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="167"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
    <message>
        <location filename="../src/cutadapt/CutadaptWorker.cpp" line="168"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::FastQCPrompter</name>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="81"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="82"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="84"/>
        <source>Builds FastQC report for file(s) %1.</source>
        <translation>Строит отчет FastQC для файлов %1.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::FastQCWorker</name>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="91"/>
        <source>FastQC Quality Control</source>
        <translation>Контроль качества (FastQC)</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="91"/>
        <source>Builds quality control reports.</source>
        <translation>этот элемент создаёт отчёты по результатам контроля качества.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="95"/>
        <source>Short reads</source>
        <translation>Короткие риды</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="95"/>
        <source>Short read data</source>
        <translation>Короткие риды</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="104"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="104"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="108"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="108"/>
        <source>Specify the output file name.</source>
        <translation>Укажите имя выходного файла.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="110"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="110"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="112"/>
        <source>List of adapters</source>
        <translation>Список адаптеров</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="112"/>
        <source>Specifies a non-default file which contains the list of adapter sequences which will be explicitly searched against the library. The file must contain sets of named adapters in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</source>
        <translation>добавить произвольный файл со списком адаптеров, для которых будет проведён поиск в данных.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="117"/>
        <source>List of contaminants</source>
        <translation>Список контаминирующих последовательностей</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="117"/>
        <source>Specifies a non-default file which contains the list of contaminants to screen overrepresented sequences against. The file must contain sets of named contaminants in the form name[tab]sequence.  Lines prefixed with a hash will be ignored.</source>
        <translation>добавить произвольный файл со контаминирующих последовательностей, в которых будет проведён поиск сверхпредстваленных последовательностей.</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="136"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="137"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="138"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
    <message>
        <location filename="../src/fastqc/FastqcWorker.cpp" line="150"/>
        <source>Auto</source>
        <translation>Авто</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="440"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="441"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="444"/>
        <source>%1 from %2 with bedtool genomecov.</source>
        <translation>%1 из %2 с bedtool genomecov.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovTask</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="681"/>
        <source>Unknown file format: </source>
        <translation>Unknown file format: </translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GenomecovWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="461"/>
        <source>Genome Coverage</source>
        <translation>Расчет покрытия генома</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="461"/>
        <source>Calculates genome coverage using bedtools genomecov.</source>
        <translation>Рассчитывает покрытие генома при помощи bedtools genomecov.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="465"/>
        <source>Input File</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="465"/>
        <source>Set of files to NGS slop</source>
        <translation>Набор файлов для NGS</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="466"/>
        <source>Output File</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="466"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="479"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="479"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="483"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="483"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="485"/>
        <source>Output file name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="485"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extension.</source>
        <translation>Имя выходного файла. По умолчанию будет использовано имя первого файла с соответствующим расширением.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="487"/>
        <source>Genome</source>
        <translation>Геном</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="487"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt;. (-g)</source>
        <translation>Для того, чтобы предотвратить расширение интервалов за пределами границ хромосом, bedtools slop требует файл генома, определяющий длину каждой хромосомы или контига.Формат файла: &lt;chromName&gt; &lt;TAB&gt; &lt;chromSize&gt;. (-g)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="489"/>
        <source>Report mode</source>
        <translation>Режим отчета</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="489"/>
        <source>&lt;b&gt;%1 (%2)&lt;/b&gt; - %3 
&lt;b&gt;%4 (%5)&lt;/b&gt; - %6 
&lt;b&gt;%7 (%8)&lt;/b&gt; - %9 
&lt;b&gt;%10 (%11)&lt;/b&gt; - %12 
&lt;b&gt;%13 (%14)&lt;/b&gt; - %15 
</source>
        <translation>&lt;b&gt;%1 (%2)&lt;/b&gt; - %3 
&lt;b&gt;%4 (%5)&lt;/b&gt; - %6 
&lt;b&gt;%7 (%8)&lt;/b&gt; - %9 
&lt;b&gt;%10 (%11)&lt;/b&gt; - %12 
&lt;b&gt;%13 (%14)&lt;/b&gt; - %15</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="510"/>
        <source>Split</source>
        <translation>Разделить</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="512"/>
        <source>Strand</source>
        <translation>Прямой</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="512"/>
        <source>Calculate coverage of intervals from a specific strand. With BED files, requires at least 6 columns (strand is column 6). (-strand)</source>
        <translation>Рассчитывает покрытие Calculate coverage of intervals from a specific strand. With BED files, requires at least 6 columns (strand is column 6). (-strand)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="514"/>
        <source>5 prime</source>
        <translation>5 начало</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="516"/>
        <source>3 prime</source>
        <translation>3 начало</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="518"/>
        <source>Max</source>
        <translation>Макс</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="518"/>
        <source>Combine all positions with a depth &gt;= max into a single bin in the histogram. (-max)</source>
        <translation>Смешивает все позиции с глубиной&gt; = макс в одну ячейку на гистограмме. (-max)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="520"/>
        <source>Scale</source>
        <translation>Шкала</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="520"/>
        <source>Scale the coverage by a constant factor.Each coverage value is multiplied by this factor before being reported. Useful for normalizing coverage by, e.g., reads per million (RPM). Default is 1.0; i.e., unscaled. (-scale)</source>
        <translation>шкалировать покрытие с учётом постоянного фактора. Каждое значение покрытия будет умножено на этот фактор. Применима для расчёта нормализованного покрытия, например, числа ридов на миллион (RPM). Значение по умолчанию — 1 (без масштабирования). (Параметр -scale)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="522"/>
        <source>Trackline</source>
        <translation>Трэк</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="522"/>
        <source>Adds a UCSC/Genome-Browser track line definition in the first line of the output. (-trackline)</source>
        <translation>добавляет определение трэка  UCSC Genome Browser в первую строку выходного файла. (Параметр -trackline)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="524"/>
        <source>Trackopts</source>
        <translation>След</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="524"/>
        <source>Writes additional track line definition parameters in the first line. (-trackopts)</source>
        <translation>Записывает дополнительную линию следа в первую линию. (-trackopts)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="561"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="559"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="510"/>
        <source>Treat “split” BAM or BED12 entries as distinct BED intervals when computing coverage. For BAM files, this uses the CIGAR “N” and “D” operations to infer the blocks for computing coverage. For BED12 files, this uses the BlockCount, BlockStarts, and BlockEnds fields (i.e., columns 10,11,12). (-split)</source>
        <translation>Определяет BAM или BED12 записи как различные BED промежутки когда вычисляется покрытие. (-split)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="514"/>
        <source>Calculate coverage of 5’ positions (instead of entire interval). (-5)</source>
        <translation>Вычисляет покрытие позиций 5&apos;. (-5)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="516"/>
        <source>Calculate coverage of 3’ positions (instead of entire interval). (-3)</source>
        <translation>Вычисляет покрытие позиций 3&apos;. (-3)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="560"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::GffreadPrompter</name>
    <message>
        <location filename="../src/cufflinks/GffreadWorker.cpp" line="242"/>
        <source>Extract transcript sequences from the genomic sequence from &lt;u&gt;%1&lt;/u&gt; using transcripts from &lt;u&gt;%2&lt;/u&gt; and save them to the file %3.</source>
        <translation>Получает транскрипционные последовательности из геномной последовательности из &lt;u&gt;%1&lt;/u&gt; используя транскрипции из &lt;u&gt;%2&lt;/u&gt; и сохраняет их в файл %3.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HeadCropStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/HeadCropStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;HEADCROP&lt;/h4&gt;&lt;p&gt;This step removes the specified number of bases, regardless of quality, from the beginning of the read.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Length&lt;/b&gt;: the number of bases to remove from the start of the read.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/HeadCropStep.cpp" line="48"/>
        <source>The number of bases to remove from the start of the read.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HmmerBuildPrompter</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="125"/>
        <source>For each MSA from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>Для каждого MSA из&lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="127"/>
        <source>%1 builds a HMMER profile.</source>
        <translation>%1 построит HMMER профиль.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HmmerBuildWorker</name>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="63"/>
        <source>Input MSA</source>
        <translation>Входное MSA</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="63"/>
        <source>Input multiple sequence alignment for building statistical model.</source>
        <translation>Входное множественное выравнивание для построения статистической модели.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="64"/>
        <source>HMM3 profile</source>
        <translation>HММ3 профиль</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="64"/>
        <source>Produced HMM3 profile URL</source>
        <translation>Профиль построенный с помощью HMM3</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="74"/>
        <source>Random seed</source>
        <translation>Затравка</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="74"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation>Генератор случайной затравки. 0 - означает, что один раз будет использована случайная затравка.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="78"/>
        <source>HMM3 Build</source>
        <translation>Построение профиля с помощью HMMER3</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="78"/>
        <source>Builds a HMM3 profile from a multiple sequence alignment.&lt;p&gt;The HMM3 profile is a statistical model which captures position-specific information about how conserved each column of the alignment is, and which residues are likely.</source>
        <translation>Строит HMM3 профиль из множественного выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="94"/>
        <source>HMMER3 Tools</source>
        <translation>HMMER3 инструменты</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerBuildWorker.cpp" line="190"/>
        <source>Built HMMER profile</source>
        <translation>Построить профиль с помощью HMMER</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HmmerSearchPrompter</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="182"/>
        <source>For each sequence from &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>Для каждой последовательности из &lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="183"/>
        <source>using all profiles provided by &lt;u&gt;%1&lt;/u&gt;,</source>
        <translation>используя все профили произведенные &lt;u&gt;%1&lt;/u&gt;,</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="187"/>
        <source>%1 search HMMER signals %2. &lt;br&gt;Output the list of found regions annotated as &lt;u&gt;%4&lt;/u&gt;.</source>
        <translation>%1 искать сигналы HMM3 %2. &lt;br&gt;Аннотировать найденные регионы как &lt;u&gt;%4&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::HmmerSearchWorker</name>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="82"/>
        <source>Filter by</source>
        <translation>Фильтровать по</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="83"/>
        <source>Parameter to filter results by.</source>
        <translation>Параметр, по которому будут отфильтрованы результаты.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="84"/>
        <source>HMMER profile</source>
        <translation>HMMER профиль</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="84"/>
        <source>HMMER profile(s) URL(s) to search with.</source>
        <translation>HMMER профили для поиска.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="85"/>
        <source>Input sequence</source>
        <translation>Входная последовательность</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="85"/>
        <source>An input sequence (nucleotide or protein) to search in.</source>
        <translation>Входная последовательность.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="86"/>
        <source>HMMER annotations</source>
        <translation>HMMER аннотации</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="86"/>
        <source>Annotations marking found similar sequence regions.</source>
        <translation>Аннотации размечающие найденные регионы совпадения с профилем HMMER.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="98"/>
        <source>Result annotation</source>
        <translation>Имя аннотации</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="98"/>
        <source>A name of the result annotations.</source>
        <translation>Имя аннотации для разметки результатов.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="99"/>
        <source>Seed</source>
        <translation>Затравка</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="99"/>
        <source>Random generator seed. 0 - means that one-time arbitrary seed will be used.</source>
        <translation>Генератор случайной затравки. 0 - означает, что один раз будет использована случайная затравка.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="100"/>
        <source>Filter by high E-value</source>
        <translation>Фильтр на большие Е</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="100"/>
        <source>Report domains with e-value less than.</source>
        <translation>Область с меньшими значениями Е.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="101"/>
        <source>Filter by low score</source>
        <translation>Фильтр на малые веса</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="101"/>
        <source>Report domains with score greater than.</source>
        <translation>Области с весами большими чем.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="117"/>
        <source>HMM3 Search</source>
        <translation>Поиск с помощью HMMER3</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="117"/>
        <source>Searches each input sequence for significantly similar sequence matches to all specified HMM profiles. In case several profiles were supplied, searches with all profiles one by one and outputs united set of annotations for each sequence.</source>
        <translation>Ищет значимые совпадения с указанными HMM профайлами в каждой входной последовательности и выдаёт набор аннотированных регионов. Профайлов может быть несколько, поиск отработает для каждого профайла и выдаст общий набор аннотаций для каждой последовательности.</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="157"/>
        <source>HMMER3 Tools</source>
        <translation>HMMER3 инструменты</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="230"/>
        <source>Value for attribute name is empty, default name used</source>
        <translation>Имя не указано, использовано значение по умолчанию</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="281"/>
        <source>Find HMMER signals in %1</source>
        <translation>Поиск HMMER сигналов в &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="286"/>
        <source>Bad sequence supplied to input: %1</source>
        <translation>Некорректная последовательность: %1</translation>
    </message>
    <message>
        <location filename="../src/hmmer/HmmerSearchWorker.cpp" line="316"/>
        <source>Found %1 HMMER signals</source>
        <translation>Найдено %1 сигналов HMMMER</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::IlluminaClipAdditionalSettingsDialog</name>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipStep.cpp" line="268"/>
        <source>Apply</source>
        <translation>Применить</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipStep.cpp" line="270"/>
        <source>True</source>
        <translation>Истина</translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipStep.cpp" line="271"/>
        <source>False</source>
        <translation>Ложь</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::IlluminaClipSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipStep.cpp" line="242"/>
        <source>Open FASTA with adapters</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::IlluminaClipStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/IlluminaClipStep.cpp" line="47"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;ILLUMINACLIP&lt;/h4&gt;&lt;p&gt;This step is used to find and remove Illumina adapters.&lt;/p&gt;&lt;p&gt;Trimmomatic first compares short sections of an adapter and a read. If they match enough, the entire alignment between the read and adapter is scored. For paired-end reads, the &quot;palindrome&quot; approach is also used to improve the result. See Trimmomatic manual for details.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Adapter sequences&lt;/b&gt;: a FASTA file with the adapter sequences. Files for TruSeq2 (GAII machines), TruSeq3 (HiSeq and MiSeq machines) and Nextera kits for SE and PE reads are now available by default. The naming of the various sequences within the specified file determines how they are used.&lt;/li&gt;&lt;li&gt;&lt;b&gt;Seed mismatches&lt;/b&gt;: the maximum mismatch count in short sections which will still allow a full match to be performed.&lt;/li&gt;&lt;li&gt;&lt;b&gt;Simple clip threshold&lt;/b&gt;: a threshold for simple alignment mode. Values between 7 and 15 are recommended. A perfect match of a 12 base sequence will score just over 7, while 25 bases are needed to score 15.&lt;/li&gt;&lt;li&gt;&lt;b&gt;Palindrome clip threshold&lt;/b&gt;: a threshold for palindrome alignment mode. For palindromic matches, a longer alignment is possible. Therefore the threshold can be in the range of 30. Even though this threshold is very high (requiring a match of almost 50 bases) Trimmomatic is still able to identify very, very short adapter fragments.&lt;/li&gt;&lt;/ul&gt;&lt;p&gt;There are also two optional parameters for palindrome mode: &lt;b&gt;Min adapter length&lt;/b&gt; and &lt;b&gt;Keep both reads&lt;/b&gt;.&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::LeadingStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/LeadingStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;LEADING&lt;/h4&gt;&lt;p&gt;This step removes low quality bases from the beginning. As long as a base has a value below this threshold the base is removed and the next base will be investigated.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Quality threshold&lt;/b&gt;: the minimum quality required to keep a base.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/LeadingStep.cpp" line="49"/>
        <source>The minimum quality required to keep a base.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSPrompter</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="559"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="566"/>
        <source>Uses &lt;u&gt;%1&lt;/u&gt; as treatment</source>
        <translation>Использует &lt;u&gt;%1&lt;/u&gt; в качестве обработки</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="568"/>
        <source> and &lt;u&gt;%1&lt;/u&gt; as control</source>
        <translation> и &lt;u&gt;%1&lt;/u&gt; в качестве контроля</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="571"/>
        <source> to call peaks.</source>
        <translation>вызова сигналов.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="573"/>
        <source> Outputs all files to &lt;u&gt;%1&lt;/u&gt; folder</source>
        <translation> Сохраняет все выходные файлы в&lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="575"/>
        <source> and pileup with &lt;u&gt;%1&lt;/u&gt; span</source>
        <translation> и накапливается в диапазоне &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MACSWorker</name>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="190"/>
        <source>Bad model fold region: %1. Default region is used</source>
        <translation>Bad model fold region: %1. Default region is used</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="190"/>
        <source>unrecognized parsing error</source>
        <translation>unrecognized parsing error</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="255"/>
        <source>Input control and treatment annotations are the same</source>
        <translation>Входные контрольные и обработанные аннотации одинаковы</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="269"/>
        <source>Treatment features</source>
        <translation>Treatment особенности</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="270"/>
        <source>ChIP-seq treatment features.</source>
        <translation>ChIP-seq treatment особенности.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="272"/>
        <source>Control features</source>
        <translation>Контрольные особенности</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="273"/>
        <source>Control features (Optional).</source>
        <translation>Контрольные особенности.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="278"/>
        <source>MACS data</source>
        <translation>Данные MACS</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="279"/>
        <source>ChIP-seq treatment features and control features (optional) to call peaks with MACS.</source>
        <translation>ChIP-seq treatment особенности и контрольные особенности для вызова пиков при помощи MACS.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="287"/>
        <source>Peak regions</source>
        <translation>Пиковые регионы</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="288"/>
        <source>Peak locations. Typically used in gene association study like CEAS, or correlation calculation.</source>
        <translation>Расположение пиков. Обычно используется в CEAS или расчете корреляции.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="290"/>
        <source>Peak summits</source>
        <translation>Пределы пиков</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="291"/>
        <source>Peak summits locations for every peaks. Typically used in DNA motif analysis or conservation check.</source>
        <translation>Расположение пределов для каждого пика.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="293"/>
        <source>Treatment fragments pileup (wiggle)</source>
        <translation>Накопление treatment фрагментов (wiggle)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="294"/>
        <source>Wiggle format files which can be imported to UCSC genome browser/GMOD/Affy IGB.</source>
        <translation>Файлы формата wiggle, которые могут быть импортированы в геномный браузер/GMOD/Affy IGB.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="297"/>
        <source>MACS output data</source>
        <translation>Выходные данные MACS</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="298"/>
        <source>ChIP-seq peaks and summits. Pileup data (optional)</source>
        <translation>ChIP-seq пики и пределы. Накопленные данные</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="310"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="311"/>
        <source>Folder to save MACS output files.</source>
        <translation>Папка для сохранения выходных файлов MACS.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="313"/>
        <source>Name</source>
        <translation>Имя</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="314"/>
        <source>The name string of the experiment. MACS will use this string NAME to create output files like &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; and so on. So please avoid any confliction between these filenames and your existing files (--name).</source>
        <translation>Имя эксперимента. MACS будет использовать это имя для создания выходных файлов, например &apos;NAME_peaks.xls&apos;, &apos;NAME_negative_peaks.xls&apos;, &apos;NAME_peaks.bed&apos;, &apos;NAME_summits.bed&apos;, &apos;NAME_model.r&apos; (--name).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="318"/>
        <source>Wiggle output</source>
        <translation>Выходной файл в формате wiggle</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="319"/>
        <source>If this flag is on, MACS will store the fragment pileup in wiggle format for the whole genome data instead of for every chromosomes (--wig) (--single-profile).</source>
        <translation>Если этот флаг установлен, MACS будет сохранять накопленные фрагменты в формате wiggle для всего генома вместо каждой отдельной хромосомы (--wig) (--single-profile).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="321"/>
        <source>Wiggle space</source>
        <translation>Пространство wiggle</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="322"/>
        <source>By default, the resolution for saving wiggle files is 10 bps,i.e., MACS will save the raw tag count every 10 bps. You can change it along with &apos;--wig&apos; option (--space).</source>
        <translation>По умолчанию, разрешение для сохранения файлов 10 нукл. Вы можете изменить это при помощи опции &apos;--wig&apos; (--space).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="325"/>
        <source>Genome size (Mbp)</source>
        <translation>Размер генома (Mbp)</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="326"/>
        <source>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt; It&apos;s the mappable genome size or effective genome size which is defined as the genome size which can be sequenced. Because of the repetitive features on the chromosomes, the actual mappable genome size will be smaller than the original size, about 90% or 70% of the genome size (--gsize).</source>
        <translation>Homo sapience - 2700 Mbp&lt;br&gt;Mus musculus - 1870 Mbp&lt;br&gt;Caenorhabditis elegans - 90 Mbp&lt;br&gt;Drosophila melanogaster  - 120 Mbp&lt;br&gt;. Это графический размер генома или эффективный размер генома, который определяется как размер генома, который может быть упорядочен. Из-за повторяющихся особенностей на хромосомах, фактическое отображаемыми размер генома будет меньше исходного размера, приблизительно 90% или 70% от размера генома (--gsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="335"/>
        <source>P-value</source>
        <translation>P-значение</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="336"/>
        <source>P-value cutoff. Default is 0.00001, for looser results, try 0.001 instead (--pvalue).</source>
        <translation>P-значение среза. По умолчанию 0.00001 (--pvalue).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="338"/>
        <source>Q-value</source>
        <translation>Q-значение</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="339"/>
        <source>Minimum FDR (q-value) cutoff for peak detection.</source>
        <translation>Минимальный срез FDR для поиска пика.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="341"/>
        <source>Use model</source>
        <translation>Использовать модель</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="342"/>
        <source>Whether or not to use MACS paired peaks model (--nomodel).</source>
        <translation>Использовать или нет модель парных пиков MACS (--nomodel).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="344"/>
        <source>Model fold</source>
        <translation>Моедль сложения</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="345"/>
        <source>Select the regions within MFOLD range of high-confidence enrichment ratio against. &lt;b&gt;Model fold&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is true, which is the foldchange to chose paired peaks to build paired peaks model. Users need to set a lower(smaller) and upper(larger) number for fold change so that MACS will only use the peaks within these foldchange range to build model (--mfold).</source>
        <translation>&lt;b&gt;Модель сложения&lt;/b&gt; доступна когда параметр&lt;b&gt;Использовать модель&lt;/b&gt; истина. Необходимо выбрать нижнюю и верхнюю границы для этой модели, MACS будет использовать только пики из этого интервала для построения модели (--mfold).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="350"/>
        <source>Shift size</source>
        <translation>Величина сдвига</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="351"/>
        <source>An arbitrary shift value used as a half of the fragment size when model is not built. &lt;b&gt;Shift size&lt;/b&gt; is available when &lt;b&gt;Use model&lt;/b&gt; is false, which will represent the HALF of the fragment size of your sample. If your sonication and size selection size is 300 bps, after you trim out nearly 100 bps adapters, the fragment size is about 200 bps, so you can specify 100 here (--shiftsize).</source>
        <translation>Величина сдвига используется как половина размера фрагмента в том случае, когда модель не построена. &lt;b&gt;Величина сдвига&lt;/b&gt; доступна когда параметр &lt;b&gt;Использовать модель&lt;/b&gt; не истина (--shiftsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="356"/>
        <source>Keep duplicates</source>
        <translation>Сохранять дубликаты</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="357"/>
        <source>It controls the MACS behavior towards duplicate tags at the exact same location -- the same coordination and the same strand. The default &lt;b&gt;auto&lt;/b&gt; option makes MACS calculate the maximum tags at the exact same location based on binomal distribution using 1e-5 as pvalue cutoff; and the &lt;b&gt;all&lt;/b&gt; option keeps every tags. If an &lt;b&gt;integer&lt;/b&gt; is given, at most this number of tags will be kept at the same location (--keep-dup).</source>
        <translation>Этот параметр контролирует поведение MACS для дубликатов, найденных в одном месте, с одинаковыми координатами и в одинаковой цепи (--keep-dup).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="361"/>
        <source>Band width</source>
        <translation>Ширина полосы</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="362"/>
        <source>The band width which is used to scan the genome for model building. You can set this parameter as the sonication fragment size expected from wet experiment. Used only while building the shifting model (--bw).</source>
        <translation>Ширина полосы, которая используется для сканирования генома для построения модели. Используется только при построении модели сдвига (--bw).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="366"/>
        <source>Extended fragment pileup</source>
        <translation>Накопления расширенного фрагмента</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="367"/>
        <source>Whether or not to generate extended fragment pileup, local lambda and score tracks at every bp.</source>
        <translation>Генерировать или нет накопления разширенного фрагмента.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="371"/>
        <source>Tag size (optional)</source>
        <translation>Размер ярлыка</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="372"/>
        <source>Length of reads. Determined from first 10 reads if not specified (input &lt;b&gt;0&lt;/b&gt;) (--tsize).</source>
        <translation>Длина ридов. Определеятся из первых 10 ридов (--tsize).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="376"/>
        <source>Use lambda</source>
        <translation>Использовать лямбда</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="377"/>
        <source>Whether to use local lambda model which can use the local bias at peak regions to throw out false positives (--nolambda).</source>
        <translation>Следует ли использовать локальную модель лямбда (--nanolambda).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="379"/>
        <source>Small nearby region</source>
        <translation>Небольшой ближайший регион</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="380"/>
        <source>The small nearby region in basepairs to calculate dynamic lambda. This is used to capture the bias near the peak summit region. Invalid if there is no control data (--slocal).</source>
        <translation>Небольшой ближайший регион в парах оснований для расчета динамической лямбда. Это используется, чтобы захватить смещения вблизи области пика. (--local).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="383"/>
        <source>Large nearby region</source>
        <translation>Большой ближайший регион</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="384"/>
        <source>The large nearby region in basepairs to calculate dynamic lambda.  This is used to capture the surround bias (--llocal).</source>
        <translation>Большой ближайший регион в парах оснований для расчета динамической лямбда. Это используется, чтобы захватить объемные смещения (--local).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="387"/>
        <source>Auto bimodal</source>
        <translation>Авто бимодальный</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="388"/>
        <source>Whether turn on the auto pair model process.If set, when MACS failed to build paired model, it will use the nomodelsettings, the “Shift size” parameter to shift and extend each tags (--on-auto).</source>
        <translation>Автоматический посик пары. Если включено, когда MACS не удалось построить парную модель, он будет использовать nomodelsettings, параметр «Shift size» для переключения и расширения каждого тега (--on-auto).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="392"/>
        <source>Scale to large</source>
        <translation>Масштаб для больших</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="393"/>
        <source> When set, scale the small sample up to the bigger sample.By default, the bigger dataset will be scaled down towards the smaller dataset,which will lead to smaller p/qvalues and more specific results.Keep in mind that scaling down will bring down background noise more (--to-large).</source>
        <translation>Когда установлено, масштаб увеличивается до большого. (--to-large).</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="398"/>
        <source>Shift control</source>
        <translation>Контроль сдвига</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="399"/>
        <source>When set, control tags will be shifted just as ChIP tags according to their strand before the extension of d, slocal and llocal. By default, control tags are extended centeredat their current positions regardless of strand. You may consider to turn this option on whilecomparing two ChIP datasets of different condition but the same factor.</source>
        <translation>Если установлен ярлыки управления будут сдвинуты как ChIP ярлыки.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="404"/>
        <source>Half-extend</source>
        <translation>Расширение наполовину</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="405"/>
        <source>When set, MACS extends 1/2 d size for each fragment centered at its middle point.</source>
        <translation>Когда установлен, MACS расширяет размер на 1/2 от размера для каждого фрагмента.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="407"/>
        <source>Broad</source>
        <translation>Широкий</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="408"/>
        <source>If set, MACS will try to call broad peaks by linking nearby highly enriched regions.The linking region is controlled by another cutoff through “Broad cutoff”.The maximum linking region length is 4 times of d from MACS.</source>
        <translation>Если установлено, MACS будет пытаться вызвать широкие пики.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="412"/>
        <source>Broad cutoff</source>
        <translation>Широкий срез</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="413"/>
        <source>Cutoff for broad region. This option is not available unless “Broad” is set.If “P-value” is set, this is a pvalue cutoff, otherwise, it&apos;s a qvalue cutoff.</source>
        <translation>Срез для широких регионов.</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="537"/>
        <source>Find Peaks with MACS</source>
        <translation>Поиск пиков с помощью MACS</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="538"/>
        <source>Performs peak calling for ChIP-Seq data.</source>
        <translation>Выполняет поиск пиков для данных  ChIP-Seq.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTPrompter</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="134"/>
        <source> from %1</source>
        <translation> из %1</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="136"/>
        <source>Aligns each MSA supplied &lt;u&gt;%1&lt;/u&gt; with &lt;u&gt;&quot;MAFFT&quot;&lt;/u&gt;.</source>
        <translation>Выравнивает каждое MSA из &lt;u&gt;%1&lt;/u&gt; при помощи &quot;&lt;u&gt;MAFFT&lt;/u&gt;&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MAFFTWorker</name>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="63"/>
        <source>Input MSA</source>
        <translation>Входное MSA</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="63"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation>Множественное выравнивание, которое будет обработано.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="64"/>
        <source>Multiple sequence alignment</source>
        <translation>Множественное выравнивание</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="64"/>
        <source>Result of alignment.</source>
        <translation>Результат выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="73"/>
        <source>Gap Open Penalty</source>
        <translation>Штраф за открытие пробела</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="73"/>
        <source>Gap Open Penalty.</source>
        <translation>Штраф за открытие пробела.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="74"/>
        <source>Offset</source>
        <translation>Смещение</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="74"/>
        <source>Works like gap extension penalty.</source>
        <translation>Работает как штраф за расширение пробела.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="75"/>
        <source>Maximum number of iterative refinement.</source>
        <translation>Максимальное число итерационных уточнений.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="76"/>
        <source>External tool path.</source>
        <translation>Путь до внешнего инструмента.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="77"/>
        <source>Folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="179"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to MAFFT.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="75"/>
        <source>Max Iteration</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="76"/>
        <source>Tool Path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="77"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="85"/>
        <source>Align with MAFFT</source>
        <translation>Выравнивание с помощью MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="208"/>
        <source>Aligned %1 with MAFFT</source>
        <translation>Aligned %1 with MAFFT</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTWorker.cpp" line="85"/>
        <source>MAFFT is a multiple sequence alignment program for unix-like operating systems. &lt;p&gt;&lt;dfn&gt;It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of &amp;lt;&amp;#126;200 sequences), FFT-NS-2 (fast; for alignment of &amp;lt;&amp;#126;10,000 sequences), etc. &lt;/dfn&gt;&lt;/p&gt;</source>
        <translation>MAFFT - это программа для выравнивания последовательностей для Unix-подобных операционных систем. &lt;p&gt;&lt;dfn&gt;MAFFT предлагает широкий выбор методов выравнивания, L-INS-i (точно; для выравнивания около 200 последовательностей),  FFT-NS-2 (быстро; для выравнивания около 10.000 последовательностей).&lt;/dfn&gt;&lt;/p&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MaxInfoStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/MaxInfoStep.cpp" line="34"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;MAXINFO&lt;/h4&gt;&lt;p&gt;This step performs an adaptive quality trim, balancing the benefits of retaining longer reads against the costs of retaining bases with errors. See Trimmomatic manual for details.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Target length&lt;/b&gt;: the read length which is likely to allow the location of the read within the target sequence. Extremely short reads, which can be placed into many different locations, provide little value. Typically, the length would be in the order of 40 bases, however, the value also depends on the size and complexity of the target sequence.&lt;/li&gt;&lt;li&gt;&lt;b&gt;Strictness&lt;/b&gt;: the balance between preserving as much read length as possible vs. removal of incorrect bases. A low value of this parameter (&lt;0.2) favours longer reads, while a high value (&gt;0.8) favours read correctness.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::MinLenStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/MinLenStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;MINLEN&lt;/h4&gt;&lt;p&gt;This step removes reads that fall below the specified minimal length. If required, it should normally be after all other processing steps. Reads removed by this step will be counted and included in the &quot;dropped reads&quot; count.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Length&lt;/b&gt;: the minimum length of reads to be kept.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/MinLenStep.cpp" line="50"/>
        <source>The minimum length of reads to be kept.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::NoSettingsWidget</name>
    <message>
        <location filename="../src/trimmomatic/util/NoSettingsWidget.cpp" line="34"/>
        <source>There are no settings for this step.</source>
        <translation></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GenePrompter</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="317"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="322"/>
        <source>Uses annotations from &lt;u&gt;%1&lt;/u&gt; as peak regions </source>
        <translation>Использует аннотации из &lt;u&gt;%1&lt;/u&gt; в качестве пиковых регионов </translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="323"/>
        <source> to annotate with genes nearby from &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> для аннотации генов рядом с &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::Peak2GeneWorker</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="190"/>
        <source>Treatment features</source>
        <translation>Treatment особенности</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="191"/>
        <source>Result peaks of MACS.</source>
        <translation>MACS пики.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="195"/>
        <source>Peak2gene data</source>
        <translation>Данные Peak2gene</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="196"/>
        <source>MACS peaks to get the refgenes near the peak summit/center.</source>
        <translation>Пики MACS, чтобы получить refgenes вблизи пика в центре.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="204"/>
        <source>Gene regions</source>
        <translation>Регионы гена</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="205"/>
        <source>Annotation for each gene, containing all the peaks nearby.</source>
        <translation>Аннотация для каждого гена, содержащая все близлежащие пики.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="207"/>
        <source>Peak regions</source>
        <translation>Пиковые регионы</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="208"/>
        <source>Annotation for each peak, containing all the genes nearby.</source>
        <translation>Аннотрация для каждого пика, содержащая все близлежащие гены.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="210"/>
        <source>Gene regions URL</source>
        <translation>Регионы гена</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="211"/>
        <source>URL to file with annotation for each gene, containing all the peaks nearby.</source>
        <translation>Файл с аннотациями для каждого гена, содержащие все пики.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="213"/>
        <source>Peak regions URL</source>
        <translation>Пиковые регионы</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="214"/>
        <source>URL to file with annotation for each peak, containing all the genes nearby.</source>
        <translation>Файл с аннотациями для каждого гена, содержащий все близлежащие пики.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="217"/>
        <source>Peak2gene output data</source>
        <translation>Выходные данные Peak2gene</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="218"/>
        <source>Genes containing all the peaks nearby and peaks containing all the genes nearby.</source>
        <translation>Гены, содержащие все близлежащие пики и пики, содержащие все близлежащие гены.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="231"/>
        <source>Output type</source>
        <translation>Тип выходного файла</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="232"/>
        <source>Select which type of genes need to output. &lt;b&gt;up&lt;/b&gt; for genes upstream to peak summit, &lt;b&gt;down&lt;/b&gt; for genes downstream to peak summit, &lt;b&gt;all&lt;/b&gt; for both &lt;b&gt;up&lt;/b&gt; and &lt;b&gt;down&lt;/b&gt; (--op).</source>
        <translation>выберите тип генов для выходных файлов &lt;b&gt;up&lt;/b&gt; для вышележащих генов, &lt;b&gt;down&lt;/b&gt; для нижележащих генов, &lt;b&gt;all&lt;/b&gt; для обоих &lt;b&gt;up&lt;/b&gt; и &lt;b&gt;down&lt;/b&gt;. Соответствует опции --op.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="237"/>
        <source>Official gene symbols</source>
        <translation>Официальные символы генов</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="238"/>
        <source>Output &lt;b&gt;official gene symbol&lt;/b&gt; instead of &lt;b&gt;refseq name&lt;/b&gt; (--symbol).</source>
        <translation>выходной &lt;b&gt;официальный символ гена&lt;/b&gt; вместо &lt;b&gt;имени refseq&lt;/b&gt;. Соответствует опции --symbol.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="240"/>
        <source>Distance</source>
        <translation>Расстояние</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="241"/>
        <source>Set a number which unit is base. It will get the refGenes in n bases from peak center (--distance).</source>
        <translation>установите число, которое является базовым. Он будет получать refGenes в n базах из центра пика. Соответствует опции --distance.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="243"/>
        <source>Genome file</source>
        <translation>Файл генома</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="244"/>
        <source>Select a genome file (sqlite3 file) to search refGenes (--genome).</source>
        <translation>выберите файл генома для (sqlite3 файл) для поиска refGenes. Соответствует опции --genome.</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="297"/>
        <source>Annotate Peaks with peak2gene</source>
        <translation>Аннотация пиков с помощью peak2gene</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneWorker.cpp" line="298"/>
        <source>Gets refGenes near the ChIP regions identified by a peak-caller.</source>
        <translation>этот элемент получает refGenes рядом с регионами ChIP найденными поиском пиков.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosComboBoxWithChecksDelegate</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="143"/>
        <source>Hint:</source>
        <translation>Совет:</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="143"/>
        <source>Use &apos;cistrome.xml&apos; to descrease the computation time. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation>Используйте &apos;cistrome.xml&apos; чтобы понизить время вычисления. Это полная коллекция мотивов из других баз данных с аналогичными мотивами.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosPrompter</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="344"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="352"/>
        <source>Uses regions from &lt;u&gt;%1&lt;/u&gt; to find motifs enriched in them.</source>
        <translation>Использует регионы из &lt;u&gt;%1&lt;/u&gt; для поиска мотивов обогащенных в них.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="353"/>
        <source> Genome assembly: &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation> Геномная сборка: &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="355"/>
        <source> Finds de novo motifs.</source>
        <translation>Ищет мотивы de novo.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="358"/>
        <source> Uses &lt;u&gt;%1&lt;/u&gt; known motifs.</source>
        <translation> Использует известные мотивы &lt;u&gt;%1&lt;/u&gt;.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="361"/>
        <source> Outputs all result files to &lt;u&gt;%1&lt;/u&gt; folder</source>
        <translation>Записывает все выходные файлы в &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SeqPosWorker</name>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="228"/>
        <source>Input regions</source>
        <translation>Входные регионы</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="229"/>
        <source>Peak summits.</source>
        <translation>Пределы пиков.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="233"/>
        <source>SeqPos data</source>
        <translation>Данные SeqPos</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="234"/>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="325"/>
        <source>Finds motifs enriched in a set of regions.</source>
        <translation>Находит мотивы, обогащенные в наборе областей.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="242"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="243"/>
        <source>The folder to store seqpos results.</source>
        <translation>Папка для сохранения результатов seqpos.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="245"/>
        <source>Genome assembly version</source>
        <translation>Версия геномной сборки</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="246"/>
        <source>UCSC database version (GENOME).</source>
        <translation>ВЕрсия базы данных UCSC (GENOME).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="248"/>
        <source>De novo motifs</source>
        <translation>Мотивы de novo</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="249"/>
        <source>Run de novo motif search (-d).</source>
        <translation>Запускает поиск de novo мотивов (-d).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="251"/>
        <source>Motif database</source>
        <translation>База данных мотивов</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="252"/>
        <source>Known motif collections (-m). Warning: computation time increases with selecting additional databases. It is recommended to use cistrome.xml. It is a comprehensive collection of motifs from the other databases with similar motifs deleted.</source>
        <translation>Известная коллекция мотивов (-m). Предупреждение: время вычисления возрастает при выборе этих баз данных. Рекомендуется использовать cistrome.xml.</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="255"/>
        <source>Output file name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="256"/>
        <source>Name of the output file which stores new motifs found during a de novo search (-n).</source>
        <translation>Имя выходной файла (-n).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="258"/>
        <source>Region width</source>
        <translation>Ширина региона</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="259"/>
        <source>Width of the region to be scanned for motifs; depends on a resolution of assay (-w).</source>
        <translation>Ширина региона (-w).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="261"/>
        <source>Pvalue cutoff</source>
        <translation>Срез Pзначения</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="262"/>
        <source>Pvalue cutoff for the motif significance (-p).</source>
        <translation>Срез Pзначения (-p).</translation>
    </message>
    <message>
        <location filename="../src/seqpos/SeqPosWorker.cpp" line="324"/>
        <source>Collect Motifs with SeqPos</source>
        <translation>Сбор мотивов с помощью SeqPos</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ShortReadsAlignerPrompter</name>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="426"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="432"/>
        <source>Aligns upstream oriented reads from &lt;u&gt;%1&lt;/u&gt; and downstream oriented reads from &lt;u&gt;%2&lt;/u&gt; </source>
        <translation>Выравнивает верховые риды из &lt;u&gt;%1&lt;/u&gt; и низовые риды из &lt;u&gt;%2&lt;/u&gt; </translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="434"/>
        <source>Maps input reads from &lt;u&gt;%1&lt;/u&gt; </source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="440"/>
        <source> to reference sequence with index &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="443"/>
        <source> to reference sequence &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlidingWindowStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/SlidingWindowStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;SLIDINGWINDOW&lt;/h4&gt;&lt;p&gt;This step performs a sliding window trimming, cutting once the average quality within the window falls below a threshold. By considering multiple bases, a single poor quality base will not cause the removal of high quality data later in the read.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Window size&lt;/b&gt;: the number of bases to average across.&lt;/li&gt;&lt;li&gt;&lt;b&gt;Quality threshold&lt;/b&gt;: the average quality required.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedPrompter</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="73"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="74"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="76"/>
        <source>Increases the size of each feature in files from %1 with bedtool slop.</source>
        <translation>Увеличивает размер каждого элемента в файлах из%1 при помощи bedtools.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SlopbedWorker</name>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="93"/>
        <source>slopBed</source>
        <translation>Продление аннотаций (slopBed)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="93"/>
        <source>Increases the size of each feature in files using bedtools slop.</source>
        <translation>Увеличивает размер каждой аннотации в файлах при помощи bedtools slop.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="97"/>
        <source>Input File</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="97"/>
        <source>Set of files to bedtools slop</source>
        <translation>Набор файлов для bedtools slop</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="98"/>
        <source>Output File</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="98"/>
        <source>Output file</source>
        <translation>Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="111"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="111"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="115"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="115"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="117"/>
        <source>Output file name</source>
        <translation>Имя выходного файла</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="117"/>
        <source>A name of an output file. If default of empty value is provided the output name is the name of the first file with additional extension.</source>
        <translation>Имя выходного файла. По умолчанию будет использовано имя первого файла с соответствующим расширением.</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="119"/>
        <source>Genome</source>
        <translation>Геном</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="119"/>
        <source>In order to prevent the extension of intervals beyond chromosome boundaries, bedtools slop requires a genome file defining the length of each chromosome or contig. The format of the file is: &lt;chromName&gt;&lt;TAB&gt;&lt;chromSize&gt; (-g).</source>
        <translation>Для того, чтобы предотвратить расширение интервалов за пределами границ хромосом, для bedtools slopотстойный требуется файл генома, определяющий длину каждой хромосомы или контига. Формат файла: &lt;chromName&gt; &lt;TAB&gt; &lt;chromSize&gt; (-g).</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="121"/>
        <source>Each direction increase</source>
        <translation>Увеличение каждого направления</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="121"/>
        <source>Increase the BED/GFF/VCF entry by the same number base pairs in each direction. If this parameter is used -l and -l are ignored. Enter 0 to disable. (-b)</source>
        <translation>Увеличивает BED/ GFF / VCF на одинаковое число пар оснований в каждом направлении. Если этот параметр используется -l и -l игнорируются. Введите 0 для отключения. (-b)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="123"/>
        <source>Subtract from start</source>
        <translation>Вычесть из начала</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="123"/>
        <source>The number of base pairs to subtract from the start coordinate. Enter 0 to disable. (-l)</source>
        <translation>Число пар основания, которое нужно вычесть, начиная от стартовой координаты. Введите 0 для отключения. (-l)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="125"/>
        <source>Add to end</source>
        <translation>Добавить в конец</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="125"/>
        <source>The number of base pairs to add to the end coordinate. Enter 0 to disable. (-r)</source>
        <translation>Число пар основания для добавления в конец. Введите 0 для отключения. (-r)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="127"/>
        <source>Strand-based</source>
        <translation>На основе цепи</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="127"/>
        <source>Define -l and -r based on strand. For example. if used, -l 500 for a negative-stranded feature, it will add 500 bp to the end coordinate. (-s)</source>
        <translation>Определяет -l и -r на основе цепи. (-s)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="129"/>
        <source>As fraction</source>
        <translation>Как доля</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="131"/>
        <source>Print header</source>
        <translation>Вывести заголовок</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="131"/>
        <source>Print the header from the input file prior to results. (-header)</source>
        <translation>Вывести заголовок из входного файла в результат. (-header)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="133"/>
        <source>Filter start&gt;end fields</source>
        <translation>Фильтровать области начало&gt;конец</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="133"/>
        <source>Remove lines with start position greater than end position</source>
        <translation>Удаляет линии, в которых стартовая позиция больши конечной позиции</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="168"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="166"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="129"/>
        <source>Define -l and -r as a fraction of the feature’s length. E.g. if used on a 1000bp feature, -l 0.50, will add 500 bp “upstream”. (-pct)</source>
        <translation>Определяет -l и -r как долю от длины аннотации. (-pct)</translation>
    </message>
    <message>
        <location filename="../src/bedtools/BedToolsWorkersLibrary.cpp" line="167"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffDatabaseDialog</name>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="51"/>
        <source>Select</source>
        <translation>Выбрать</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="52"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffDatabasePropertyWidget</name>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="90"/>
        <source>Select genome</source>
        <translation>Выбрать геном</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="123"/>
        <source>%1 and %2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="124"/>
        <source>The list of genomes is not available.
Make sure %1 and %2 tools are set in the UGENE Application Settings and can be validated.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseDelegate.cpp" line="125"/>
        <source>Do you want to do it now?</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffPrompter</name>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="87"/>
        <source>unset</source>
        <translation>не указан</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="88"/>
        <source> from &lt;u&gt;%1&lt;/u&gt;</source>
        <translation> из &lt;u&gt;%1&lt;/u&gt;</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="90"/>
        <source>Annotates and filters variations %1 with SnpEff.</source>
        <translation>Аннотирует и фильтрует вариации %1 при помощи SnpEff.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SnpEffWorker</name>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="97"/>
        <source>SnpEff Annotation and Filtration</source>
        <translation>Аннотация и фильтрация с помощью SnpEff</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="97"/>
        <source>Annotates and filters variations with SnpEff.</source>
        <translation>этот элемент аннотирует и фильтрует вариации при помощи SnpEff.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="101"/>
        <source>Variations</source>
        <translation>Вариации</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="101"/>
        <source>Set of variations</source>
        <translation>Набор вариаций</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="102"/>
        <source>Annotated variations</source>
        <translation>Аннотированные вариации</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="115"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="115"/>
        <source>Select an output folder. &lt;b&gt;Custom&lt;/b&gt; - specify the output folder in the &apos;Custom folder&apos; parameter. &lt;b&gt;Workflow&lt;/b&gt; - internal workflow folder. &lt;b&gt;Input file&lt;/b&gt; - the folder of the input file.</source>
        <translation>выберите выходную папку. &lt;b&gt;Частная&lt;/b&gt; - укажите папку в параметре &apos;Частная папка&apos;. &lt;b&gt;Заданная&lt;/b&gt; - внутренняя папка дизайнера. &lt;b&gt;Входной файл&lt;/b&gt; - папка входного файла.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="119"/>
        <source>Custom folder</source>
        <translation>Частная папка</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="119"/>
        <source>Select the custom output folder.</source>
        <translation>Укажите частную выходную папку.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="121"/>
        <source>Input format</source>
        <translation>Входной формат</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="121"/>
        <source>Select the input format of variations.</source>
        <translation>выберите входной формат вариаций.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="123"/>
        <source>Output format</source>
        <translation>Выходной формат</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="123"/>
        <source>Select the format of annotated output files.</source>
        <translation>выберите формат выходных аннотированных файлов.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="125"/>
        <source>Genome</source>
        <translation>Геном</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="125"/>
        <source>Select the target genome. Genome data will be downloaded if it is not found.</source>
        <translation>выберите референсный геном. Если он не найден, данные генома будут загружены из сети Интернет.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="127"/>
        <source>Upstream/downstream length</source>
        <translation>Верхняя/нижняя длина</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="127"/>
        <source>Upstream and downstream interval size. Eliminate any upstream and downstream effect by using 0 length</source>
        <translation>верхний и нижний размеры интервала</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="129"/>
        <source>Canonical transcripts</source>
        <translation>Канонические транскрипты</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="129"/>
        <source>Use only canonical transcripts</source>
        <translation>использовать только канонические транскрипты</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="131"/>
        <source>HGVS nomenclature</source>
        <translation>Номенклатура HGVS</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="131"/>
        <source>Annotate using HGVS nomenclature</source>
        <translation>аннотировать при помощи номенклатуры HGVS</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="133"/>
        <source>Annotate Loss of function variations</source>
        <translation>Аннотировать варианты loss of function</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="133"/>
        <source>Annotate Loss of function variations (LOF) and Nonsense mediated decay (NMD)</source>
        <translation>аннотирует варианты, связанные с потерей функции (loss of function) и изменением деградации, связанной с преждевременным стоп-кодоном (NMD)</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="135"/>
        <source>Annotate TFBSs motifs</source>
        <translation>Аннотировать мотивы TFBS</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="135"/>
        <source>Annotate transcription factor binding site motifs (only available for latest GRCh37)</source>
        <translation>аннотирует сайты связывания транскрипционных факторов (опция доступна только для последней версии сборки GRCh37)</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="155"/>
        <source>Input file</source>
        <translation>Входной файл</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="156"/>
        <source>Workflow</source>
        <translation>Схема</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffWorker.cpp" line="157"/>
        <source>Custom</source>
        <translation>Частный</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SpadesPrompter</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="608"/>
        <source>Assemble de novo the input data into contigs and scaffolds.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::SpadesWorker</name>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="451"/>
        <source>Scaffolds URL</source>
        <translation>Путь до Scaffolds</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="452"/>
        <source>Output scaffolds URL.</source>
        <translation>Путь до выходногог scaffolds.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="259"/>
        <source>Incorrect input data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="263"/>
        <source>Incorrect port values</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="341"/>
        <source>Some input data elements sent data while some elements already finished their work. Check that all input data elements have the same datasets quantity.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="412"/>
        <source>Input %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="413"/>
        <source>Input %1 to be assembled with SPAdes.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="419"/>
        <source>File URL 1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="420"/>
        <source>File URL 1.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="428"/>
        <source>File URL 2</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="429"/>
        <source>File URL 2.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="455"/>
        <source>Contigs URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="456"/>
        <source>Output contigs URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="460"/>
        <source>Output assembly files.</source>
        <translation>Выходные файлы сборки.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="471"/>
        <source>Input data</source>
        <translation type="unfinished">Входные данные</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="490"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="491"/>
        <source>Folder to save Spades output files.</source>
        <translation>Папка для сохранения выходных файлов Spades.</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="498"/>
        <source>Memory limit</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="503"/>
        <source>Select the input dataset type: standard isolate (the default value) or multiple displacement amplification (corresponds to --sc).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="507"/>
        <source>By default, SPAdes performs both read error correction and assembly. You can select leave one of only (corresponds to --only-assembler, --only-error-correction).&lt;br&gt;&lt;br&gt;                              Error correction is performed using BayesHammer module in case of Illumina input reads and IonHammer in case of IonTorrent data. Note that you should not use error correction                               in case input reads do not have quality information(e.g. FASTA input files are provided).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="559"/>
        <source>Standard isolate</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="560"/>
        <source>MDA single-cell</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="564"/>
        <source>Error correction and assembly</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="574"/>
        <source>In general, SPAdes (St. Petersburg genome assembler) is an assembly toolkit containing various assembly pipelines.                           This workflow element provides GUI for the main SPAdes executable script. One can specify Illumina, IonTorrent or                           PacBio reads as input. Hybrid assemblies are also possible, for example, with Oxford Nanopore or Sanger reads.&lt;br&gt;&lt;br&gt;                          To use the element, configure the type of input in the &quot;Input data&quot; parameter. The corresponding input ports will appear                           on the element. Provide URL(s) to the corresponding FASTA or FASTQ file(s) to these ports.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="459"/>
        <source>Output File</source>
        <translation type="unfinished">Выходной файл</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="494"/>
        <source>Number of threads</source>
        <translation>Число потоков</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="495"/>
        <source>Number of threads (-t).</source>
        <translation>число потоков (-t).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="499"/>
        <source>Memory limit (-m).</source>
        <translation>Предел памяти (-m).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="502"/>
        <source>Dataset type</source>
        <translation>Тип набора данных</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="506"/>
        <source>Running mode</source>
        <translation>Режим запуска</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="512"/>
        <source>K-mers</source>
        <translation>K-mers</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="513"/>
        <source>k-mer sizes (-k).</source>
        <translation>Размеры k-mer (-k).</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="545"/>
        <source>Auto</source>
        <translation type="unfinished">Авто</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="553"/>
        <source> Gb</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="565"/>
        <source>Assembly only</source>
        <translation type="unfinished">Только сборка</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="566"/>
        <source>Error correction only</source>
        <translation type="unfinished">Только исправление ошибок</translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesWorker.cpp" line="573"/>
        <source>Assemble Reads with SPAdes</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::StringTieWorker</name>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="228"/>
        <source>Uses a BAM file with RNA-Seq read mappings to assemble transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="240"/>
        <source>Input URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="241"/>
        <source>URL(s) of input file(s) in FASTA format with DNA sequences that need to be assembled</source>
        <translation type="unfinished">Пути входных файлов в формате FASTA</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="246"/>
        <source>Output URL Transcripts</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="247"/>
        <source>Output URL Transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="249"/>
        <source>Output URL Gene Abundance</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="250"/>
        <source>Output URL Gene Abundance.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="255"/>
        <source>Input BAM file(s)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="256"/>
        <source>URL(s) to sorted BAM file(s) with RNA-Seq read mappings. Note that every spliced read alignment (i.e. an alignment across at least one junction) in the input file must contain the tag XS to indicate the genomic strand that produced the RNA from which the read was sequenced. Alignments produced by TopHat and HISAT2 (when run with --dta option) already include this tag, but if you use a different read mapper you should check that this XS tag is included for spliced alignments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="263"/>
        <source>StringTie output data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="263"/>
        <source>For each input BAM file the port outputs an URL to a GTF file with assembled transcripts, produced by StringTie. If &quot;Report gene abundance&quot; is &quot;True&quot;, the port also output an URL to a text file with gene abundances (in a tab-delimited format).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="277"/>
        <source>Reference annotations</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="278"/>
        <source>Use the reference annotation file (in GTF or GFF3 format) to guide the assembly process (-G). The output will include expressed reference transcripts as well as any novel transcripts that are assembled.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="282"/>
        <source>Reads orientation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="283"/>
        <source>Select the NGS libraries type: unstranded, stranded fr-secondstrand (--fr), or stranded fr-firststand (--rf).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="286"/>
        <source>Label</source>
        <translation type="unfinished">Ярлык</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="287"/>
        <source>Use the specified string as the prefix for the name of the output transcripts (-l).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="290"/>
        <source>Min isoform fraction</source>
        <translation type="unfinished">Минимальное содержание изоформ</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="291"/>
        <source>Specify the minimum isoform abundance of the predicted transcripts as a fraction of the most abundant transcript assembled at a given locus (-f). Lower abundance transcripts are often artifacts of incompletely spliced precursors of processed transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="296"/>
        <source>Min assembled transcript length</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="297"/>
        <source>Specify the minimum length for the predicted transcripts (-m).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="300"/>
        <source>Min anchor length for junctions</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="301"/>
        <source>Junctions that don&apos;t have spliced reads that align across them with at least this amount of bases on both sides are filtered out (-a).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="305"/>
        <source>Min junction coverage</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="306"/>
        <source>There should be at least this many spliced reads that align across a junction (-j). This number can be fractional, since some reads align in more than one place. A read that aligns in n places will contribute 1/n to the junction coverage.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="311"/>
        <source>Trim transcripts based on coverage</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="312"/>
        <source>By default StringTie adjusts the predicted transcript&apos;s start and/or stop coordinates based on sudden drops in coverage of the assembled transcript. Set this parameter to &quot;False&quot; to disable the trimming at the ends of the assembled transcripts (-t).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="317"/>
        <source>Min coverage for assembled transcripts</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="318"/>
        <source>Specify the minimum read coverage allowed for the predicted transcripts (-c). A transcript with a lower coverage than this value is not shown in the output. This number can be fractional, since some reads align in more than one place. A read that aligns in n places will contribute 1/n to the coverage.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="324"/>
        <source>Min locus gap separation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="325"/>
        <source>Reads that are mapped closer than this distance are merged together in the same processing bundle (-g).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="328"/>
        <source>Fraction covered by multi-hit reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="329"/>
        <source>Specify the maximum fraction of muliple-location-mapped reads that are allowed to be present at a given locus (-M). A read that aligns in n places will contribute 1/n to the coverage.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="333"/>
        <source>Skip assembling for sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="334"/>
        <source>Ignore all read alignments (and thus do not attempt to perform transcript assembly) on the specified reference sequences (-x). The value can be a single reference sequence name (e.g. &quot;chrM&quot;) or a comma-delimited list of sequence names (e.g. &quot;chrM,chrX,chrY&quot;). This can speed up StringTie especially in the case of excluding the mitochondrial genome, whose genes may have very high coverage in some cases, even though they may be of no interest for a particular RNA-Seq analysis. The reference sequence names are case sensitive, they must match identically the names of chromosomes/contigs of the target genome against which the RNA-Seq reads were aligned in the first place.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="343"/>
        <source>Abundance for reference transcripts only</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="344"/>
        <source>Limits the processing of read alignments to only estimate and output the assembled transcripts matching the reference transcripts (-e). With this option, read bundles with no reference transcripts will be entirely skipped, which may provide a considerable speed boost when the given set of reference transcripts is limited to a set of target genes, for example. The parameter is only available if the &quot;Reference annotations&quot; file is specified. It is recommended to use it when Ballgown table files are produced.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="351"/>
        <source>Multi-mapping correction</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="352"/>
        <source>Enables or disables (-u) multi-mapping correction.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="355"/>
        <source>Verbose log</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="356"/>
        <source>Enable detailed logging, if required (-v). The messages will be written to the UGENE log (enabling of &quot;DETAILS&quot; and &quot;TRACE&quot; logging may be required) and to the dashboard.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="361"/>
        <source>Number of threads</source>
        <translation type="unfinished">Число потоков</translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="362"/>
        <source>Specify the number of processing threads (CPUs) to use for transcript assembly (-p).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="365"/>
        <source>Output transcripts file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="366"/>
        <source>StringTie&apos;s primary output GTF file with assembled transcripts.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="369"/>
        <source>Enable gene abundance output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="370"/>
        <source>Select &quot;True&quot; to generate gene abundances output (-A). The output is written to a tab-delimited text file. Also, the file URL is passed to an output slot of the workflow element.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="375"/>
        <source>Output gene abundances file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="376"/>
        <source>Specify the name of the output file with gene abundances (-A).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="379"/>
        <source>Enable covered reference transcripts output</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="380"/>
        <source>Select &quot;True&quot; to generate a file with reference transcripts that are fully covered by reads (-C). Thus, the parameter is only available if the &quot;Reference annotations&quot; file is specified.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="384"/>
        <source>Output covered reference transcripts file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="385"/>
        <source>Specify the name of the output file with reference transcripts that are fully covered by reads (-C).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="388"/>
        <source>Enable output for Ballgown</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="389"/>
        <source>Select &quot;True&quot; to generate table files (*.ctab) that can be used as input to Ballgown (-b). The files contain coverage data for the reference transcripts. The parameter is only available if the &quot;Reference annotations&quot; file is specified. It is also recommended to set &quot;Abundance for reference transcripts only&quot; to &quot;True&quot;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="395"/>
        <source>Output folder for Ballgown</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="396"/>
        <source>Specify a folder for table files (*.ctab) that can be used as input to Ballgown.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="453"/>
        <source>Unstranded</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="454"/>
        <source>Forward (FR)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="455"/>
        <source>Reverse (RF)</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="515"/>
        <source>Enabled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="516"/>
        <source>Disabled</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="544"/>
        <source>Assemble Transcripts with StringTie</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringTieWorker.cpp" line="544"/>
        <source>StringTie is a fast and highly efficient assembler of RNA-Seq alignments into potential transcripts. It uses a novel network flow algorithm as well as an optional de novo assembly step to assemble and quantitate full-length transcripts representing multiple splice variants for each gene locus.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::StringtieGeneAbundanceReportPrompter</name>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportPrompter.cpp" line="32"/>
        <source>Create a general report for several StringTie gene abundance output files.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::StringtieGeneAbundanceReportTask</name>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="53"/>
        <source>StringTie Gene Abundance Report Task</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="76"/>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="93"/>
        <source>The directory &quot;%1&quot; did not created</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="136"/>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="143"/>
        <source>Too long line while reading a file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="148"/>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="260"/>
        <source>Bad line format of input: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="170"/>
        <source>Cannot open a file: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="229"/>
        <source>Unexpected error while parsing input data</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="236"/>
        <source>Bad file format, there is no %2 column: &quot;%1&quot;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportTask.cpp" line="254"/>
        <source>Cannot open a file: %1
Error is :</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::StringtieGeneAbundanceReportWorker</name>
    <message>
        <location filename="../src/stringtie/StringtieGeneAbundanceReportWorker.cpp" line="59"/>
        <source>An empty URL to StringTie report passed to the &apos;%1&apos;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeePrompter</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="132"/>
        <source> from %1</source>
        <translation> из %1</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="134"/>
        <source>For each MSA&lt;u&gt;%1&lt;/u&gt;, build the alignment using &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt; and send it to output.</source>
        <translation>Для каждого MSA&lt;u&gt;%1&lt;/u&gt;, строится выравнивание при помощи &lt;u&gt;&quot;T-Coffee&quot;&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TCoffeeWorker</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="63"/>
        <source>Input MSA</source>
        <translation>Входное MSA</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="63"/>
        <source>Multiple sequence alignment to be processed.</source>
        <translation>Множественное выравнивание, которое будет обработано.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="64"/>
        <source>Multiple sequence alignment</source>
        <translation>Множественное выравнивание</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="64"/>
        <source>Result of alignment.</source>
        <translation>Результат выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="73"/>
        <source>Gap Open Penalty</source>
        <translation>Штраф за открытие пробела</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="73"/>
        <source>Gap Open Penalty. Must be negative, best matches get a score of 1000.</source>
        <translation>Штраф за открытие пробела. Должно быть отрицательным, лучшие совпадения оцениваются в 1000.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="74"/>
        <source>Gap Extension Penalty</source>
        <translation>Штраф за расширение пробела</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="74"/>
        <source>Gap Extension Penalty. Positive values give rewards to gaps and prevent the alignment of unrelated segments.</source>
        <translation>Штраф за расширение пробела. Положительные значения дают дополнительные очки для пробелов и предотвращают выравнивание не связанных между собой сегментов.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="75"/>
        <source>Max Iteration</source>
        <translation>Количество итераций</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="75"/>
        <source>Number of iteration on the progressive alignment.&lt;br&gt;0 - no iteration, -1 - Nseq iterations.</source>
        <translation>Число итераций.&lt;br&gt;0 - нет итераций, -1 - итераций Nseq.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="77"/>
        <source>Tool Path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="77"/>
        <source>External tool path.</source>
        <translation>Путь до внешнего инструмента.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="178"/>
        <source>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</source>
        <translation>An empty MSA &apos;%1&apos; has been supplied to T-Coffee.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="78"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="78"/>
        <source>folder for temporary file.s</source>
        <translation>Папка для временных файлов</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="86"/>
        <source>Align with T-Coffee</source>
        <translation>Выравнивание с помощью T-Coffee</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="86"/>
        <source>T-Coffee is a multiple sequence alignment package. </source>
        <translation>&lt;i&gt;T-Coffee&lt;/i&gt; это пакет для множественного выравнивания. </translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeWorker.cpp" line="210"/>
        <source>Aligned %1 with T-Coffee</source>
        <translation>Aligned %1 with T-Coffee</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ToPhred33Step</name>
    <message>
        <location filename="../src/trimmomatic/steps/ToPhred33Step.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;TOPHRED33&lt;/h4&gt;&lt;p&gt;This step (re)encodes the quality part of the FASTQ file to base 33.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::ToPhred64Step</name>
    <message>
        <location filename="../src/trimmomatic/steps/ToPhred64Step.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;TOPHRED64&lt;/h4&gt;&lt;p&gt;This step (re)encodes the quality part of the FASTQ file to base 64.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TopHatPrompter</name>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="547"/>
        <source> to reference sequence with index &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="550"/>
        <source> to reference sequence &lt;u&gt;%1&lt;/u&gt;.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TopHatWorker</name>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="114"/>
        <location filename="../src/tophat/TopHatWorker.cpp" line="125"/>
        <source>Input reads</source>
        <translation>Входные риды</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="115"/>
        <source>Input RNA-Seq reads</source>
        <translation>Входные риды RNA-Seq</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="119"/>
        <source>TopHat output</source>
        <translation>Выходные данные TopHat</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="120"/>
        <source>Accepted hits, junctions, insertions and deletions</source>
        <translation>Допустимы хиты, переходы, вставки и удаления</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="126"/>
        <source>TopHat input reads. Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation>Входные риды TopHat. Оставьте этот слот пустым если вы хотите выровнять риды напрямую из файла и укажите слот &quot;Путь для входных ридов&quot;.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="131"/>
        <source>Input reads url</source>
        <translation>Путь для входных ридов</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="132"/>
        <source>TopHat input reads url. When running TopHat with paired-end reads, this should be the *_1 (&quot;left&quot;) set of reads.</source>
        <translation>Путь для входных ридов TopHat.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="136"/>
        <source>Input paired reads</source>
        <translation>Входные парные риды</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="137"/>
        <source> Set this slot empty if you want to align reads directly from a file and specify the &quot;Input reads url&quot; slot. Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads. Reads MUST appear in the same order as the *_1 reads.</source>
        <translation>Оставьте этот слот пустым если вы хотите выровнять риды напрямую из файла и укажите слот &quot;Путь для входных ридов&quot;.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="144"/>
        <source>Input paired reads url</source>
        <translation>Путь для входных парных ридов</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="145"/>
        <source>Only used when running TopHat with paired end reads, and contains the *_2 (&quot;right&quot;) set of reads.</source>
        <translation>Используется только когда TopHat запускается для парных ридов, и содержит набор ридов *_2 (&quot;right&quot;).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="150"/>
        <source>Dataset name</source>
        <translation>Имя набора данных</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="151"/>
        <source>Use it only when sequences slot(or slots) is specified. Group input reads into chunks for several Tophat runs.
Set it empty if you want to run Tophat once for all input reads</source>
        <translation>Используйте этот параметр только когда указан слот типа последовательность. Оставьте его пустым если хотите запустить задачу для всех входных ридов</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="170"/>
        <source>Accepted hits</source>
        <translation>Принимаемые хиты</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="171"/>
        <source>Accepted hits found by TopHat</source>
        <translation>Принимаемые хиты найденные TopHat</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="173"/>
        <source>Sample name</source>
        <translation>Пример имени</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="174"/>
        <source>Sample name for running Cuffdiff</source>
        <translation>Пример имени для запуска Cuffdiff</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="176"/>
        <source>Accepted hits url</source>
        <translation>Путь до принимаемых хитов</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="177"/>
        <source>The url to the assembly file with the accepted hits</source>
        <translation>Путь до файла сборки с принимаемыми хитами</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="188"/>
        <source>Map RNA-Seq Reads with TopHat</source>
        <translation>Поиск сайтов сплайсинга с помощью TopHat</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="189"/>
        <source>TopHat is a program for mapping RNA-Seq reads to a long reference sequence. It uses Bowtie or Bowtie2 to map the reads and then analyzes the mapping results to identify splice junctions between exons.&lt;br/&gt;&lt;br/&gt;Provide URL(s) to FASTA or FASTQ file(s) with NGS RNA-Seq reads to the input port of the element, set up the reference sequence in the parameters. The result is saved to the specified BAM file, URL to the file is passed to the output port. Several UCSC BED tracks are also produced: junctions, insertions, and deletions.</source>
        <translation>TopHat является быстрым инструментом для определения сплайсинга по данным секвенирования РНК. TopHat выравнивает риды с помощью Bowtie и затем анализирует результаты выравнивания для определения границ экзонов.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="200"/>
        <source>Output folder</source>
        <translation>Выходная папка</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="201"/>
        <source>The base name of output folder. It could be modified with a suffix.</source>
        <translation>Базовое имя для выходной папки.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="204"/>
        <source>Samples map</source>
        <translation>Карта образцов</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="205"/>
        <source>The map which divide all input datasets into samples. Every sample has the unique name.</source>
        <translation>карта, которая разделяет все входные наборы данных на образцы. Каждый образец имеет уникальное название.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="208"/>
        <source>Reference input type</source>
        <translation>Тип референса</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="209"/>
        <source>Select &quot;Sequence&quot; to input a reference genome as a sequence file. &lt;br/&gt;Note that any sequence file format, supported by UGENE, is allowed (FASTA, GenBank, etc.). &lt;br/&gt;The index will be generated automatically in this case. &lt;br/&gt;Select &quot;Index&quot; to input already generated index files, specific for the tool.</source>
        <translation>Выберите &quot;Последовательность&quot; чтобы ввести референсный геном как файл. &lt;br/&gt;Любой формат файла, который поддерживается UGENE допустим (FASTA, GenBank, и т.д.). &lt;br/&gt;Индекс будет сгенерирован автоматически. &lt;br/&gt;Выберите &quot;Индекс&quot; чтобы ввести уже готовый файл с индексом.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="215"/>
        <source>Reference genome</source>
        <translation>Референсный геном</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="216"/>
        <source>Path to indexed reference genome.</source>
        <translation>Путь до индексированного референсного генома.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="219"/>
        <source>Bowtie index folder</source>
        <translation>Папка для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="220"/>
        <source>The folder with the Bowtie index for the reference sequence.</source>
        <translation>папка, в которой расположен индекс Bowtie для референсной последовательности.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="228"/>
        <source>Bowtie index basename</source>
        <translation>Базовое имя для индекса Bowtie</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="229"/>
        <source>The basename of the Bowtie index for the reference sequence.</source>
        <translation>базовое имя для индекса Bowtie для референсной последовательности.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="242"/>
        <source>Mate inner distance</source>
        <translation>Внутренняя дистанция в паре</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="243"/>
        <source>The expected (mean) inner distance between mate pairs.</source>
        <translation>ожидаемая (средняя) внутренняя дистанция между ридами в паре (mate pair).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="246"/>
        <source>Mate standard deviation</source>
        <translation>Стандартное отклонение в паре</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="247"/>
        <source>The standard deviation for the distribution on inner distances between mate pairs.</source>
        <translation>стандартное отклонение для распределения в внутренних дистанций между ридами в паре (mate pair).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="250"/>
        <source>Library type</source>
        <translation>Тип библиотеки</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="251"/>
        <source>Specifies RNA-Seq protocol.</source>
        <translation>задает протокол секвенирования РНК.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="254"/>
        <source>No novel junctions</source>
        <translation>Не искать новые сайты сплайсинга</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="255"/>
        <source>Only look for reads across junctions indicated in the supplied GFF or junctions file. This parameter is ignored if &lt;i&gt;Raw junctions&lt;/i&gt; or &lt;i&gt;Known transcript file&lt;/i&gt; is not set.</source>
        <translation>Искать только такие сплайсированные риды, которые соответствуют информации в приложенном файле (GFF или файле с описанием сайтов сплайсинга). Если ни один из этих файлов не указан, параметр будет проигнорирован.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="260"/>
        <source>Raw junctions</source>
        <translation>Новые сайты сплайсинга</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="261"/>
        <source>The list of raw junctions.</source>
        <translation>список обнаруженных сайтов сплайсинга.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="264"/>
        <source>Known transcript file</source>
        <translation>Файл с известными транскриптами</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="265"/>
        <source>A set of gene model annotations and/or known transcripts.</source>
        <translation>файл с аннотацией моделей генов и/или известными транскриптами.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="268"/>
        <source>Max multihits</source>
        <translation>Максимальное количество множественных выравниваний</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="269"/>
        <source>Instructs TopHat to allow up to this many alignments to the reference for a given read, and suppresses all alignments for reads with more than this many alignments.</source>
        <translation>Tophat не будет анализировать все выравнивания для ридов с большим количеством выравниваний, чем указанный параметр.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="274"/>
        <source>Segment length</source>
        <translation>Длина сегмента</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="275"/>
        <source>Each read is cut up into segments, each at least this long. These segments are mapped independently.</source>
        <translation>каждый рид разрезается на сегменты указанной длины. Эти сегменты выравниваются независимо.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="287"/>
        <source>Fusion search</source>
        <translation>Поиск химер</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="288"/>
        <source>Turn on fusion mapping.</source>
        <translation>искать выравнивание для химерных ридов.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="291"/>
        <source>Transcriptome only</source>
        <translation>Только транскриптом</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="292"/>
        <source>Only align the reads to the transcriptome and report only those mappings as genomic mappings.</source>
        <translation>выравнивать риды только на транскриптом и выдавать только эти выравнивания при выравнивании на геном.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="296"/>
        <source>Transcriptome max hits</source>
        <translation>Максимальное число выравниваний</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="297"/>
        <source>Maximum number of mappings allowed for a read, when aligned to the transcriptome (any reads found with more than this number of mappings will be discarded).</source>
        <translation>максимальное число выравниваний, разрешенных для рида при выравнивании на транскриптом (риды с большим числом выравниваний будут исключены из дальнейшего анализа).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="302"/>
        <source>Prefilter multihits</source>
        <translation>Предварительный фильтр множественных выравниваний</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="303"/>
        <source>When mapping reads on the transcriptome, some repetitive or low complexity reads that would be discarded in the context of the genome may appear to align to the transcript sequences and thus may end up reported as mapped to those genes only. This option directs TopHat to first align the reads to the whole genome in order to determine and exclude such multi-mapped reads (according to the value of the &lt;i&gt;Max multihits&lt;/i&gt; option).</source>
        <translation>при выравнивании ридов на транскриптом, некоторые повторяющиеся риды или риды, содержащие повторы, могут быть выявлены как выровненные, хотя при выравнивании на геном они были бы отброшены. При выборе этой опции TopHat сначала выравнивает риды на весь геном, чтобы определить и исключить риды, которые выравниваются несколько раз (число выравниваний определяется значением опции).</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="312"/>
        <source>Min anchor length</source>
        <translation>Минимальная длина якоря</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="313"/>
        <source>The &lt;i&gt;anchor length&lt;/i&gt;. TopHat will report junctions spanned by reads with at least this many bases on each side of the junction. Note that individual spliced alignments may span a junction with fewer than this many bases on one side. However, every junction involved in spliced alignments is supported by at least one read with this many bases on each side.</source>
        <translation>&lt;i&gt;длина якоря&lt;/i&gt;. TopHat будет указывать в отчете сайты сплайсинга в том случае, если с каждой стороны от стыка прочитано не менее указанного числа нуклеотидов. В отдельных выравниваниях число нуклеотидов может быть меньше, однако для любого указанного в отчете сайта сплайсинга будет существовать хотя бы одно выравнивание с не менее чем указанным числом нуклеотидов с каждой стороны.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="321"/>
        <source>Splice mismatches</source>
        <translation>Замены около сайтов сплайсинга</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="322"/>
        <source>The maximum number of mismatches that may appear in the &lt;i&gt;anchor&lt;/i&gt; region of a spliced alignment.</source>
        <translation>максимально допустимое число несоответствий в нуклеотидах около сайтов сплайсинга..</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="326"/>
        <source>Read mismatches</source>
        <translation>Замены в ридах</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="327"/>
        <source>Final read alignments having more than these many mismatches are discarded.</source>
        <translation>финальные выравнивания, число несоответствий в которых больше указанного, будут отброшены.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="331"/>
        <source>Segment mismatches</source>
        <translation>Замены в сегментах</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="332"/>
        <source>Read segments are mapped independently, allowing up to this many mismatches in each segment alignment.</source>
        <translation>сегменты рида выравниваются независимо, и при этом при выравнивании каждого сегмента допускается указанное максимальное число несоответствий.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="337"/>
        <source>Solexa 1.3 quals</source>
        <translation>Качество в шкале Solexa 1.3</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="338"/>
        <source>As of the Illumina GA pipeline version 1.3, quality scores are encoded in Phred-scaled base-64. Use this option for FASTQ files from pipeline 1.3 or later.</source>
        <translation>в пайплайне Illumina GA версии 1.3 показатели качества кодируются в шкале Phred64. Используйте эту опцию для FASTQ файлов из пайплайна 1.3 или более поздних версии.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="343"/>
        <source>Bowtie version</source>
        <translation>Версия Bowtie</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="344"/>
        <source>Specifies which Bowtie version should be used.</source>
        <translation>указывает, какую версию Bowtie необходимо использовать.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="347"/>
        <source>Bowtie -n mode</source>
        <translation>Режим Bowtie -n</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="348"/>
        <source>TopHat uses &lt;i&gt;-v&lt;/i&gt; in Bowtie for initial read mapping (the default), but with this option, &lt;i&gt;-n&lt;/i&gt; is used instead. Read segments are always mapped using &lt;i&gt;-v&lt;/i&gt; option.</source>
        <translation>по умолчанию TopHat использует -v в Bowtie для выравнивания ридов, но при выборе этой опции будет использован параметр -n. Выравнивание сегментов ридов всегда осуществляется с параметром -v.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="354"/>
        <source>Bowtie tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="355"/>
        <source>The path to the Bowtie external tool.</source>
        <translation>Путь до внешнего инструмента Bowtie в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="358"/>
        <source>SAMtools tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="359"/>
        <source>The path to the SAMtools tool. Note that the tool is available in the UGENE External Tool Package.</source>
        <translation>Путь до инструмента SAMtools. Обратите внимание, что этот доступен в пакете внешних инструментов UGENE.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="363"/>
        <source>TopHat tool path</source>
        <translation>Путь</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="364"/>
        <source>The path to the TopHat external tool in UGENE.</source>
        <translation>Путь до внешнего инструмента TopHat в UGENE.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="367"/>
        <source>Temporary folder</source>
        <translation>Временная папка</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="368"/>
        <source>The folder for temporary files.</source>
        <translation>Папка для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="422"/>
        <source>Sequence</source>
        <translation>Последовательность</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="423"/>
        <source>Index</source>
        <translation>Индекс</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="430"/>
        <source>Use -n mode</source>
        <translation>Использовать -n моду</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="431"/>
        <source>Use -v mode</source>
        <translation>Использовать -v моду</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="542"/>
        <source>Maps RNA-seq reads</source>
        <translation>Выравнивает риды RNA-seq на референсную последовательность</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="553"/>
        <source> and finds splice junctions.</source>
        <translation> и ищет регионы сплайсинга.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="621"/>
        <source>Incorrect value of the library type parameter for Cufflinks!</source>
        <translation>Incorrect value of the library type parameter for Cufflinks!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="650"/>
        <source>Unrecognized value of the Bowtie mode option!</source>
        <translation>Unrecognized value of the Bowtie mode option!</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TrailingStep</name>
    <message>
        <location filename="../src/trimmomatic/steps/TrailingStep.cpp" line="36"/>
        <source>&lt;html&gt;&lt;head&gt;&lt;/head&gt;&lt;body&gt;&lt;h4&gt;TRAILING&lt;/h4&gt;&lt;p&gt;This step removes low quality bases from the end. As long as a base has a value below this threshold the base is removed and the next base (i.e. the preceding one) will be investigated. This approach can be used removing the special Illumina &quot;low quality segment&quot; regions (which are marked with quality score of 2), but SLIDINGWINDOW or MAXINFO are recommended instead.&lt;/p&gt;&lt;p&gt;Input the following values:&lt;/p&gt;&lt;ul&gt;&lt;li&gt;&lt;b&gt;Quality threshold&lt;/b&gt;: the minimum quality required to keep a base.&lt;/li&gt;&lt;/ul&gt;&lt;/body&gt;&lt;/html&gt;</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/steps/TrailingStep.cpp" line="52"/>
        <source>The minimum quality required to keep a base.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TrimmomaticPropertyDialog</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticDelegate.cpp" line="186"/>
        <source>Apply</source>
        <translation>Применить</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::TrimmomaticWorker</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorker.cpp" line="119"/>
        <source>Copy adapters to working folder</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticWorker.cpp" line="150"/>
        <source>Process &quot;Trimmomatic&quot; with one dataset</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusPrompter</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="175"/>
        <source>Apply VCF variants from &lt;u&gt;%1&lt;/u&gt; to fasta file &lt;u&gt;%2&lt;/u&gt; and save consensus sequence to &lt;u&gt;%3&lt;/u&gt;.</source>
        <translation>Применяются вариации VCF из &lt;u&gt;%1&lt;/u&gt; к файлу в формате fasta&lt;u&gt;%2&lt;/u&gt;, консенсус сохраняется в &lt;u&gt;%3&lt;/u&gt;.</translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::VcfConsensusWorker</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="76"/>
        <source>Input fasta slot is empty</source>
        <translation>Input fasta slot is empty</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="79"/>
        <source>Input vcf slot is empty</source>
        <translation>Input vcf slot is empty</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="126"/>
        <source>Input FASTA and VCF</source>
        <translation>Входные FASTA и VCF</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="127"/>
        <source>FASTA url</source>
        <translation>Путь до FASTA</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="128"/>
        <source>VCF url</source>
        <translation>Путь до VCF</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="134"/>
        <source>Fasta consensus url</source>
        <translation>Путь до консенсуса в формате Fasta</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="142"/>
        <source>Output FASTA consensus</source>
        <translation>Записать консенсус в формате FASTA</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="142"/>
        <source>The path to the output file with the result consensus.</source>
        <translation>путь до выходного файла с консенсусом.</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="152"/>
        <source>Create VCF Consensus</source>
        <translation>Создание консенсуса с помощью VCF</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusWorker.cpp" line="153"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation>этот элемент применяет список вариаций в формате VCF к файлу в формате FASTA для создания консенсуса.</translation>
    </message>
</context>
<context>
    <name>U2::MACSSupport</name>
    <message>
        <location filename="../src/macs/MACSSupport.cpp" line="45"/>
        <source>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, which analyzes datagenerated by short read sequencers such as Solexa&apos;s Genome Analyzer. MACS empirically models the shiftsize of ChIP-Seq tags, and uses it to improve the spatial resolution of predicted binding sites.MACS also uses a dynamic Poisson distribution to effectively capture local biases in the genome,allowing for more robust predictions. MACS compares favorably to existing ChIP-Seq peak-finding algorithms, and is freely available.</source>
        <translation>&lt;i&gt;MACS&lt;/i&gt; - Model-based Analysis of ChIP-Seq data, который анализирует данные полученные из коротких ридов подобно Solexa&apos;s Genome Analyzer. MACS эмпирически моделирует ChIP-Seq ярлыки, и использует их для улучшения пространственного разрешения прогнозируемых обязательных участков.</translation>
    </message>
</context>
<context>
    <name>U2::MACSTask</name>
    <message>
        <location filename="../src/macs/MACSTask.cpp" line="53"/>
        <source>MACS peak calling</source>
        <translation>MACS peak calling</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTLogParser</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="408"/>
        <source>MAFFT has switched to the memsave mode. UGENE is unable to track its progress.</source>
        <translation>MAFFT has switched to the memsave mode. UGENE is unable to track its progress.</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupport</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="67"/>
        <source>&lt;i&gt;MAFFT&lt;/i&gt; is a multiple sequence alignment program for unix-like operating systems. </source>
        <translation>&lt;i&gt;MAFFT&lt;/i&gt;это программа для выравнивания множественных наборов последовательностей.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="79"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="80"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportContext</name>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="134"/>
        <source>Align with MAFFT...</source>
        <translation>Выравнивание с помощью MAFFT...</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="159"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupport.cpp" line="160"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="45"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="46"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTSupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="96"/>
        <source>MAFFT alignment started</source>
        <translation>MAFFT alignment started</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="117"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="126"/>
        <source>Subdir for temporary files exists. Can not remove this folder.</source>
        <translation>Subdir for temporary files exists. Can not remove this folder.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="131"/>
        <source>Can not create folder for temporary files.</source>
        <translation>Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="144"/>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="146"/>
        <source>Can not open output file: </source>
        <translation>Can not open output file: </translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="146"/>
        <source> May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation> May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="181"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation>Output file &apos;%1&apos; not found</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="183"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="191"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="266"/>
        <source>Failed to apply the result of aligning with MAFFT: alignment object is not available!</source>
        <translation>Failed to apply the result of aligning with MAFFT: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="271"/>
        <source>MAFFT alignment successfully finished</source>
        <translation>MAFFT alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="290"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="74"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="75"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="84"/>
        <source>Open an alignment file</source>
        <translation>Открыть выравнивание</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="97"/>
        <source>Save an multiple alignment file</source>
        <translation>Сохранить выравнивание</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="116"/>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="118"/>
        <source>Kalign with Align</source>
        <translation>Выравнивание с помощью Kalign</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="116"/>
        <source>Input file is not set!</source>
        <translation>Не указан входной файл!</translation>
    </message>
    <message>
        <location filename="../src/mafft/MAFFTSupportRunDialog.cpp" line="118"/>
        <source>Output file is not set!</source>
        <translation>Не указан выходной файл!</translation>
    </message>
</context>
<context>
    <name>U2::MAFFTWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/mafft/MAFFTSupportTask.cpp" line="323"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::MafftAddToAlignmentTask</name>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="68"/>
        <source>Align sequences to alignment task</source>
        <translation>Align sequences to alignment task</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="102"/>
        <source>Align sequences to an existing alignment by MAFFT started</source>
        <translation>Align sequences to an existing alignment by MAFFT started</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="177"/>
        <source>Output file &apos;%1&apos; not found</source>
        <translation>Output file &apos;%1&apos; not found</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="179"/>
        <source>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file &apos;%3&apos; not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="186"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="265"/>
        <source>Row for updating doesn&apos;t found</source>
        <translation>Row for updating doesn&apos;t found</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="282"/>
        <source>Additional enhancement of short sequences alignment performed</source>
        <translation>Additional enhancement of short sequences alignment performed</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="291"/>
        <source>The following sequence(s) were not aligned as they do not contain meaningful characters: &quot;%1&quot;.</source>
        <translation>The following sequence(s) were not aligned as they do not contain meaningful characters: &quot;%1&quot;.</translation>
    </message>
    <message>
        <location filename="../src/mafft/MafftAddToAlignmentTask.cpp" line="303"/>
        <source>MAFFT alignment successfully finished</source>
        <translation>MAFFT alignment successfully finished</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="194"/>
        <source>Generating output trees from MrBayes</source>
        <translation>Создание выходных дереьев черезе MrBayes</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="203"/>
        <source>Output file is not found</source>
        <translation>Output file is not found</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="230"/>
        <source>No trees are found</source>
        <translation>Деревья не найдены</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesPrepareDataForCalculation</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="41"/>
        <source>Generating input file for MrBayes</source>
        <translation>Создание входного файла для MRBayes</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupport</name>
    <message>
        <location filename="../src/mrbayes/MrBayesSupport.cpp" line="62"/>
        <source>&lt;i&gt;MrBayes&lt;/i&gt; is a program for the Bayesian estimation of phylogeny.Bayesian inference of phylogeny is based upon a quantity called the posterior probability distribution of trees, which is the probability of a tree conditioned on the observations. The conditioning is accomplished using Bayes&apos;s theorem. The posterior probability distribution of trees is impossible to calculate analytically; instead, MrBayes uses a simulation technique called Markov chain Monte Carlo (or MCMC) to approximate the posterior probabilities of trees.</source>
        <translation>&lt;i&gt;MrBayes&lt;/i&gt; это программа для байесовского анализа и выбора модели в широком диапазоне филогенетических и эволюционных моделей. MrBayes использует методы Монте-Карло для оценки распределения параметров модели.</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesSupportTask</name>
    <message>
        <location filename="../src/mrbayes/MrBayesTask.cpp" line="87"/>
        <source>MrBayes tree calculation</source>
        <translation>MrBayes вычисление деревьев</translation>
    </message>
</context>
<context>
    <name>U2::MrBayesWidget</name>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="65"/>
        <source>Substitution model</source>
        <translation>Модель замен</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="69"/>
        <source>Rate Matrix (fixed)</source>
        <translation>Матрица оценок</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="180"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/mrbayes/MrBayesDialogWidget.cpp" line="181"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::PathLineEdit</name>
    <message>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="850"/>
        <location filename="../src/ExternalToolSupportSettingsController.cpp" line="852"/>
        <source>Select a file</source>
        <translation>Выберите файл</translation>
    </message>
</context>
<context>
    <name>U2::Peak2GeneFormatLoader</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneFormatLoader.cpp" line="105"/>
        <source>Incorrect start position at line %1: &apos;%2&apos;</source>
        <translation>Incorrect start position at line %1: &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneFormatLoader.cpp" line="111"/>
        <source>Incorrect end position at line %1: &apos;%2&apos;</source>
        <translation>Incorrect end position at line %1: &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneFormatLoader.cpp" line="129"/>
        <source>Incorrect peak score at line %1: &apos;%2&apos;</source>
        <translation>Incorrect peak score at line %1: &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/peak2gene/Peak2GeneFormatLoader.cpp" line="139"/>
        <source>Incorrect NA value at line %1: &apos;%2&apos;</source>
        <translation>Incorrect NA value at line %1: &apos;%2&apos;</translation>
    </message>
</context>
<context>
    <name>U2::Peak2GeneSupport</name>
    <message>
        <location filename="../src/peak2gene/Peak2GeneSupport.cpp" line="52"/>
        <source>&lt;i&gt;peak2gene&lt;/i&gt; - Gets refGenes near the ChIP regions identified by a peak-caller</source>
        <translation>&lt;i&gt;peak2gene&lt;/i&gt; - получает refGenes рядом с регионами ChIP идентифицированных пиками</translation>
    </message>
</context>
<context>
    <name>U2::PerlSupport</name>
    <message>
        <location filename="../src/perl/PerlSupport.cpp" line="47"/>
        <source>Perl scripts interpreter</source>
        <translation>Perl scriptsинтерпретатор</translation>
    </message>
</context>
<context>
    <name>U2::PhmmerSearchDialog</name>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="67"/>
        <source>Search</source>
        <translation>Искать</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="68"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="122"/>
        <source>Select query sequence file</source>
        <translation>Выберите файл последовательности</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="169"/>
        <source>Query sequence file path is empty</source>
        <translation>Укажите последовательность</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="187"/>
        <source>Error</source>
        <translation>Ошибка</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="187"/>
        <source>Cannot create an annotation object. Please check settings</source>
        <translation>Невозможно создать аннотацию. Проверьте настройки</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchDialog.cpp" line="194"/>
        <source>Error: bad arguments!</source>
        <translation>Ошибка: некорректные аргументы!</translation>
    </message>
</context>
<context>
    <name>U2::PhmmerSearchTask</name>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="50"/>
        <source>Search with phmmer</source>
        <translation>Поиск при помощи phmmer</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="104"/>
        <source>Query sequence: </source>
        <translation>Последовательность:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="107"/>
        <source>Task was not finished</source>
        <translation>Task was not finished</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="113"/>
        <source>Result annotation table: </source>
        <translation>Таблица аннотаций:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="115"/>
        <source>Result annotation group: </source>
        <translation>Группа аннотаций:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="116"/>
        <source>Result annotation name: </source>
        <translation>Имя аннотации:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="118"/>
        <source>Results count: </source>
        <translation>Количество:</translation>
    </message>
    <message>
        <location filename="../src/hmmer/PhmmerSearchTask.cpp" line="150"/>
        <source>Cannot create a folder for temporary files.</source>
        <translation>Cannot create a folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLGetCalculatedTreeTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="192"/>
        <source>Generating output trees from PhyML</source>
        <translation>Generating output trees from PhyML</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="201"/>
        <source>Output file is not found</source>
        <translation>Output file is not found</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="227"/>
        <source>No trees are found</source>
        <translation>Деревья не найдены</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="232"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="238"/>
        <source>No result tree in PhyML output</source>
        <translation>No result tree in PhyML output</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLLogParser</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="171"/>
        <source>PhyML finished with error</source>
        <translation>PhyML finished with error</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="187"/>
        <source>UGENE internal error</source>
        <translation>UGENE internal error</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLPrepareDataForCalculation</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="41"/>
        <source>Generating input file for PhyML</source>
        <translation>Generating input file for PhyML</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="67"/>
        <location filename="../src/phyml/PhyMLTask.cpp" line="71"/>
        <source>Internal UGENE error</source>
        <translation>Internal UGENE error</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="75"/>
        <source>Can not open tmp file</source>
        <translation>Can not open tmp file</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupport</name>
    <message>
        <location filename="../src/phyml/PhyMLSupport.cpp" line="51"/>
        <source>&lt;i&gt;PhyML&lt;/i&gt; is a simple, fast, and accurate algorithm to estimate large phylogenies by maximum likelihood</source>
        <translation>&lt;i&gt;PhyML&lt;/i&gt; это простой, быстрый и точный алгоритм для построения больших филогенетических деревьев методом максимального правдоподобия</translation>
    </message>
</context>
<context>
    <name>U2::PhyMLSupportTask</name>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="93"/>
        <source>PhyML tree calculation</source>
        <translation>PhyML tree calculation</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLTask.cpp" line="143"/>
        <source>UGENE internal error</source>
        <translation>UGENE internal error</translation>
    </message>
</context>
<context>
    <name>U2::PhyMlWidget</name>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="234"/>
        <source>Open an alignment file</source>
        <translation>Открыть выравнивание</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="290"/>
        <source>File with the starting tree is not set.</source>
        <translation>Файл с начальным деревом не указан.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="292"/>
        <source>File with the starting tree doesn&apos;t exist.</source>
        <translation>Файла со стартовым деревом не существует.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="308"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/phyml/PhyMLDialogWidget.cpp" line="309"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputFastaFilesTask</name>
    <message>
        <location filename="../src/blast_plus/PrepareInputFastaFilesTask.cpp" line="36"/>
        <source>Prepare input FASTA files</source>
        <translation>Prepare input FASTA files</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/PrepareInputFastaFilesTask.cpp" line="89"/>
        <source>File &apos;%1&apos; was skipped. Cannot detect the file format.</source>
        <translation>File &apos;%1&apos; was skipped. Cannot detect the file format.</translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForCAP3Task</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="231"/>
        <source>Failed to initialize sequence writer.</source>
        <translation>Failed to initialize sequence writer.</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="257"/>
        <source>Failed to write sequence %1</source>
        <translation>Failed to write sequence %1</translation>
    </message>
</context>
<context>
    <name>U2::PrepareInputForSpideyTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="191"/>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="204"/>
        <source>Failed to write DNA sequence  %1</source>
        <translation>Failed to write DNA sequence  %1</translation>
    </message>
</context>
<context>
    <name>U2::PrepareReferenceSequenceTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/PrepareReferenceSequenceTask.cpp" line="46"/>
        <source>Prepare reference sequence</source>
        <translation>Prepare reference sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/PrepareReferenceSequenceTask.cpp" line="85"/>
        <source>No reference sequence in the file: </source>
        <translation>No reference sequence in the file: </translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/PrepareReferenceSequenceTask.cpp" line="86"/>
        <source>More than one sequence in the reference file: </source>
        <translation>More than one sequence in the reference file: </translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/PrepareReferenceSequenceTask.cpp" line="89"/>
        <source>Unable to cast gobject to sequence object</source>
        <translation>Unable to cast gobject to sequence object</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/PrepareReferenceSequenceTask.cpp" line="90"/>
        <source>The input reference sequence &apos;%1&apos; contains characters that don&apos;t belong to DNA alphabet.</source>
        <translation>The input reference sequence &apos;%1&apos; contains characters that don&apos;t belong to DNA alphabet.</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleBioSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="126"/>
        <source> (or biopython) is a python module for biological computations.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleDjangoSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="104"/>
        <source>: Python module for the %1 tool</source>
        <translation>: Python модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleNumpySupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="113"/>
        <source>: Python module for the %1 tool</source>
        <translation>: Python модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::PythonModuleSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="93"/>
        <source>Python module is not installed. Install module or set path to another Python scripts interpreter with installed module in the External Tools settings</source>
        <translation>Python модуль не установлен. Установите модель или укажите другой путь с установленным модулем в настройках внешних инструментов</translation>
    </message>
</context>
<context>
    <name>U2::PythonSupport</name>
    <message>
        <location filename="../src/python/PythonSupport.cpp" line="66"/>
        <source>Python scripts interpreter</source>
        <translation>Python интерпретатор</translation>
    </message>
</context>
<context>
    <name>U2::RModuleCelegansdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="184"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleDrosophila2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="191"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleGodbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="142"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleGostatsSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="135"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133adbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="149"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133bdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="156"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu133plus2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="163"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleHgu95av2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="170"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleMouse430a2dbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="177"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgceegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="212"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgdmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="219"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrghsegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="198"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleOrgmmegdbSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="205"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleSeqlogoSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="226"/>
        <source>: Rscript module for the %1 tool</source>
        <translation>: Rscript модуль для инструмента %1</translation>
    </message>
</context>
<context>
    <name>U2::RModuleSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="120"/>
        <source>R module is not installed. Install module or set path to another R scripts interpreter with installed module in the External Tools settings</source>
        <translation>R модуль не установлен. Установите модель или укажите другой путь с установленным модулем в настройках внешних инструментов</translation>
    </message>
</context>
<context>
    <name>U2::RSupport</name>
    <message>
        <location filename="../src/R/RSupport.cpp" line="90"/>
        <source>Rscript interpreter</source>
        <translation>Rscript интерпретатор</translation>
    </message>
</context>
<context>
    <name>U2::RegisterCustomToolTask</name>
    <message>
        <location filename="../src/custom_tools/RegisterCustomToolTask.cpp" line="36"/>
        <source>Register custom external tool</source>
        <translation>Register custom external tool</translation>
    </message>
    <message>
        <location filename="../src/custom_tools/RegisterCustomToolTask.cpp" line="55"/>
        <source>Can&apos;t register a custom external tool &apos;%1&apos;</source>
        <translation>Can&apos;t register a custom external tool &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::RemoveGapsFromSequenceTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/RemoveGapsFromSequenceTask.cpp" line="65"/>
        <source>Remove gaps from the sequence</source>
        <translation>Remove gaps from the sequence</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/RemoveGapsFromSequenceTask.cpp" line="81"/>
        <source>Find gaps in the sequence</source>
        <translation>Find gaps in the sequence</translation>
    </message>
</context>
<context>
    <name>U2::RunCap3AndOpenResultTask</name>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="128"/>
        <source>CAP3 run and open result task</source>
        <translation>CAP3 run and open result task</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="136"/>
        <source>Invalid CAP3 task</source>
        <translation>Invalid CAP3 task</translation>
    </message>
    <message>
        <location filename="../src/cap3/CAP3SupportTask.cpp" line="151"/>
        <source>Project loader is NULL</source>
        <translation>Project loader is NULL</translation>
    </message>
</context>
<context>
    <name>U2::SaveAlignmentTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="62"/>
        <source>Export alignment to &apos;%1&apos;</source>
        <translation>Export alignment to &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="66"/>
        <source>An alignment is empty</source>
        <translation>An alignment is empty</translation>
    </message>
</context>
<context>
    <name>U2::SaveMSA2SequencesTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="103"/>
        <source>Export alignment to sequence: %1</source>
        <translation>Export alignment to sequence: %1</translation>
    </message>
</context>
<context>
    <name>U2::SaveSequenceTask</name>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="133"/>
        <source>Save sequence</source>
        <translation>Save sequence</translation>
    </message>
    <message>
        <location filename="../src/utils/ExportTasks.cpp" line="163"/>
        <source>&apos;%&apos; format is not registered</source>
        <translation>&apos;%&apos; format is not registered</translation>
    </message>
</context>
<context>
    <name>U2::SeqPosSupport</name>
    <message>
        <location filename="../src/seqpos/SeqPosSupport.cpp" line="51"/>
        <source>&lt;i&gt;SeqPos&lt;/i&gt; - Finds motifs enriched in a set of regions.</source>
        <translation>&lt;i&gt;SeqPos&lt;/i&gt; - ищет мотивы, обогащенные в наборе регионов.</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffDatabaseListModel</name>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseListModel.cpp" line="97"/>
        <source>Genome</source>
        <translation>Геном</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseListModel.cpp" line="99"/>
        <source>Organism</source>
        <translation>Организм</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffDatabaseListTask</name>
    <message>
        <location filename="../src/snpeff/SnpEffDatabaseListTask.cpp" line="38"/>
        <source>SnpEff Database List task</source>
        <translation>SnpEff Database List task</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffParser</name>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="60"/>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="82"/>
        <source>A problem occurred during allocating memory for running SnpEff. Check the &quot;Tasks memory limit&quot; parameter in the UGENE Application Settings.It is recommended to set this value to the available RAM on the computer.</source>
        <translation>A problem occurred during allocating memory for running SnpEff. Check the &quot;Tasks memory limit&quot; parameter in the UGENE Application Settings.It is recommended to set this value to the available RAM on the computer.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="76"/>
        <source>There is not enough memory to complete the SnpEff execution.It is recommended to run SnpEff on a computer with RAM 4Gb or more.</source>
        <translation>There is not enough memory to complete the SnpEff execution.It is recommended to run SnpEff on a computer with RAM 4Gb or more.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="87"/>
        <source>Failed to download SnpEff database. Check your internet connection.</source>
        <translation>Failed to download SnpEff database. Check your internet connection.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="92"/>
        <source>Genome database &apos;%1&apos; is not found.</source>
        <translation>Genome database &apos;%1&apos; is not found.</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffSupport</name>
    <message>
        <location filename="../src/snpeff/SnpEffSupport.cpp" line="56"/>
        <source>&lt;i&gt;SnpEff&lt;/i&gt;: Genetic variant annotation and effect prediction toolbox.</source>
        <translation>&lt;i&gt;SnpEff&lt;/i&gt;: инструмент генетического варианта аннотации и прогнозирования эффекта.</translation>
    </message>
    <message>
        <location filename="../src/snpeff/SnpEffSupport.cpp" line="101"/>
        <source>Failed to get SnpEff database list</source>
        <translation>Failed to get SnpEff database list</translation>
    </message>
</context>
<context>
    <name>U2::SnpEffTask</name>
    <message>
        <location filename="../src/snpeff/SnpEffTask.cpp" line="206"/>
        <source>SNPEff dataDir is not initialized.</source>
        <translation>SNPEff dataDir is not initialized.</translation>
    </message>
</context>
<context>
    <name>U2::SpadesSupport</name>
    <message>
        <location filename="../src/spades/SpadesSupport.cpp" line="43"/>
        <source>&lt;i&gt;SPAdes&lt;/i&gt; - St. Petersburg genome assembler - is intended for both standard isolates and single-cell MDA bacteria assemblies. Official site: http://bioinf.spbau.ru/spades</source>
        <translation>&lt;i&gt;SPAdes&lt;/i&gt; - геномный сборщик, разработанный в Санкт-Петербурге. Официальный сайт: http://bioinf.spbau.ru/spades</translation>
    </message>
</context>
<context>
    <name>U2::SpadesTask</name>
    <message>
        <location filename="../src/spades/SpadesTask.cpp" line="60"/>
        <source>Folder does not exist: </source>
        <translation>Folder does not exist: </translation>
    </message>
    <message>
        <location filename="../src/spades/SpadesTask.cpp" line="123"/>
        <location filename="../src/spades/SpadesTask.cpp" line="130"/>
        <source>File %1 has not been found in output folder %2</source>
        <translation>File %1 has not been found in output folder %2</translation>
    </message>
</context>
<context>
    <name>U2::SpideyAlignmentTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="81"/>
        <source>Output file not found</source>
        <translation>Выходной файл не найден</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="86"/>
        <source>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="98"/>
        <source>Failed to open result file %1</source>
        <translation>Failed to open result file %1</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupport</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="78"/>
        <source>&lt;i&gt;Spidey&lt;/i&gt; is mRNA-to-DNA alignment program.                    &lt;br&gt;Binaries can be downloaded from http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</source>
        <translation>&lt;i&gt;Spidey&lt;/i&gt;это программа для выранивания mRNA-to-DNA.                    &lt;br&gt;Программа может быть скачана http://www.ncbi.nlm.nih.gov/spidey/spideyexec.html</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="80"/>
        <source>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</source>
        <translation>&lt;br&gt;&lt;br&gt; Wheelan SJ, Church DM, Ostell JM.</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="81"/>
        <source>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</source>
        <translation>&lt;br&gt;Spidey: a tool for mRNA-to-genomic alignments</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="82"/>
        <source>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</source>
        <translation>&lt;br&gt;Genome Res. 2001 Nov;11(11):1952-7.</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportContext</name>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="115"/>
        <source>Align sequence to mRNA</source>
        <translation>Выровнять последовательность на мРНК</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="143"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/spidey/SpideySupport.cpp" line="144"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::SpideySupportTask</name>
    <message>
        <location filename="../src/spidey/SpideySupportTask.cpp" line="244"/>
        <source>Failed to align mRNA to genomic sequence: no alignment is found.</source>
        <translation>Failed to align mRNA to genomic sequence: no alignment is found.</translation>
    </message>
</context>
<context>
    <name>U2::StringTieSupport</name>
    <message>
        <location filename="../src/stringtie/StringTieSupport.cpp" line="34"/>
        <source>&lt;i&gt;StringTie&lt;/i&gt; is a fast and highly efficient assembler of RNA-Seq alignments into potential transcripts. It uses a novel network flow algorithm as well as an optional de novo assembly step to assemble and quantitate full-length transcripts representing multiple splice variants for each gene locus. Its input can include not only the alignments of raw reads used by other transcript assemblers, but also alignments longer sequences that have been assembled from those reads.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::StringTieTask</name>
    <message>
        <location filename="../src/stringtie/StringTieTask.cpp" line="51"/>
        <source>Assemble Transcripts with StringTie task</source>
        <translation>Assemble Transcripts with StringTie task</translation>
    </message>
</context>
<context>
    <name>U2::TBlastXPlusSupportTask</name>
    <message>
        <location filename="../src/blast_plus/TBlastXPlusSupportTask.cpp" line="63"/>
        <source>Unexpected settings combination</source>
        <translation>Unexpected settings combination</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupport</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="71"/>
        <source>&lt;i&gt;T-Coffee&lt;/i&gt; is a multiple sequence alignment package.</source>
        <translation>&lt;i&gt;T-Coffee&lt;/i&gt; это пакет для множественного выравнивания.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="81"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="82"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportContext</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="136"/>
        <source>Align with T-Coffee...</source>
        <translation>Выравнивание с помощью T-Coffee...</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="161"/>
        <source>Path for %1 tool not selected.</source>
        <translation>Путь для %1 не установлен.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupport.cpp" line="162"/>
        <source>Do you want to select it now?</source>
        <translation>Выбрать сейчас?</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="46"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="47"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeSupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="98"/>
        <source>Unsupported alphabet: %1</source>
        <translation>Unsupported alphabet: %1</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="102"/>
        <source>T-Coffee alignment started</source>
        <translation>T-Coffee alignment started</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="123"/>
        <source>Saving data to temporary file &apos;%1&apos;</source>
        <translation>Saving data to temporary file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="132"/>
        <source>Subdir for temporary files exists. Can not remove this folder.</source>
        <translation>Subdir for temporary files exists. Can not remove this folder.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="137"/>
        <source>Can not create folder for temporary files.</source>
        <translation>Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="183"/>
        <source>Output file %1 not found</source>
        <translation>Output file %1 not found</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="185"/>
        <source>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</source>
        <translation>Output file %3 not found. May be %1 tool path &apos;%2&apos; not valid?</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="193"/>
        <source>Loading output file &apos;%1&apos;</source>
        <translation>Loading output file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="272"/>
        <source>Failed to apply the result of TCoffee: alignment object is not available!</source>
        <translation>Failed to apply the result of TCoffee: alignment object is not available!</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="277"/>
        <source>T-Coffee alignment successfully finished</source>
        <translation>T-Coffee alignment successfully finished</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="290"/>
        <source>Can not remove folder for temporary files.</source>
        <translation>Can not remove folder for temporary files.</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportRunDialog</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="74"/>
        <source>Align</source>
        <translation>Выровнять</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="75"/>
        <source>Cancel</source>
        <translation>Отмена</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="84"/>
        <source>Open an alignment file</source>
        <translation>Открыть выравнивание</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="97"/>
        <source>Save an multiple alignment file</source>
        <translation>Сохранить выравнивание</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="116"/>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="118"/>
        <source>Kalign with Align</source>
        <translation>Выравнивание с помощью Kalign</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="116"/>
        <source>Input file is not set!</source>
        <translation>Не указан входной файл!</translation>
    </message>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportRunDialog.cpp" line="118"/>
        <source>Output file is not set!</source>
        <translation>Не указан выходной файл!</translation>
    </message>
</context>
<context>
    <name>U2::TCoffeeWithExtFileSpecifySupportTask</name>
    <message>
        <location filename="../src/tcoffee/TCoffeeSupportTask.cpp" line="323"/>
        <source>Unrecognized input alignment file format</source>
        <translation>Unrecognized input alignment file format</translation>
    </message>
</context>
<context>
    <name>U2::TabixSupport</name>
    <message>
        <location filename="../src/samtools/TabixSupport.cpp" line="52"/>
        <source>&lt;i&gt;Tabix&lt;/i&gt; is a generic indexer for TAB-delimited genome position files</source>
        <translation>&lt;i&gt;Tabix&lt;/i&gt;это индексатор для TAB-файлов с разделителями позиции генома</translation>
    </message>
</context>
<context>
    <name>U2::TabixSupportTask</name>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="38"/>
        <source>Generate index with Tabix task</source>
        <translation>Generate index with Tabix task</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="47"/>
        <source>Tabix indexing started</source>
        <translation>Tabix indexing started</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="50"/>
        <source>Input file &apos;%1&apos; is already bgzipped</source>
        <translation>Input file &apos;%1&apos; is already bgzipped</translation>
    </message>
    <message>
        <location filename="../src/samtools/TabixSupportTask.cpp" line="61"/>
        <source>Saving data to file &apos;%1&apos;</source>
        <translation>Saving data to file &apos;%1&apos;</translation>
    </message>
</context>
<context>
    <name>U2::TopHatSupportTask</name>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="53"/>
        <source>Running TopHat task</source>
        <translation>Running TopHat task</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="112"/>
        <source>Can&apos;t create directory for index files </source>
        <translation>Can&apos;t create directory for index files </translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="183"/>
        <source>An unexpected error has occurred during preparing the TopHat task!</source>
        <translation>An unexpected error has occurred during preparing the TopHat task!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="184"/>
        <source>Preparing TopHatSupportTask internal error: unable to get a sequence object!</source>
        <translation>Preparing TopHatSupportTask internal error: unable to get a sequence object!</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="312"/>
        <source>TopHat was not able to map reads to the reference.</source>
        <translation>TopHat was not able to map reads to the reference.</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatSupportTask.cpp" line="336"/>
        <source>There are no accepted hits in the result</source>
        <translation>There are no accepted hits in the result</translation>
    </message>
</context>
<context>
    <name>U2::TrimmomaticSupport</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticSupport.cpp" line="50"/>
        <source>&lt;i&gt;Trimmomatic&lt;/i&gt; is a flexible read trimming tool for Illumina NGS data.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::TrimmomaticTask</name>
    <message>
        <location filename="../src/trimmomatic/TrimmomaticTask.cpp" line="43"/>
        <source>Improve reads with Trimmomatic</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::UserAppsSettings</name>
    <message>
        <location filename="../src/ExternalToolSupportSettings.cpp" line="150"/>
        <source>Temporary UGENE dir is empty</source>
        <translation>Temporary UGENE dir is empty</translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupport</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupport.cpp" line="51"/>
        <source>Apply VCF variants to a fasta file to create consensus sequence.</source>
        <translation>Применяет VCF вариации для файла fasta для создания консенсуса.</translation>
    </message>
</context>
<context>
    <name>U2::VcfConsensusSupportTask</name>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="40"/>
        <source>Create Vcf Consensus</source>
        <translation>Create Vcf Consensus</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="49"/>
        <source>VcfConsensus started</source>
        <translation>VcfConsensus started</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="51"/>
        <source>AppSettings is NULL</source>
        <translation>AppSettings is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="53"/>
        <source>UserAppsSettings is NULL</source>
        <translation>UserAppsSettings is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="55"/>
        <source>Temporary folder is not set!</source>
        <translation>Temporary folder is not set!</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="60"/>
        <source>Can not create folder for temporary files.</source>
        <translation>Не удаётся создать папку для временных файлов.</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="64"/>
        <source>Saving temporary data to file &apos;%1&apos;</source>
        <translation>Saving temporary data to file &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="83"/>
        <source>ExternalToolRegistry is NULL</source>
        <translation>ExternalToolRegistry is NULL</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="87"/>
        <source>There is no VcfConsensus external tool registered</source>
        <translation>There is no VcfConsensus external tool registered</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="88"/>
        <source>There is no Tabix external tool registered</source>
        <translation>There is no Tabix external tool registered</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="109"/>
        <source>Trying to get path of NULL external tool</source>
        <translation>Trying to get path of NULL external tool</translation>
    </message>
    <message>
        <location filename="../src/vcftools/VcfConsensusSupportTask.cpp" line="113"/>
        <source>Path to %1</source>
        <translation>Path to %1</translation>
    </message>
</context>
<context>
    <name>U2::Workflow::BlastAndSwReadTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="89"/>
        <source>Map one read with BLAST &amp; SW task</source>
        <translation>Map one read with BLAST &amp; SW task</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="120"/>
        <source>A problem occurred while mapping &quot;%1&quot; to &quot;%2&quot;.</source>
        <translation>A problem occurred while mapping &quot;%1&quot; to &quot;%2&quot;.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="130"/>
        <source>%1 was skipped. No BLAST results.</source>
        <translation>%1 was skipped. No BLAST results.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="174"/>
        <source>%1 was skipped. Low similarity: %2. Minimum similarity was set to %3</source>
        <translation>%1 was skipped. Low similarity: %2. Minimum similarity was set to %3</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="275"/>
        <source>Read doesn&apos;t contain meaningful data</source>
        <translation>Read doesn&apos;t contain meaningful data</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="358"/>
        <source>The %1 algorithm is not found. Add the %1 plugin.</source>
        <translation>The %1 algorithm is not found. Add the %1 plugin.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="361"/>
        <source>The %1 algorithm is not found. Check that the %1 plugin is up to date.</source>
        <translation>The %1 algorithm is not found. Check that the %1 plugin is up to date.</translation>
    </message>
</context>
<context>
    <name>U2::Workflow::BlastReadsSubTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="53"/>
        <source>Map reads with BLAST &amp; SW task</source>
        <translation>Map reads with BLAST &amp; SW task</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp" line="65"/>
        <source>The task uses a temporary folder to process the data. The folder path is required not to have spaces. Please set up an appropriate path for the &quot;Temporary files&quot; parameter on the &quot;Directories&quot; tab of the UGENE Application Settings.</source>
        <translation>The task uses a temporary folder to process the data. The folder path is required not to have spaces. Please set up an appropriate path for the &quot;Temporary files&quot; parameter on the &quot;Directories&quot; tab of the UGENE Application Settings.</translation>
    </message>
</context>
<context>
    <name>U2::Workflow::ComposeResultSubTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/ComposeResultSubTask.cpp" line="65"/>
        <source>Compose alignment</source>
        <translation>Compose alignment</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/ComposeResultSubTask.cpp" line="197"/>
        <source>No read satisfy minimum similarity criteria.</source>
        <translation>No read satisfy minimum similarity criteria.</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/ComposeResultSubTask.cpp" line="311"/>
        <source>The related chromatogram not found</source>
        <translation>The related chromatogram not found</translation>
    </message>
</context>
<context>
    <name>U2::Workflow::FormatDBSubTask</name>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/FormatDBSubTask.cpp" line="47"/>
        <source>Format DB task wrapper</source>
        <translation>Format DB task wrapper</translation>
    </message>
    <message>
        <location filename="../src/blast_plus/align_worker_subtasks/FormatDBSubTask.cpp" line="65"/>
        <source>The task uses a temporary folder to process the data. It is required that the folder path doesn&apos;t have spaces. Please set up an appropriate path for the &quot;Temporary files&quot; parameter on the &quot;Directories&quot; tab of the UGENE Application Settings.</source>
        <translation>The task uses a temporary folder to process the data. It is required that the folder path doesn&apos;t have spaces. Please set up an appropriate path for the &quot;Temporary files&quot; parameter on the &quot;Directories&quot; tab of the UGENE Application Settings.</translation>
    </message>
</context>
<context>
    <name>U2::Workflow::IntegralBusPort</name>
    <message>
        <location filename="../src/cufflinks/CuffdiffWorker.cpp" line="57"/>
        <source>&apos;%1&apos; slot must be not binded</source>
        <translation>&apos;%1&apos; слот не должен быть связан</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="60"/>
        <source>One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>Один из этих слотов не должен быть пустым: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/cufflinks/CufflinksWorker.cpp" line="65"/>
        <source>Only one of these slots must be binded: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>Только один из этих слотов должен быть связан: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/macs/MACSWorker.cpp" line="230"/>
        <location filename="../src/utils/BaseShortReadsAlignerWorker.cpp" line="269"/>
        <source>The slot must be not empty: &apos;%1&apos;</source>
        <translation>Слот не должен быть пустым: &apos;%1&apos;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="782"/>
        <source>Error! One of these slots must be not empty: &apos;%1&apos;, &apos;%2&apos;</source>
        <translation>Ошибка! Один из этих слотов не должен быть пустым: &apos;%1&apos;, &apos;%2&apos;</translation>
    </message>
    <message>
        <location filename="../src/tophat/TopHatWorker.cpp" line="788"/>
        <source>Error! You can not bind one of sequences slots and one of url slots simultaneously</source>
        <translation>Ошибка! Вы не можете связать слот последовательности и слот пути одновременно</translation>
    </message>
</context>
</TS>
